#### Figure 1: Accumulation of nonsynonymous S1 mutations is correlated with clade success
**A** For every clade in the phylogeny, mutations relative to the root of the phylogeny are tallied and plotted against the date of the base of that clade. Nonsynonymous S1, synonymous S1, and nonsynonymous RdRp mutations are plotted separately. Nonsynonymous mutations include nonsynonymous SNPs and deletions. The primary axis (left, black ticks) displays mutations per codon, and the secondary axis (right, gray ticks) shows the absolute number of mutations accumulated in each clade. Each point is colored according to the lineage it belongs to. Points are fit by linear regression. **B** For every clade, mutation accumulation (as in A) is plotted against logistic growth rate and the points are fit by linear regression. **C** The R-value of the correlation between mutation accumulation and logistic growth rate is compared to an expected distribution to yield a p-value. Expected R-values are determined from randomizing mutations across the phylogeny using a multinomial draw with mutation likelihood proportional to relative branch length. The results of 1000 iterations are shown.

#### Figure 1 supplement 1: Phylogeny of 9544 SARS-CoV-2 genomes
Screenshot of the phylogeny used for the primary analyses in this manuscript. Tips and branches are colored according to emerging lineage. Emerging lineages are labeled by PANGO lineage and WHO Variant of Interest (VOI) or Variant of Concern (VOC) designation. An interactive version of this phylogeny can be accessed at nextstrain.org/groups/blab/ncov/adaptive-evolution/2021-05-15

#### Figure 1 supplement 2: Correlation between nonsynonymous mutation accumulation and clade success is strongest in S1
Nonsynonymous mutation accumulation (mutations per codon) is plotted against logistic growth rate for 8 genes (or subunits), as in Fig1B. 

#### Figure 1 supplement 3: Deletions contribute to protein-coding changes in S1, N and Nsp6
For each gene nonsynonymous mutation accumulation is separated into nonsynonymous SNPs (left) and deletions (right). Accumulation of these mutations is plotted against logistic growth rate for 8 genes (or subunits), as in Fig1B.


 <br /> <br /> <br />
#### Figure 2: Ratio of nonsynonymous to synonymous divergence is highest S1
The ratio of nonsynonymous to synonymous divergence within various coding regions is plotted over time. Each point shows the mean and 95% confidence interval of this ratio for all isolates present in a 0.2 year window (ending at the date indicated on the x-axis). Nonsynonymous divergence is calculated as the nonsynonymous Hamming distance between the sequence of an isolate and the root, normalized by the total possible number of nonsynonymous sites. The same is done for synonymous divergence. Divergence is calculated for various locations within the genome.

#### Figure 2 supplement 1: Ratio of nonsynonymous to synonymous divergence in influenza H3N2
The mean and 95% confidence intervals for nonsynonymous/synonymous divergence ratios in the H3N2 genes HA1 and PB1 are shown over a 12-year period. 


 <br /> <br /> <br />
#### Figure 3: S1 substitutions are temporally clustered
**A** Time line showing accumulation of S1 nonsynonymous, synonymous, and RdRp nonsynonymous mutations between the root and a representative tip (isolate USA/ME-HETL-J3202/2021) with wait times between mutations illustrated in below. The exact date of a mutation is chosen by randomly selecting a date along the branch the mutation occurs on. **B** Distribution of wait times between S1 nonsynonymous, S1 synonymous, and RdRp nonsynonymous mutations. Empirical wait times (in color) are plotted along with expected wait times (gray). Expected wait times determined from randomizing mutations across the phylogeny using a multinomial draw with mutation likelihood proportional to relative branch length. The results of 1000 iterations are shown. **C** ***i*** The mean empirical wait time from 1000 iterations of the analysis (colored bar) is compared to the distribution of mean expected wait times (gray) to yield a p-value. ***ii*** The proportion of observed wait times under 0.3 years (colored bar) is compared to the distribution of expected wait times under 0.3 years (gray).

#### Figure 3 supplement 1: Temporal accumulation of S1 mutations on representative paths through the tree.
The total number of accumulated S1 nonsynonymous mutations is counted at every branch along a path through the tree. This is plotted for 3 representative paths from the root to an isolate in an emerging lineage clade. The isolate and emerging lineage are labeled on each panel.

#### Figure 3 supplement 2: Distribution of expected wait times is affected by the number of mutations that occur across the phylogeny.
The phylogeny was randomized with varying numbers of mutations to display the expected wait time distributions if 50, 100, 200, 300, 400 or 500 mutations occur on internal branches of the phylogeny. Each randomization is run for 10 iterations. The empirical number of S1 nonsynonymous, S1 synonymous, and RdRp nonsynonymous mutations observed on internal branches of the phylogeny are indicated.

 <br /> <br /> <br />
#### Figure 4: A 3-amino acid deletion in Nsp6 displays convergent evolution and occurs in successful clades
**A** Every mutation observed on internal branches of the phylogeny is plotted according to the number of times this mutation occurs on the tree and the mean logistic growth rate of all clades it occurs in. Convergently-evolved mutations that appear 4 or more times across the phylogeny are shown in the inset. The average growth rate of all clades is shown with a dotted line. **B** Observed mean growth rates of convergently-evolved mutations are compared to the mean growth rate expected for a mutation occurring on the phylogeny the same number of times. Convergently-evolved mutations that have a mean growth rate falling at or above the 90th percentile of the expected values are labeled. **C** The analysis shown in panel A was completed for 5 time points, spanning two months. Each date represents the maximum date of sequences included in the analysis. Mutations that occur at least 4 times (convergent mutations) and result in a higher-than-average mean growth rate are shown in dark yellow. Mutations that display convergent evolution but do not result in high growth rates are in light yellow. The primary analysis was done using 2021-05-15 (outlined in red). 

#### Figure 4 supplement 1: Every occurrence of the 3-amino acid deletion in Nsp6 resulted in an emerging lineage
Every occurrence of the convergently-evolved mutations is colored according to the emerging lineage it occurs at the base of. Multiple emerging lineages descending from the branch a mutation occurs on is represented by light purple.

#### Figure 4 supplement 2: Analyses of convergent evolution shown 1 month before and 1 month after the primary analysis
**A** Same as Fig4A, completed using sequences up to April 15, 2021 (1 month before the primary analysis). **B** Same as Fig4B, completed using sequences up to April 15, 2021. **C** Same as Fig4A, completed using sequences up to June 15, 2021 (1 month after the primary analysis). **D** Same as Fig4B, completed using sequences up to June 15, 2021.

 <br /> <br /> <br />
#### Figure 5: Clades with the 3-amino acid deletion in Nsp6 have a high number of S1 mutations.
**A** The mean number of S1 mutations (top), or RdRp mutations (bottom), that occur in clades that have (blue/orange) or do not have (gray) the 3-amino acid deletion in Nsp6. The expected difference is shown on the left, and empirical data is shown on the right. Expectation is based off of 100 randomizations of the tree. Error bars shown standard deviation. **B** The difference in the number of nonsynonymous S1 (dark blue), S1 synonymous (light blue), and nonsynonymous RdRp (orange) mutations that occur before versus after a convergently-evolved mutation is shown. Error bars show 95% confidence intervals. 
