# Adaptive Evolution of SARS-CoV-2
**Kathryn Kistler** <sup>1,2</sup>, **John Huddleston** <sup>1,2</sup> , **Trevor Bedford** <sup>1,2</sup>  
<sup>1</sup> *Molecular and Cellular Biology Program, University of Washington, Seattle, United States*  
<sup>2</sup> *Vaccine and Infectious Disease Division, Fred Hutchinson Cancer Research Center, Seattle, United States*

## Abstract
Despite the appearance of variant SARS-CoV-2 viruses with altered receptor-binding or antigenic phenotypes, traditional methods for detecting adaptive evolution from sequence data do not pick up strong signals of positive selection. Here, we present a new method for identifying adaptive evolution on short evolutionary time scales with densely sampled populations and apply it to SARS-CoV-2 to present a comprehensive analysis of adaptively-evolving regions of the genome. We find that spike S1 is a focal point of adaptive evolution, but also identify positively-selected mutations in other genes that are sculpting the evolutionary trajectory of SARS-CoV-2. Protein-coding mutations in S1 are temporally-clustered and, in 2021, the ratio of nonsynonymous to synonymous divergence in S1 is more than 4 times greater than in the equivalent influenza HA1 subunit.
