{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "be28518f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import math\n",
    "import json\n",
    "import string\n",
    "import requests\n",
    "from augur.utils import json_to_tree\n",
    "from os import path\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Align import AlignInfo\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f1f9c70c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Download tree json\n",
    "# tree_url = \"https://data.nextstrain.org/ncov_gisaid_global.json\"\n",
    "tree_url = \"https://data.nextstrain.org/ncov_global_2021-05-19.json\"\n",
    "# tree_url = \"https://data.nextstrain.org/ncov_global_2021-05-28.json\"\n",
    "\n",
    "tree_json = requests.get(tree_url).json()\n",
    "\n",
    "#Put tree in Bio.Phylo format\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9af645b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def divergence_weighted(cov, gene, window, clade, min_seqs, year_max=None, year_min=None):\n",
    "    #Find fraction of sites that differ from root and average among all viruses at each time point\n",
    "    input_file_root = '../'+str(cov)+'/auspice/seasonal_corona_'+str(cov)+'_'+str(gene)+'_root-sequence.json'\n",
    "    input_file_alignment = '../'+str(cov)+'/results/aligned_'+str(cov)+'_'+str(gene)+'.fasta'\n",
    "    metafile = '../'+str(cov)+'/results/metadata_'+str(cov)+'_'+str(gene)+'.tsv'\n",
    "    \n",
    "    #Subset data based on time windows\n",
    "    meta = pd.read_csv(metafile, sep = '\\t')\n",
    "    meta.drop(meta[meta['date']=='?'].index, inplace=True)\n",
    "    meta.dropna(subset=['date'], inplace=True)\n",
    "    meta['year'] = meta['date'].str[:4].astype('int')\n",
    "    \n",
    "    if year_max:\n",
    "        meta.drop(meta[meta['year']>year_max].index, inplace=True)\n",
    "    if year_min:\n",
    "        meta.drop(meta[meta['year']<year_min].index, inplace=True)\n",
    "\n",
    "    \n",
    "    date_range = meta['year'].max() - meta['year'].min()\n",
    "    \n",
    "    \n",
    "    if clade!= None:\n",
    "        clade_df = separate_clades(cov, gene)\n",
    "        meta = meta.merge(clade_df, on='strain')\n",
    "        meta.drop(meta[meta['clade']!=clade].index, inplace=True)\n",
    "\n",
    "    #Group viruses by time windows\n",
    "    virus_time_subset = {}\n",
    "    if window == 'all':\n",
    "        years = str(meta['year'].min()) + '-' + str(meta['year'].max())\n",
    "        virus_time_subset[years] = meta['strain'].tolist()\n",
    "    else:\n",
    "        date_window_start = meta['year'].min()\n",
    "        date_window_end = meta['year'].min() + window\n",
    "        while date_window_end <= meta['year'].max():\n",
    "            years = str(date_window_start) + '-' + str(date_window_end)\n",
    "            strains = meta[(meta['year']>=date_window_start) & (meta['year']<date_window_end)]['strain'].tolist()\n",
    "            virus_time_subset[years] = strains\n",
    "            #sliding window\n",
    "            date_window_end += 1\n",
    "            date_window_start += 1     \n",
    "    \n",
    "    root_seq, root_aa_seq, first_window_years = find_founder_consensus(virus_time_subset,input_file_alignment, min_seqs)\n",
    "    \n",
    "    nonsyn_denominator, syn_denominator = find_nonsyn_syn_denominators(root_seq, root_aa_seq, cov, clade)\n",
    "\n",
    "    \n",
    "    #initiate lists to record all time windows\n",
    "    year_windows = []\n",
    "    seqs_in_window = []\n",
    "    nonsyn_divergences = []\n",
    "    syn_divergences = []\n",
    "    nonsyn_divergences_window_average = []\n",
    "    syn_divergences_window_average = []\n",
    "\n",
    "    \n",
    "    for years, subset_viruses in virus_time_subset.items():\n",
    "    #don't use windows with fewer than min_seqs \n",
    "        if len(subset_viruses) >= min_seqs:\n",
    "            year_windows.append(years)\n",
    "            seqs_in_window.append(len(subset_viruses))\n",
    "            \n",
    "\n",
    "            syn_div_allviruses_in_window = []\n",
    "            nonsyn_div_allviruses_in_window  = []\n",
    "            with open(input_file_alignment, \"r\") as aligned_handle:\n",
    "                for virus in SeqIO.parse(aligned_handle, \"fasta\"):         \n",
    "                    #Only viruses in time window\n",
    "                    if virus.id in subset_viruses:    \n",
    "                        #check\n",
    "                        if len(virus.seq) != len(root_seq):\n",
    "                            print(virus)\n",
    "                        elif len(virus.seq) == len(root_seq):\n",
    "                            count_total_unambiguous = 0\n",
    "                            count_subs = 0\n",
    "                            count_syn_subs = 0\n",
    "                            count_nonsyn_subs = 0\n",
    "                            for pos in range(len(root_seq)):\n",
    "                                root_nt = str(root_seq[pos])\n",
    "                                virus_nt = str(virus.seq[pos])\n",
    "                                #skip ambiguous sites\n",
    "                                if virus_nt != 'N':\n",
    "                                    if root_nt != 'N':\n",
    "                                        count_total_unambiguous+=1\n",
    "                                        if virus_nt != root_nt:\n",
    "                                            count_subs+=1\n",
    "                                            #determine syn or nonsyn\n",
    "                                            codon = math.floor(pos/3)\n",
    "                                            codon_pos = pos-(codon*3)\n",
    "                                            if codon_pos == 0:\n",
    "                                                codon_nt = virus.seq[pos:(pos+3)]\n",
    "                                            elif codon_pos == 1:\n",
    "                                                codon_nt = virus.seq[(pos-1):(pos+2)]\n",
    "                                            elif codon_pos == 2:\n",
    "                                                codon_nt = virus.seq[(pos-2):(pos+1)]\n",
    "                                            codon_aa = codon_nt.translate()\n",
    "                                            root_aa = root_aa_seq[codon]\n",
    "                                            #skip ambiguous\n",
    "                                            if root_aa != 'X':\n",
    "                                                if codon_aa != root_aa:\n",
    "                                                    count_nonsyn_subs+=1\n",
    "                                                elif codon_aa == root_aa:\n",
    "                                                    count_syn_subs+=1\n",
    "\n",
    "\n",
    "                            #Multiply div by fraction of sites that were unambiguously sequenced\n",
    "                            unambiguous_ratio = count_total_unambiguous/len(root_seq)\n",
    "                            syn_div_for_virus = (count_syn_subs*unambiguous_ratio)/(syn_denominator*unambiguous_ratio)\n",
    "#                             syn_div_for_virus = (count_syn_subs)/(syn_denominator)\n",
    "                            syn_div_allviruses_in_window.append(syn_div_for_virus)\n",
    "                            nonsyn_div_for_virus = (count_nonsyn_subs*unambiguous_ratio)/(nonsyn_denominator*unambiguous_ratio)\n",
    "#                             nonsyn_div_for_virus = (count_nonsyn_subs)/(nonsyn_denominator)\n",
    "                            nonsyn_div_allviruses_in_window.append(nonsyn_div_for_virus)\n",
    "\n",
    "\n",
    "\n",
    "                mean_syn_div_in_window = sum(syn_div_allviruses_in_window)/len(syn_div_allviruses_in_window)\n",
    "                mean_nonsyn_div_in_window = sum(nonsyn_div_allviruses_in_window)/len(nonsyn_div_allviruses_in_window)\n",
    "\n",
    "                syn_divergences.append(syn_div_allviruses_in_window)\n",
    "                nonsyn_divergences.append(nonsyn_div_allviruses_in_window)\n",
    "\n",
    "                syn_divergences_window_average.append(mean_syn_div_in_window)\n",
    "                nonsyn_divergences_window_average.append(mean_nonsyn_div_in_window)\n",
    "\n",
    "            \n",
    "    return year_windows, seqs_in_window, syn_divergences, nonsyn_divergences, syn_divergences_window_average, nonsyn_divergences_window_average\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b28464f4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2b6635f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "341f9dae",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89a58cd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
