{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33194542",
   "metadata": {},
   "outputs": [],
   "source": [
    "from augur.utils import json_to_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.gridspec as gridspec\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import MutableSeq\n",
    "from collections import Counter\n",
    "import requests\n",
    "import random\n",
    "import math\n",
    "import json\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71e410d0",
   "metadata": {},
   "source": [
    "Make an expectation for wait times by randomly putting mutations on tree and then calculating wait times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 255,
   "id": "8f4a46bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_url = 'https://nextstrain-blab.s3.amazonaws.com/ncov_adaptive-evolution_2021-05-15.json'\n",
    "\n",
    "tree_json = requests.get(tree_url).json()\n",
    "\n",
    "#Put tree in Bio.Phylo format\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30469c3d",
   "metadata": {},
   "source": [
    "Find the actual number of S1 and RdRp mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "id": "c874c9be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to find path from root to clade\n",
    "def get_parent(tree, child_clade):\n",
    "    node_path = tree.get_path(child_clade)\n",
    "    return node_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "id": "13453d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def consolidate_deletions(mutation_list):\n",
    "    \"\"\"\n",
    "    For deletion mutations, consider adjacent sites as part of the same deletion\n",
    "    \"\"\"\n",
    "    \n",
    "    without_deletions = [x for x in mutation_list if x[-1]!='-' and x[0]!='-']\n",
    "    #consolidate deletions and reversions\n",
    "    deletions_only = [x for x in mutation_list if x[-1]=='-' or x[0]=='-']\n",
    "    deletions_only.sort(key=lambda x:x[1:-1])\n",
    "    \n",
    "    \n",
    "    #keep track of start of separate deletions\n",
    "    separate_deletions = []\n",
    "\n",
    "    # if there are deletions, count a run of consecutive sites as a single deletion/mutation\n",
    "    if len(deletions_only) != 0:\n",
    "        separate_deletions.append(deletions_only[0])\n",
    "\n",
    "        deletion_tracker = int(deletions_only[0][1:-1])\n",
    "        \n",
    "        for deletion in deletions_only[1:]:\n",
    "\n",
    "            deleted_pos = int(deletion[1:-1])\n",
    "            if deleted_pos == deletion_tracker+1:\n",
    "                pass\n",
    "            else:\n",
    "                separate_deletions.append(deletion)\n",
    "            deletion_tracker = deleted_pos\n",
    "    \n",
    "    consolidated_mutation_list = separate_deletions + without_deletions\n",
    "            \n",
    "    return consolidated_mutation_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "id": "ab10b892",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "# make dictionary with gene name as key and reference sequence of that gene as value\n",
    "reference_sequence_aa = {}\n",
    "reference_sequence_nt = {}\n",
    "\n",
    "\n",
    "# make dictionary giving gene by genomic location \n",
    "reference_gene_locations = {}\n",
    "\n",
    "# make dictionary saying what codon within the gene a certain genomic location falls within\n",
    "# and whether the mutation is at pos 0, 1 or 2 within codon\n",
    "reference_gene_codon = {}\n",
    "\n",
    "for record in SeqIO.parse(open(\"reference_seq_edited.gb\",\"r\"), \"genbank\"):\n",
    "    genome_seq = record.seq\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            # allow RdRp to overwrite Orf1a and Orf1b, \n",
    "            # to take care of changed reading frame due to  ribosome slippage\n",
    "            # S1 and S2 will also overwrite spike\n",
    "            for pos in range(int(feature.location.start), int(feature.location.end)):\n",
    "                reference_gene_locations[pos] = feature.qualifiers['gene'][0]\n",
    "                codon_num = math.floor((pos-feature.location.start)/3)\n",
    "                pos_in_codon = ((pos-feature.location.start)-codon_num*3)\n",
    "                reference_gene_codon[pos] = (codon_num, pos_in_codon)\n",
    "                \n",
    "            gene_seq = feature.location.extract(record.seq)\n",
    "            reference_sequence_nt[feature.qualifiers['gene'][0]] = gene_seq\n",
    "            gene_seq_aa = gene_seq.translate()\n",
    "            reference_sequence_aa[feature.qualifiers['gene'][0]] = gene_seq_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "id": "c6fd7806",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dictionary with length of each gene\n",
    "gene_lengths_aa = {k:len(v) for k,v in reference_sequence_aa.items()}\n",
    "gene_lengths_aa['Nsp6'] = 290\n",
    "gene_lengths_aa['Nsp4'] = 485"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "id": "421c3209",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary of list of colors for emerging lineages, depending on total number of lineages\n",
    "color_palettes = {8: ['#4068CF','#5098B9','#6CB28C','#94BD62','#BFBB47','#DFA53B','#E67131','#DB2823'], \n",
    "                  9: ['#3E5DD0','#4A8CC2','#60AA9E','#80B974','#A6BE55','#CBB742','#E29D39','#E56A2F','#DB2823'], \n",
    "                  10: ['#3F52CD','#4681CA','#57A1AD','#70B487','#90BC65','#B4BD4C','#D3B240','#E59638','#E4642E','#DB2823'], \n",
    "                  11: ['#3F47C9','#4274CE','#4F97BB','#64AC99','#7EB976','#9EBE5A','#BEBB48','#D9AE3E','#E69036','#E35F2D','#DB2823'], \n",
    "                  12: ['#403CC5','#4067CF','#4A8BC3','#5AA4A9','#6FB488','#8BBB69','#A9BD53','#C7B944','#DDA93C','#E68A35','#E35C2C','#DB2823'], \n",
    "                  13: ['#4433BE','#3E5ACF','#457FCB','#529AB6','#64AD98','#7BB77A','#96BD60','#B3BD4D','#CDB642','#DFA43B','#E68434','#E2582C','#DB2823'], \n",
    "                  14: ['#492AB5','#3F4CCB','#4271CE,''#4C8FC0','#5AA5A8','#6DB38A','#85BA6F','#A0BE59','#BBBC49','#D2B340','#E19F3A','#E68033','#E2562B','#DB2823'], \n",
    "                  15: ['#4D21AD','#403FC6','#3F63CF','#4783C8','#539BB5','#63AC9B','#77B67F','#8EBC66','#A8BD53','#C1BA47','#D6B03F','#E39C39','#E67C33','#E1532B','#DB2823'], \n",
    "                  16: ['#571EA2','#4334BF','#3F55CE','#4376CD','#4C91C0','#59A4A9','#6AB18F','#7FB975','#97BD5F','#AFBD4F','#C7B944','#D9AD3D','#E49838','#E67932','#E1512A','#DB2823'], \n",
    "                  17: ['#5E1D9D','#462EB9','#3F4CCB','#416CCE','#4887C6','#539CB3','#62AB9C','#74B582','#89BB6B','#A0BE59','#B7BD4B','#CCB742','#DDAA3C','#E69537','#E67631','#E14F2A','#DB2823']}\n",
    "                      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "id": "92b12cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nuc_changes_from_reference(muts_on_path):\n",
    "    \"\"\"\n",
    "    From all the of the nucleotide changes that have occurred on the path from root to branch, \n",
    "    find the most recent nuc mutation at each site (giving the genotype at the branch)\n",
    "    \"\"\"\n",
    "    \n",
    "    final_muts_from_ref = {}\n",
    "\n",
    "    # overwrites genotypes at pos in historical order\n",
    "    for x in muts_on_path:\n",
    "        x_pos = int(x[1:-1])\n",
    "        final_muts_from_ref[x_pos] = x[-1]\n",
    "        \n",
    "\n",
    "    return final_muts_from_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "id": "fbd74bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_synonymous(nuc_muts_on_branch, parent_diffs_from_ref):\n",
    "    \n",
    "    parent_diffs_pos = [int(k) for k,v in parent_diffs_from_ref.items()]\n",
    "\n",
    "    \n",
    "    # make dictionary of synonymous (and noncoding) mutations to add to tree\n",
    "    syn_muts = {}\n",
    "    \n",
    "    # don't care about deletions because they are obviously not synonymous\n",
    "    for mut in nuc_muts_on_branch:\n",
    "        if mut[-1]!= '-' and mut[0]!='-':\n",
    "            mut_pos = int(mut[1:-1])\n",
    "            # find what gene this mut happens in\n",
    "            if (mut_pos-1) in reference_gene_locations.keys():\n",
    "                mut_gene = reference_gene_locations[mut_pos-1]\n",
    "                mut_codon_num = reference_gene_codon[mut_pos-1][0]\n",
    "                mut_codon_pos = reference_gene_codon[mut_pos-1][1]\n",
    "                \n",
    "                # find the reference sequence of the codon this mutation occurs in\n",
    "                codon_ref_aa = reference_sequence_aa[mut_gene][mut_codon_num]\n",
    "                \n",
    "                codon_ref_nt = reference_sequence_nt[mut_gene][(mut_codon_num*3):(mut_codon_num*3+3)]\n",
    "                \n",
    "                # check if a mutation occurred within the same codon in a parent\n",
    "                # and if so, change the reference codon sequence accordingly, \n",
    "                # to tell whether the mutation at this branch is synonymous or not\n",
    "                codon_genome_pos = list(range((mut_pos-1-mut_codon_pos),(mut_pos-1-mut_codon_pos+3)))\n",
    "                \n",
    "                parent_codon = codon_ref_nt\n",
    "                for parent_diff in parent_diffs_pos:\n",
    "                    parent_diff_zero_based = parent_diff-1\n",
    "                    if parent_diff_zero_based in codon_genome_pos:\n",
    "                        parent_diff_pos = codon_genome_pos.index(parent_diff_zero_based)\n",
    "                        parent_codon = MutableSeq(str(codon_ref_nt))\n",
    "                        parent_codon[parent_diff_pos] = parent_diffs_from_ref[parent_diff]\n",
    "                        parent_codon = parent_codon.toseq()\n",
    "                \n",
    "                \n",
    "                codon_mutated = MutableSeq(str(parent_codon))\n",
    "                #if deletion (or seq error) has happened at neighboring nucleotide\n",
    "                if '-' in codon_mutated:\n",
    "                    pass\n",
    "                else:\n",
    "                    codon_mutated[mut_codon_pos] = mut[-1]\n",
    "                    codon_mutated = codon_mutated.toseq()\n",
    "                    codon_mutated_translation = codon_mutated.translate()\n",
    "\n",
    "                    if str(codon_ref_aa) == str(codon_mutated_translation):\n",
    "                        if mut_gene in syn_muts.keys():\n",
    "                            syn_muts[mut_gene] += [mut]\n",
    "                        else:\n",
    "                            syn_muts[mut_gene] = [mut]\n",
    "                        \n",
    "                \n",
    "\n",
    "            else:\n",
    "                if 'noncoding' in syn_muts.keys():\n",
    "                    syn_muts['noncoding'] += [mut]\n",
    "                else:\n",
    "                    syn_muts['noncoding'] = [mut]\n",
    "                    \n",
    "    return syn_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "id": "2c44c1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_syn_mut_attribute(tree):\n",
    "    \n",
    "    for node in tree.find_clades():\n",
    "\n",
    "        node.node_attrs['syn_muts'] = {}\n",
    "\n",
    "        # only care if this branch has some nucleotide mutations\n",
    "        if hasattr(node, 'branch_attrs'):\n",
    "            if 'nuc' in node.branch_attrs['mutations']:\n",
    "\n",
    "                nuc_muts_on_branch = node.branch_attrs['mutations']['nuc']\n",
    "\n",
    "                node_path = get_parent(tree, node)\n",
    "\n",
    "                nucleotide_mut_path = []\n",
    "\n",
    "                # find all nucleotide mutations that happened in parents, \n",
    "                # in case they affect codons mutated on this branch\n",
    "                for parent in node_path[-1]:\n",
    "                    if hasattr(parent, 'branch_attrs'):\n",
    "                        if 'nuc' in parent.branch_attrs['mutations']:\n",
    "                            nucleotide_mut_path+=parent.branch_attrs['mutations']['nuc']\n",
    "\n",
    "                parent_diffs_from_ref = nuc_changes_from_reference(nucleotide_mut_path)\n",
    "\n",
    "                syn_muts_dict = determine_synonymous(nuc_muts_on_branch, parent_diffs_from_ref)\n",
    "\n",
    "                node.node_attrs['syn_muts'] = syn_muts_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 256,
   "id": "a26817cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_syn_mut_attribute(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "id": "f202af18",
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in tree.find_clades(terminal=False):\n",
    "\n",
    "    node.s1_nonsyn_at_node = 0\n",
    "    node.s1_syn_at_node = 0\n",
    "    node.rdrp_nonsyn_at_node = 0\n",
    "    node.s2_syn_at_node = 0\n",
    "    node.nsp6_nonsyn_at_node = 0\n",
    "    node.nsp4_nonsyn_at_node = 0\n",
    "    node.n_nonsyn_at_node = 0\n",
    "    node.e_nonsyn_at_node = 0\n",
    "    node.m_nonsyn_at_node = 0\n",
    "\n",
    "    if hasattr(node, \"node_attrs\") and 'S1' in node.node_attrs['syn_muts']:\n",
    "        node.s1_syn_at_node = len(node.node_attrs['syn_muts']['S1'])\n",
    "\n",
    "\n",
    "    if hasattr(node, 'branch_attrs'):\n",
    "\n",
    "        s1_nonsyn_at_this_node = []\n",
    "        s2_nonsyn_at_this_node = []\n",
    "        if \"S\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"S\"]:\n",
    "                if int(mut[1:-1]) in range(14,686):\n",
    "                    s1_nonsyn_at_this_node.append(mut)\n",
    "                elif int(mut[1:-1]) in range(687,1274):\n",
    "                    s2_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "        s1_consolidated = consolidate_deletions(s1_nonsyn_at_this_node)\n",
    "        node.s1_nonsyn_at_node = len(s1_consolidated)\n",
    "        s2_consolidated = consolidate_deletions(s2_nonsyn_at_this_node)\n",
    "        node.s2_nonsyn_at_node = len(s2_consolidated)\n",
    "\n",
    "\n",
    "\n",
    "        rdrp_nonsyn_at_this_node = []\n",
    "        nsp6_nonsyn_at_this_node = []\n",
    "        nsp4_nonsyn_at_this_node = []\n",
    "        if \"ORF1a\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"ORF1a\"]:\n",
    "                if int(mut[1:-1]) in range(4492,4401):\n",
    "                    rdrp_nonsyn_at_this_node.append(mut)\n",
    "                elif int(mut[1:-1]) in range(3570,3859):\n",
    "                    # exclude this ancestral mut\n",
    "                    if mut!= 'K3833N':\n",
    "                        nsp6_nonsyn_at_this_node.append(mut)\n",
    "                elif int(mut[1:-1]) in range(2777,3261):\n",
    "                    nsp4_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "\n",
    "        if \"ORF1b\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"ORF1b\"]:\n",
    "                if int(mut[1:-1]) in range(1,923):\n",
    "                    rdrp_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "        rdrp_consolidated = consolidate_deletions(rdrp_nonsyn_at_this_node)\n",
    "        node.rdrp_nonsyn_at_node = len(rdrp_consolidated)\n",
    "        nsp6_consolidated = consolidate_deletions(nsp6_nonsyn_at_this_node)\n",
    "        node.nsp6_nonsyn_at_node = len(nsp6_consolidated)\n",
    "        nsp4_consolidated = consolidate_deletions(nsp4_nonsyn_at_this_node)\n",
    "        node.nsp4_nonsyn_at_node = len(nsp4_consolidated)\n",
    "        \n",
    "        n_nonsyn_at_this_node = []\n",
    "        if \"N\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"N\"]:\n",
    "                n_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "        n_consolidated = consolidate_deletions(n_nonsyn_at_this_node)\n",
    "        node.n_nonsyn_at_node = len(n_consolidated)\n",
    "\n",
    "        e_nonsyn_at_this_node = []\n",
    "        if \"E\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"E\"]:\n",
    "                e_nonsyn_at_this_node.append(mut)\n",
    "            \n",
    "        e_consolidated = consolidate_deletions(e_nonsyn_at_this_node)\n",
    "        node.e_nonsyn_at_node = len(e_consolidated)\n",
    "        \n",
    "        m_nonsyn_at_this_node = []\n",
    "        if \"M\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"M\"]:\n",
    "                m_nonsyn_at_this_node.append(mut)\n",
    "            \n",
    "        m_consolidated = consolidate_deletions(m_nonsyn_at_this_node)\n",
    "        node.m_nonsyn_at_node = len(m_consolidated)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "id": "f9ada0b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes with S1 nonsyn mutations: 277\n",
      "Total number of S1 nonsyn mutations: 353\n",
      "Nodes with S1 syn mutations: 100\n",
      "Total number of S1 syn mutations: 104\n",
      "Nodes with RdRp mutations: 98\n",
      "Total number of RdRp mutations: 101\n",
      "Time window: 2019.982929101653-2021.3404645014461\n"
     ]
    }
   ],
   "source": [
    "# find number of nodes that have S1 mutations\n",
    "# consider only internal branches\n",
    "nodes_with_s1_nonsyn = 0\n",
    "total_s1_nonsyn = 0\n",
    "nodes_with_s1_syn = 0\n",
    "total_s1_syn = 0\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    # look only at nodes with at least 5 descendents, to exclude weird stuff at tips and sequencing errors\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        if node.s1_nonsyn_at_node!= 0:\n",
    "            nodes_with_s1_nonsyn+=1\n",
    "            total_s1_nonsyn += node.s1_nonsyn_at_node\n",
    "        if node.s1_syn_at_node!=0:\n",
    "            nodes_with_s1_syn+=1\n",
    "            total_s1_syn += node.s1_syn_at_node\n",
    "        \n",
    "        \n",
    "print(f'Nodes with S1 nonsyn mutations: {nodes_with_s1_nonsyn}')\n",
    "print(f'Total number of S1 nonsyn mutations: {total_s1_nonsyn}')\n",
    "print(f'Nodes with S1 syn mutations: {nodes_with_s1_syn}')\n",
    "print(f'Total number of S1 syn mutations: {total_s1_syn}')\n",
    "\n",
    "\n",
    "# find number of nodes that have RdRp mutations\n",
    "nodes_with_rdrp_muts = 0\n",
    "total_rdrp_muts = 0\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    # look only at nodes with at least 3 descendents, to exclude sequencing errors\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        if node.rdrp_nonsyn_at_node!= 0:\n",
    "            nodes_with_rdrp_muts+=1\n",
    "            total_rdrp_muts += node.rdrp_nonsyn_at_node\n",
    "        \n",
    "print(f'Nodes with RdRp mutations: {nodes_with_rdrp_muts}')\n",
    "print(f'Total number of RdRp mutations: {total_rdrp_muts}')\n",
    "\n",
    "first_timepoint = tree.root.node_attrs['num_date']['value']\n",
    "last_timepoint = 0\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    if node.node_attrs['num_date']['value'] > last_timepoint:\n",
    "        last_timepoint = node.node_attrs['num_date']['value']\n",
    "        \n",
    "print(f'Time window: {first_timepoint}-{last_timepoint}')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "id": "42866e2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_s2_nonsyn = 0\n",
    "total_nsp6_nonsyn = 0\n",
    "total_nsp4_nonsyn = 0\n",
    "total_e_nonsyn = 0\n",
    "total_n_nonsyn = 0\n",
    "total_m_nonsyn = 0\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    # look only at nodes with at least 3 descendents, to exclude sequencing errors\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        if node.s2_nonsyn_at_node!= 0:\n",
    "            total_s2_nonsyn += node.s2_nonsyn_at_node\n",
    "        if node.nsp6_nonsyn_at_node!= 0:\n",
    "            total_nsp6_nonsyn += node.nsp6_nonsyn_at_node\n",
    "        if node.nsp4_nonsyn_at_node!= 0:\n",
    "            total_nsp4_nonsyn += node.nsp4_nonsyn_at_node\n",
    "        if node.e_nonsyn_at_node!= 0:\n",
    "            total_e_nonsyn += node.e_nonsyn_at_node        \n",
    "        if node.m_nonsyn_at_node!= 0:\n",
    "            total_m_nonsyn += node.m_nonsyn_at_node\n",
    "        if node.n_nonsyn_at_node!= 0:\n",
    "            total_n_nonsyn += node.n_nonsyn_at_node"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "id": "aee702e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "134 119 100 12 32 280\n"
     ]
    }
   ],
   "source": [
    "print(total_s2_nonsyn, total_nsp6_nonsyn, \n",
    "      total_nsp4_nonsyn, total_e_nonsyn, \n",
    "      total_m_nonsyn, total_n_nonsyn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "id": "ed5c9f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_branchs = 0\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        total_branchs+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "id": "a4da7f8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find all branch lengths\n",
    "root_date = tree.root.node_attrs['num_date']['value']\n",
    "\n",
    "# keep track of parts of paths that have already been considered\n",
    "paths_segments_already_used_all = []\n",
    "\n",
    "# store all wait times between branchs\n",
    "branch_lengths_all = []\n",
    "branch_names_all = []\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    \n",
    "    if len(node.get_terminals()) >=3:\n",
    "\n",
    "\n",
    "        node_path = get_parent(tree, node)\n",
    "\n",
    "        # find the branch length (in decimal years) between node and previous node\n",
    "        path_dates = [n.node_attrs['num_date']['value'] for n in node_path]\n",
    "\n",
    "        if len(path_dates) > 1:\n",
    "            path_dates_with_root = [root_date] + path_dates\n",
    "            branch_lengths = [j-i for i, j in zip(path_dates_with_root[:-1], path_dates_with_root[1:])]\n",
    "\n",
    "    #         branch_lengths = [j-i for i, j in zip(path_dates[:-1], path_dates[1:])]\n",
    "\n",
    "\n",
    "            for n_index in range(len(node_path)):\n",
    "\n",
    "                # name of the node at the end of the branch\n",
    "                # branch length will be assigned to this node because \n",
    "                # it is the length of the branch leading to this node\n",
    "                branch_name = node_path[n_index].name\n",
    "                if n_index==0:\n",
    "                    preceding_node_name = tree.root.name\n",
    "                else:\n",
    "                    preceding_node_name = node_path[n_index-1].name\n",
    "\n",
    "\n",
    "                if [preceding_node_name, branch_name] not in paths_segments_already_used_all:\n",
    "\n",
    "                    branch_length = branch_lengths[n_index]\n",
    "\n",
    "                    if branch_length<0:\n",
    "                        branch_length = 0.0\n",
    "\n",
    "                    branch_lengths_all.append(branch_length)\n",
    "                    branch_names_all.append(branch_name)\n",
    "    #                 branch_names_all.append([root_name, branch_name])\n",
    "                    paths_segments_already_used_all.append([preceding_node_name, branch_name])\n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfea755f",
   "metadata": {},
   "source": [
    "Throw each of the S1 mutations onto a random branch of the tree by picking a random branch X times (where X= total number of S1 mutations). Can result in multiple hits on same branch (as with empirical data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0fbbc586",
   "metadata": {},
   "outputs": [],
   "source": [
    "def randomize_mutations_on_tree_multinomial(tree, num_mutations, branch_lengths, branch_names):\n",
    "    \n",
    "    # multinomial draw to randomize mutations on tree\n",
    "    branch_lengths_normalized = [x/sum(branch_lengths) for x in branch_lengths]\n",
    "\n",
    "    num_branches = len(branch_lengths_normalized)\n",
    "\n",
    "    hits = list(np.random.multinomial(num_mutations, branch_lengths_normalized, size=1))[0]\n",
    "    \n",
    "    muts_on_branch = dict(zip(branch_names, hits))\n",
    "\n",
    "    for node in tree.find_clades(terminal=False):\n",
    "        if len(node.get_terminals()) >=3:\n",
    "        # no branch length at root\n",
    "            if node.name == 'NODE_0000000':\n",
    "                node.random_muts = 0\n",
    "\n",
    "            else:\n",
    "                node.random_muts = muts_on_branch[node.name] \n",
    "    return tree\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f1a67c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_linege_names(old_name):\n",
    "    \n",
    "#     old_to_new = {'A.23.1': 'A.23.1', 'B.1.1.7': 'B.1.1.7 (Alpha)', 'B.1.351': 'B.1.351 (Beta)', \n",
    "#                   'B.1.427+B.1.429': 'B.1.427/429 (Epsilon)', 'B.1.525': 'B.1.525 (Eta)', 'B.1.526': 'B.1.526 (Iota)', \n",
    "#                   'B.1.617': 'B.1.617.1/2 (Delta/Kappa)', \n",
    "#                   'C.37': 'C.37 (Lambda)', 'P.1': 'P.1 (Gamma)', 'P.3': 'P.3', \n",
    "#                   'unassigned': 'basal'}\n",
    "    \n",
    "    old_to_new = {'unassigned': 'basal'}\n",
    "    if old_name in old_to_new.keys():\n",
    "        new_name = old_to_new[old_name]\n",
    "    else:\n",
    "        new_name = old_name\n",
    "    \n",
    "    return new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0fc89b66",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_linear_reg_stats(mut_location, nonsyn_syn, x_axis):\n",
    "    \n",
    "    df_subset = muts_information_df.copy()\n",
    "    df_subset = df_subset[(df_subset['mut_location']==mut_location)&(df_subset['nonsyn_syn']==nonsyn_syn)]\n",
    "    \n",
    "    if x_axis == 'logistic_growth':\n",
    "        df_subset = df_subset.dropna()\n",
    "    \n",
    "    \n",
    "    slope, intercept, r_value, p_value, std_err = stats.linregress(df_subset[x_axis], df_subset['muts_per_codon'])\n",
    "    slope = round(slope, 4)\n",
    "    r_value = round(r_value, 2)\n",
    "    \n",
    "    return slope, r_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "id": "11053433",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_growth_randomizations(iterations, save_filename):\n",
    "\n",
    "#     genes = ['S1','RdRp']\n",
    "    genes = ['S2', 'Nsp6', 'Nsp4', 'E', 'M', 'N']\n",
    "\n",
    "    growth_stats_dict = {'growth_rate_stats':[]}\n",
    "\n",
    "    for gene in genes:\n",
    "        if gene == 'S1':\n",
    "            nonsyn_syn_options=['nonsyn', 'syn']\n",
    "        else:\n",
    "            nonsyn_syn_options=['nonsyn']\n",
    "\n",
    "        for nonsyn_syn in nonsyn_syn_options:\n",
    "\n",
    "            for x in range(iterations):\n",
    "                #Download tree json\n",
    "                tree_json = requests.get(tree_url).json()\n",
    "\n",
    "                #Put tree in Bio.Phylo format\n",
    "                tree = json_to_tree(tree_json)\n",
    "                \n",
    "                # randomize mutation positions on tree\n",
    "                if gene == 'RdRp':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_rdrp_muts, branch_lengths_all, branch_names_all)\n",
    "\n",
    "                elif gene == 'S1':\n",
    "                    if nonsyn_syn == 'nonsyn':\n",
    "                        tree = randomize_mutations_on_tree_multinomial(tree, total_s1_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                    elif nonsyn_syn == 'syn':\n",
    "                        tree = randomize_mutations_on_tree_multinomial(tree, total_s1_syn, branch_lengths_all, branch_names_all)\n",
    "                \n",
    "                elif gene == 'S2':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_s2_nonsyn, branch_lengths_all, branch_names_all) \n",
    "                elif gene == 'Nsp6':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_nsp6_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                elif gene == 'Nsp4':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_nsp4_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                elif gene == 'E':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_e_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                elif gene == 'M':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_m_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                elif gene == 'N':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_n_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                \n",
    "                # keep track of logistic growth rate for this iteration\n",
    "                muts_information = []\n",
    "                # count accumulation of mutations on path\n",
    "                for node in tree.find_clades(terminal=False):\n",
    "                    \n",
    "                    # only care if it has a logistic growth rate\n",
    "                    if \"logistic_growth\" in node.node_attrs:\n",
    "                        logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "\n",
    "                        #Find all parents of the node\n",
    "                        parents = get_parent(tree, node)\n",
    "\n",
    "                        #Find mutations that occur in the parents, and at node\n",
    "                        parents_random_muts = 0\n",
    "\n",
    "                        for parent in parents:\n",
    "                            if hasattr(parent, \"random_muts\"):\n",
    "                                parents_random_muts+=parent.random_muts\n",
    "                                \n",
    "                        muts_information.append({'muts_per_codon': parents_random_muts/gene_lengths_aa[gene],\n",
    "                                                 'logistic_growth': logistic_growth})\n",
    "                        \n",
    "    \n",
    "                \n",
    "\n",
    "                muts_information_df = pd.DataFrame(muts_information)\n",
    "                slope, intercept, r_value, p_value, std_err = stats.linregress(muts_information_df['logistic_growth'], \n",
    "                                                                               muts_information_df['muts_per_codon'])\n",
    "\n",
    "                growth_stats_dict['growth_rate_stats'].append({'data': 'randomized', 'iteration':x,\n",
    "                                                               'gene': gene, 'nonsyn_syn': nonsyn_syn, \n",
    "                                                               'r_value':r_value})\n",
    "\n",
    "    with open(save_filename, 'w') as outfile:\n",
    "        json.dump(growth_stats_dict, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 436,
   "id": "68e2ba43",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_growth_randomizations(200, 'growth_rate_stats/growth_stats-supp_200itsmore.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 434,
   "id": "a7112ec5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fig1c(data_files, filename=None):\n",
    "\n",
    "    # dataframe for all wait time information, \n",
    "    # each row is an iteration for expected s1 or rdrp, syn or nonsyn\n",
    "    to_plot = pd.DataFrame(columns = ['data', 'iteration', 'gene', 'nonsyn_syn', 'r_value'])\n",
    "    \n",
    "    for data_file in data_files:\n",
    "        with open(data_file) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            df = pd.DataFrame(data['growth_rate_stats'])\n",
    "            to_plot = pd.concat([to_plot, df])\n",
    "            \n",
    "    num_iterations = len(to_plot)/3\n",
    "    \n",
    "    with open('growth_rate_stats/growth_stats_observed.json') as observed_json:\n",
    "        data_observed = json.load(observed_json)\n",
    "        to_plot_observed = pd.DataFrame(data_observed['growth_rate_stats'])\n",
    "\n",
    "    \n",
    "    fig, (ax1, ax2, ax3) = plt.subplots(1,3, figsize=(14,1.25), sharey=True)\n",
    "#     sns.set_style(\"white\")\n",
    "\n",
    "    s1_expected = to_plot[(to_plot['gene']=='S1')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax1 = sns.histplot(data=s1_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax1)\n",
    "    s1_observed_r = to_plot_observed[(to_plot_observed['gene']=='S1'\n",
    "                                     )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "\n",
    "    ax1.axvline(x=s1_observed_r, color = \"#4770eb\", linewidth=4)\n",
    "    s1_pvalue= len(s1_expected[s1_expected['r_value']>s1_observed_r])/num_iterations\n",
    "    if s1_pvalue < 0.005:\n",
    "        s1_pvalue = '< 0.005'\n",
    "        ax1.annotate(f'p-value: {s1_pvalue}', xy=((s1_observed_r), 0.65), \n",
    "                     xycoords='axes fraction',ha='left', size=11, fontweight='bold')\n",
    "    else:\n",
    "        ax1.annotate(f'p-value: {round(s1_pvalue, 2)}', xy=((s1_observed_r), 0.65), \n",
    "                     xycoords='axes fraction',ha='left', size=11, fontweight='bold')\n",
    "    \n",
    "    \n",
    "    s1_syn_expected = to_plot[(to_plot['gene']=='S1')&(to_plot['nonsyn_syn']=='syn')]\n",
    "    ax2 = sns.histplot(data=s1_syn_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax2)\n",
    "    s1_syn_observed_r = to_plot_observed[(to_plot_observed['gene']=='S1'\n",
    "                                   )&(to_plot_observed['nonsyn_syn']=='syn')]['r_value'].values[0]\n",
    "    ax2.axvline(x=s1_syn_observed_r, color = \"#5094e3\", linewidth=4)\n",
    "    s1_syn_pvalue= len(s1_syn_expected[s1_syn_expected['r_value']>s1_syn_observed_r])/num_iterations\n",
    "    ax2.annotate(f'p-value: {round(s1_syn_pvalue, 2)}', xy=((s1_syn_observed_r+0.5), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    \n",
    "    rdrp_expected = to_plot[(to_plot['gene']=='RdRp')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax3 = sns.histplot(data=rdrp_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax3)\n",
    "    rdrp_observed_r = to_plot_observed[(to_plot_observed['gene']=='RdRp'\n",
    "                                   )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "    ax3.axvline(x=rdrp_observed_r, color = \"#ff8c3a\", linewidth=4)\n",
    "    \n",
    "    rdrp_pvalue= len(rdrp_expected[rdrp_expected['r_value']>rdrp_observed_r])/num_iterations\n",
    "    ax3.annotate(f'p-value: {round(rdrp_pvalue, 2)}', xy=((rdrp_observed_r+0.15), 0.65), \n",
    "                 xycoords='axes fraction',ha='left', size=11, fontweight='bold')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ax1.set_title(f'S1 nonsynonymous', fontsize = 13, fontweight='bold')\n",
    "    ax2.set_title(f'S1 synonymous', fontsize = 13, fontweight='bold')\n",
    "    ax3.set_title(f'RdRp nonsynonymous', fontsize = 13, fontweight='bold')\n",
    "    \n",
    "    ax1.set_xlabel('$R^2$', fontsize=12)\n",
    "    ax2.set_xlabel('$R^2$', fontsize=12)\n",
    "    ax3.set_xlabel('$R^2$', fontsize=12)\n",
    "    \n",
    "    for ax in [ax1, ax2, ax3]:\n",
    "        ax.set_ylabel('')\n",
    "        plt.setp(ax.get_yticklabels(), visible=False)\n",
    "        ax.tick_params(axis='x', which='major', labelsize=10)\n",
    "    \n",
    "    \n",
    "    sns.despine(left=False, bottom=False)\n",
    "    plt.subplots_adjust(wspace=0.1)\n",
    "    \n",
    "    plt.figtext(0.04, 0.94, 'C', fontsize=20, fontweight='bold')\n",
    "\n",
    "    if filename:\n",
    "        fig.savefig(filename, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 435,
   "id": "0a0f329a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA3EAAACLCAYAAADPlRCuAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAqbElEQVR4nO3deXRNV8PH8W+ma4gWMVSEUKWXGCKSFqlIUGIoEUVLRatK+2rKiypKTa0xJVXDU42p8mp5UiktT7WGaqJIUTw1VxWJijlIQiSS9w8rty4iCUlubvL7rGWt3Nx9zt3nxPndvc/ZZx+bjIyMDERERERERMQq2Fq6AiIiIiIiIpJz6sSJiIiIiIhYEXXiRERERERErIg6cSIiIiIiIlZEnTgRERERERErok6ciIiIiIiIFbHPTeFdu3bx1VdfsXv3bi5cuEDJkiWpU6cOnTp1omfPnhgMhvyqp4iIiIiIiAA2OXlOXHp6OlOmTCE8PDzLMm5ubixevJjy5cvnaQVFRERERETkHznqxM2aNYsFCxYAUK5cOfr06UOtWrU4ceIE4eHhXL58GYCWLVsSFhaWvzUWEREREREpxrLtxJ04cYIOHTqQnp5O+fLl+frrr6lWrZrp/djYWLp3705CQgIAkZGR1K9fP18rLSIiIiIiUlxle0/cypUrSU9PB2DAgAFmHTiA6tWrM3DgQI4dO0bTpk1xdnbOn5qKiIiIiIhI9p24nTt3mn728fG5b5n+/fvnXY1EREREREQkS9k+YiA2Ntb0c/Xq1fO1MlJ0HThwgP79+9OkSRM8PDwICAjgm2++uadcYmIiLVu25IUXXrBALUXEWiljRIomo9Fo9q9Bgwa0bt2a0NBQsroj6Pfff8doNDJq1CgA5syZY7aOunXr4unpSVBQEIcOHSrIzRHJM9leiUtMTDT9XKJEiXytjBRNiYmJ9OvXj9TUVLp27UqZMmVYu3Yto0aNwtHRkXbt2gFw6tQphg8fztmzZ3n88cctXGsRsRbKGJGirUKFCvTp0weAlJQU1qxZw2effUaNGjXo1q1bjtfTqVMnateuTUZGBn/++Sfr1q0jODiYjRs3YmNjk1/VF8kX2V6JK1mypOnn5OTkfK2MFE0nT57kypUrNGvWjPHjxzN8+HDmzZuHn58f58+fByAmJoaOHTty7dq1bNc3atQojEYj//nPf2jfvj2enp4MGzaMpKQkU5kvv/wSf39/3N3d6dq1Kz/88IPpvaCgIDw8PFi3bh2tWrWiadOmTJo0iVu3bgGwefNmOnfujLu7O82aNWP06NHcuHGD+Ph46tWrR69evUzrWrNmDUajkbCwMGJiYjAajcyaNYvg4GDc3d0JCAhg7969pvKHDx+mX79+eHh40KpVKyZPnmw6riIjIzEajSxevJigoCAaN25Mr169+Ouvv0hJScHT0xNfX1/Tmcddu3ZhNBqZMGGC6bOnT5/Oa6+9RsOGDQkKCuLgwYP07duXhg0b8vLLLxMfH5/jfWQ0Grl06RIAixYtwmg0EhkZCUBERAT+/v40bNiQFi1aMGPGDNO9syIFLa8z5rfffqNnz540btyYZ599lrfffptLly5lexzGxcVhNBoZO3YsY8aMwcPDA39/fzZv3mxa9+nTpwkODqZJkya0aNGCUaNGmY6z7DLE39+fRo0ambIuPj4eo9HIgAEDTJ89evRoBg8eTKNGjQgMDGT//v0EBwfTqFEjOnfuzJEjR0x1Wb9+PQEBAbi7u9OxY0e++uor03Zl5uzvv/9uKms0GpkzZw6QdU6K5AcnJycGDRrEoEGDGDp0KGPHjgXgv//9LwDXr19n3LhxNG3alNatWxMVFXXf9bRr145Bgwbx9ttvM2vWLOrVq0dcXByXLl0yO37fe+89GjduTGBgIPv27bvvutQWUVvE0rLtxFWtWtX0c1xc3H3LHD58mN27d2d5WVuKtzp16lC9enU2b95Mu3btmDRpEnFxccyePZtXXnkFADs7O4KDg+87/CkrU6ZMoUWLFjg5ObFu3TpWrVoFwIoVK5g4cSL29vb06dOH5ORkBg8ezKZNm0zL3rhxg1mzZtG2bVvs7e1Zvnw5W7Zs4caNG/zv//4vSUlJ9OvXj8aNGxMZGcnnn39OlSpV8PLyYu/evVy4cAGATZs2YWNjQ8eOHU3rXrx4MXZ2drRo0YLDhw8zceJEAM6fP8+rr77K7t27efHFF6lXrx7Lli3jvffeM9uu0NBQXF1dadiwIb/99hszZ86kRIkStGvXjvj4eFOjKnN77hwWtnTpUpydnalfvz6//vor3bt3p0KFCnh6erJnzx7TI0Byso+yEhsbywcffIDBYGDAgAHUrFmTRYsW5epvJ5KX8jpjhg8fzsmTJ+nbty9+fn5s3LiR6dOn5/g4XL16NWfOnMHf358TJ04wevRo0tPTuXHjBq+++iqbN2+mffv2eHt7s3r1agYMGGBquEHWGdK5c2dSUlJMDdSNGzcCt68uZPrmm2+4desWzZo14+DBg/Ts2ZPr16/j5+fH0aNHCQ0NBSA6OpohQ4Zw5coV+vTpQ4kSJZgwYcIDnweb6UE5KZIfUlNTOXnyJCdPnuTw4cOsWbMGuH3sw+3hkitXrqRGjRr4+/uzdOnS+67nwoULnDx5kuPHj7Nhwwb++usvKlasaPaM48jISC5evEi3bt34448/eOutt8xGpd1NbRG1RSwl2+GUzzzzDEePHgVg69at1K1b954yYWFhrF27lsqVKzN9+nS8vb3zvqZitQwGA1988QWzZ89m8+bNLF++nOXLl+Pk5MScOXPw8vIy/cuNESNGEBAQQNOmTQkODubUqVMALFmyBIPBQHh4OE5OTrzyyiu0adOGsLAw2rRpA9x+gP306dPx8vKiatWqTJ06ldjYWGxtbbGzs8PBwQFnZ2deeOEFRo8ebboftEuXLvz6669s2rSJwMBAoqOj8fDwwMXFxXSSw8vLi9mzZ5ORkcGzzz5rqldkZCQJCQmMHTuWoKAgAPr162f6IsnUpUsXJk+ezJUrV8yW79y5M5GRkWzcuJFGjRqxadMmqlSpgqenJ7/++isA3t7eTJ06lZ07d9KnTx8aNGhAaGgoR44coUuXLvz999853kdZsbOzw8bGBoPBgKurK506daJUqVJmJ3xEClJeZ4y9vT329vZUqFCBjh07MmjQINPMzA86Dk+fPg2Ai4sLCxcuxNbWloMHD3LkyBGuXbtGdHQ0sbGx9OvXz3SvjsFgICIigu3bt+Pg4ABknSGdO3dmzpw5bNy4kQ4dOrBp0yZKlCjB888/b3rMz5NPPsncuXM5c+YMrVq1olKlSixYsIAbN27www8/mDIgs5G7YMECjEYjiYmJtG7dms8//5y+ffs+cP9kl5Miee3EiROmYdEADg4OdOvWjR49egC3rxSXLl2aJUuW4OjoSK1atUxX6+704Ycfmr2uVq0akydPxtb2n2sazs7OhIWFmf6fL1u2jO3bt9O2bdv71k1tEbVFLCXbK3E9e/Y0/bx48WLOnDlj9v7evXtZv349AJcuXaJmzZp5W0OxehkZGZQuXZpJkyaxY8cOIiIieP3117l06RJTpkx56PVm/l977LHHAEyX0P/++2+qVKmCk5MTcPtqcvny5c0m6bnf8rdu3cJgMDB//nzKli3L+PHj6dSpEwMHDiQmJgaA9u3bYzAY2LBhA9u2bSM5OfmeCRJq1KgBgI2NDWXKlDHVK7OBV69ePVNZNzc3wHwCoczlM+/ZyVy+WbNmVK5cmQ0bNvDHH39w8uRJOnbsaDaOPzPgM4dBZ67L0dHxofZRpjuvElStWpWZM2dy8+ZNRo4cSceOHQkODtbN4WIxeZ0xn376KbVr12bGjBkEBAQQFBRk+p7L6XGY2SjMPI5v3bplarjkNAPuzpAaNWrg7u7Oli1buHjxIjt37sTPz48yZcqYlq1WrRo2NjamDHBxccHe3t5U5s48sre3x2g0AlCmTBlq1KjB+fPnSUlJuWef3DlEKbucFMlrzs7OzJo1y/R9+9xzzzFmzBgMBgNw+wpb5cqVTd91tWrVuu96+vXrx9ixY6lYsSKlSpXi/fffp1mzZmZlXF1dTcfvU089BcC5c+eyrJvaImqLWEq2nbi6deuaeuqZl5fnz5/PunXrmDVrFv369SMtLQ2A3r17F6sesOTMsmXLaNasGQsXLsTe3p5GjRoxbNgwSpcubTYuOrcyQ/bum5GrVq3K2bNnTWOo4+PjuXz5Mq6urtkuf/XqVVJSUhg1ahTbt29n3rx5nD17lnHjxgG3Q9bPz48dO3awevVq7OzsaN++vdl67ezsTD/fuW4XFxfg9vDjTJlhkxlwdy5/93bZ2trSqVMnjh8/bhq2dHdo33k28e663Cm7fZS53PXr1wFMQzYyfzYYDISEhBAdHc20adM4cOAA06dPv+9nieS3vMyYGzducPHiRQYMGEBMTIzpitro0aNJS0vL0XGYVQZkfj/mNAPuXh5unwVPTExk2rRppKamPnQGuLi4kJaWZhppk5SUxKlTp3jiiScoUaLEPRmQeW8hZJ+TInmtTJkydOrUiZkzZ9KpUye2bNnC0KFDTe9XqVKFs2fPmu5HO378+H3X07hxY4KCgli4cCHp6ekMHjyY3bt3m5U5duwYqampwO0rgJnrz4raImqLWEq2wynh9s2bycnJrFq1ikuXLjF79ux7yvj6+jJixIg8r6BYv/bt2zN//nz+9a9/cerUKZydnbM8c5QXXnnlFSZPnmy6n2XDhg2kp6fz5ptvZrtscnIyQ4YMwWAw0LNnT65du8b169fNHmLfpUsXfvzxR77//ntatGhBhQoVclSvLl26EBYWxscff8yJEyeIj4/nl19+wd/fnxo1atzzRZLVOpYsWcK3335LzZo1qV+/fo4++27Z7SNXV1e2b9/O5MmTcXNzIyIiwrTs33//zTvvvEOlSpUIDAw0nTG7cx+JFKS8zJiMjAxGjhzJlStX6NWrF7a2tly6dInKlStjb3/7K/Nhj8NWrVpRtWpVvvjiC65evcqtW7dYvXo1jRo1olmzZmbPZc1Kp06dmDZtGt9++y1lypTB19c3V9uXqXfv3mzdupU333yTjh07sn37dhISEkyNxMxG1OzZs/H19eWLL74wLZuTnBTJL5MmTWL//v1ERUWxYsUKXn75Zbp06cKcOXPo27cvzz77rGnii6zUq1ePUaNGMXHiREaOHMl3331neu/cuXO8/vrrGI1GVq5cSaVKlWjevHmu66m2SPbUFnk02V6Jg9v3B0yZMoXFixfTtm1bKlWqhL29PWXLlqV58+bMmDGDBQsWmC5ri9zpiSeeYOnSpfj4+LB161YWLVrExYsXef311+87Zv1R9e3bl3HjxpGWlkZ4eDilSpVi7ty5+Pn5ZbtslSpVCAsL46mnnuLLL7/khx9+oG3btmZndnx9fSlbtixgPqFAdpydnQkPD8fT05NVq1axf/9+XnvttVydNXJzc6N27drAvWe+ciO7ffQ///M/NG7cmOjoaGJiYszOeDZq1IhPPvmE8uXLs3TpUnbs2EFgYCCjR49+6PqIPIq8zJhSpUqxaNEinnnmGSIjI1m1ahVeXl7861//MpV52OPQ0dGRZcuW8fzzz7N+/Xqio6MJDAw03X+TE05OTqb7ztu2bfvQj/5p06YNn3zyCeXKlSM8PJwbN24wceJE00QwvXv3xsfHh//+9798//33Zidpc5KTIvmlTJkyzJw5EwcHB0JCQjh79iwDBw4kKCiIuLg4Nm7cmKPvo969e9OuXTtiY2PNLk40b94cFxcXVq9eTZ06dfjss88oXbp0ruuptkj21BZ5NDYZmlJSJFfS09Np27YtZ8+eZdu2bQX+vKk+ffqwc+dO1q1bZwpRESlYljwOR44cyerVq/n8888f+kqciJiLi4ujTZs2+Pn5sWDBAktXJ1tqi0iOhlOKyG1fffUVO3fuJC4ujo4dOxZoaK5du5Zdu3axc+dO3N3dFZoiFmDJ4/Dnn38mJiaGdevWUaVKFZ577rkC+2wRKTzUFhFQJ04kV6Kjo9m6dStNmzbl/fffL9DP3rt3L6tWraJBgwZMnTq1QD9bRG6z5HH4559/Eh4ejqurK5MmTTLdoycixYvaIgIaTikiIiIiImJVcnYntYiIiIiIiBQKGothAf3792fRokWWroaIVWo/9ILZ6/WhFS1Uk4KhvPhH4Lw4s9ffvF3NQjURKbyUGZJvZt51H+7wXyxTDwF0Jc4iLl++bOkqiIiVUF6ISG4oM0SKB3XiRERERERErIiGU4qIiFjYmjVrSE1NzZd1Ozg4EBAQkC/rFhERy1AnTiSf5FejTA0ykaInNTWVli1b5su6o6Ki8mW9ImI5amOIOnEi+SS/GmVqkImIiBRvamOI7okTERERERGxIroSJ8Vefg1JsLfX4SUiIiIieU+tTCn28vNeFBERERGRvKZOnIiISA7pyr2IiBQG+tYQERHJIV25FxGRwkATm4iIiIiIiFgRdeJERERERESsiDpxIiIiIiIiVkT3xImIiIiI5ANNhiT5Rf8DRERERETygSZDkvyi4ZQihcjixYvx8fEhOTm5QD4vIyPjntdTp07l2Wefxd/fn19++eW+yy1cuBBvb2/8/Pz49ttvTb9fs2YNvr6+eHt7s3DhQtPvQ0NDMRqNpn+TJk3Knw0SKWD5ecyeOXOGt99+m3bt2jFy5EiuXLlyT5mjR4/yxhtv4O/vz+jRo0lISDAtO2TIENq1a8e8efM4dOhQntdPRO41Z84cjEYjSUlJeb7unGTCkSNHGDhwIM8//zyvv/46Bw8eBG5/vy9dupSuXbsSEBDAjz/+mOf1k4KlTpxIMZORkcGWLVt49dVX2blzp9l733//PUuXLuWdd96hZs2aDB06lJSUFLMy+/btIyQkhJ49e9K8eXPef/994uPjiY+PZ8yYMbRs2ZLu3bsTEhLCvn37APj9999p1qwZixcvZsmSJfTp06fAtlfEWs2cOZOEhARGjBjB/v37zU6MZBo/fjy2traMGjWKw4cP8/HHHwMwa9Ys4uPjeffdd7l+/Tpjxowp6OqLSB7LSSZMmjSJ9PR0xo8fT3p6OhMnTgRg7dq1LF26lFdffZXGjRszffr0fOloSsFRJ04kl/bs2YOPjw8LFy6kc+fO9OjRg+3bt2dZLioqCoC9e/fi4+PDtm3bOHbsGP3796dNmzYEBASwatWqe5a/+wx/jx49GDt2LACnTp0iKCgIDw8Punfvzu+//w5ATEwMRqORn3766Z71paSk8O9//5uOHTvy5ptvcu3aNapUqWJWJjo6mipVqhAUFETfvn25cuWKqSN2Zxk7OzsGDRrEm2++SWpqKtu3b2f79u2kpqYyYMAA3n77bWxtbYmOjiYjI4P9+/dz4MABBg4cSFhYGOXKlcv9jhd5SJnHxezZs2nevDmtW7fm559/zrLcxo0bAdi5cydGo5EtW7Zw+PBhunXrxvjx4x/qmD19+jSDBw+mXbt2DBw4kMOHDwP/5MTdV73T0tLYvXs3rVq1om3btjRt2pSYmBizMgkJCcTFxdG6dWtatWrF888/z7Zt20hLS2P8+PHMmzcPLy8vSpYsiYODw6PvSJEiIi8zoWHDhrRo0YL/+7//u2f5OXPmMGbMmALLhLS0NF544QXeeecdfHx88PT05Ny5c6Snp/PTTz/h5uZGYGAg7777LkuWLKFkyZKPvjPFYtSJE3lIv/zyC++++y7lypXjww8/JDk5mZs3b5KcnExycjINGzbExcWFDRs2ALBhwwYqVKhA06ZN+f777zEYDEyYMIFatWrd92zag7z33numM2zOzs4MGTKE1NRU6taty5IlS3B3dzcrf/HiRfz8/AgJCaF58+asWbOGyMhIXF1dzcqdPXvW1MFycnIC4Ny5c/eUKV26NAaDwazM2bNnAShXrhwlSpTA0dGRc+fOcfnyZapVq0bv3r35+OOP2bt3L7NmzcrV9orkhc2bNzNx4kTKly/PiBEjSExM5ObNmyQlJZGUlESTJk1wdXU1DRH+7rvvqFSpEj4+PnzzzTcYDAZeeumlhzpmP/roI9LT03n33XepXLky48aNIy0tjdq1axMaGkr9+vXNyickJJCWlsbjjz8O3D6uLly4YFbmscceo1SpUuzbt4/Lly9z5MgRUlNTSUhIoEyZMlSsWJHevXtz/vx5Ro4c+Qh7TqRoyotMCA0NpU6dOsyePTtXn50fmWBvb0+vXr1wd3cnPj6eH3/8EQ8PD2xtbYmPjyc5OZlevXrRvXt3fv75Z+zs7B5h74mlaWITkYfUu3dvfH19sbGxYcyYMRw7dozvvvuO9evXAzB69Gg6dOjAsmXLSEhI4KeffqJz587Y2dnx1ltvsXXrVnbv3s3p06dJTEzM9vMy719LSUlhz549AOzatcv0fmxsLLVq1cLb2/ueZW1sbLC1teXmzZtcv349y5myMjIysLGxMfu8nJbJ/Pnu95ycnIiMjDSVW7duHb/++mu22yuS19544w3atWuHjY0NwcHBHDlyhIiICL755hsApk6dSmBgIJ999hmXLl1i/fr19OjRAzs7O4YPH86mTZtYuXJlro/Z5ORk9u/fD2B2Zfvvv//G1dUVLy+vLJd90PFoZ2fHkCFDCAkJoUuXLlSrVs1smfT0dD766CMWLFjAgAEDGDZsGCVKlMjx/sqOg4MDAQEBebY+kYKWF5mwY8cOTp06xdWrV7P9vPzOhEzx8fEMHjyYtLQ0Bg8ebCp/+vRpxo4dy2+//cbnn3+Op6cnbm5uOdlVUgipEyfykNLS0sxe29ra8tprr9G1a1cAqlatSmpqKosXL+bTTz/l2rVrdOjQAbh9H8vp06fp378/JUuWZPny5fesPzOob926BWC6Ny09PR2Al19+mcDAQK5fv46DgwOVK1fOsq5OTk789NNPrFmzhiVLlhAZGUmDBg345JNPqF69uqlcpUqV+OuvvwBMN0zfvd5KlSqRlJREamqqWZnMeiYkJFCiRAmuX79O5cqViY2NZe3atXTt2hVnZ2fS0tI0NbJYROYxm3ls2draMmjQIF5++WUAXF1duXnzJnPmzGHKlClcuXKFbt26ATB06FBOnTqFl5cXtWrVytUxm/k6ICCADh06cP36dQwGAxUrVsyyruXKlcPOzs50jF29evW+5du3b4+Hhwdw+56Xf//735QrV46YmBgee+wxvLy8uHr1KuPHj6d69erUrVs3l3sta5lDxUWsVV5kwuDBgylZsiRhYWH3rN8SmXDhwgUGDx5MUlISoaGh1KxZE7j9PV26dGl8fX2pWbMmkZGRnDhx4p5OnL29PV9//fV969D9rtdZlbsfnfTJe2pJidUobM9a+eqrryhdujQrVqzg8ccfp1atWpQuXRoXFxezcl5eXmzYsAE3NzdTmP72229UrVqVmzdvsm3bNuCfUM9UtmxZAH788UcyMjJMs86VKlUKNzc3tm7dStOmTfnyyy85ceIEmzdv5sqVKxw4cIC6deuahjpmMhgM9OjRg+7duxMVFcWiRYs4c+aMWSfO29ubNWvWEB4ezrZt2yhXrhzu7u6cO3eOY8eO0bhxY5577jnmz5/PvHnzuHDhAg4ODnh7e5OamoqdnR2ff/45Tk5O3Lp1i5YtW2Jvb8+8efM4cOAArVu3Ztu2bbzxxhsPtc9FHsWiRYtwdHRk8eLFlCtXjqeffhpHR8d7hhV7e3vz3Xff4e7uzlNPPQXAjh07qF69OmlpaaYJgXJ6zD722GPUqVOHX3/9lSZNmhAZGUlsbCwRERFcu3aNI0eOULt2bbN7RR0cHHB3d2fz5s3UrFmTmJgYWrVqBdyefe7WrVu4ubkxZMgQEhIS6NevHz/88AMtW7bEzs6O5cuX8/fff/POO+8QGRlJ2bJl79lOkeIuLzIhJSWFLVu2APdmQvny5YGCzYSPPvqIM2fOMGDAABITE9m1axceHh74+Pgwd+5c1qxZw/Hjx7GxseHpp5++Z5/cbzSPSXio2cvcPDpBJ33ynu6JE6uR+ayVvP73wMB6gLp16xIaGsrFixcZN24cpUuXvm+5F154AcB0FQ6gf//+nDlzhtmzZ5u+LE6dOmW2nL+/P56ennz22Wf88ccf1KtXz/Te7NmzcXFxYcyYMVy4cIEZM2ZgMBg4fPgw/fr1u2cykjvZ2Njg6+vLsmXLeOaZZ8ze69KlC6+//jpz587lzz//ZNasWZQoUYKtW7fSr18/4uLi8PLyYsSIEURERBAdHc3kyZN54oknqFatGlOnTmXr1q1EREQwYsQIGjVqhLOzMyEhIRw6dIiJEyfSoUMH3nrrrdztbJE80LBhQyZNmsS5c+cICQnB0dHxvuW6d799vjkwMND0u8GDBxMXF8e6dese6pj98MMPcXZ2Ztq0aVy+fJkPPvgAg8HAsWPHGDp0KAcOHLinHiNHjsTJyYkZM2bg5uZmOvkxd+5c032lwcHBAISEhNCgQQOGDh0KwKhRo6hatSofffQRycnJTJ06NcuMEimu8iITJk+ezJNPPgnA8ePHzZYLCAjgqaeeKrBMOHHiBLt37wYgLCyMoUOHmmaZDgwMpGfPnoSFhREdHc2wYcOoXbv2I+w9sTSbjAcNqpV80a1bN7N7hCRnvv7660LxwMw9e/YwePBgpk2bxnPPPVfgnx8VFWX6QimO2g81v5F7fWjWw0+KAuXFPwLnxZm9/ubtajlaLiYmhr59+/LZZ5+Zzlw/rMKSQ4VBcc+iwkqZkb28zITsFKXMqBweaPb6XNA3OV5WeZH3dCVORERERETEiuieOJFc8vDwIDo62tLVEJEcatq0KUeOHLF0NUSkkFAmSFGgTpyIlXnQzFGPQjNHiYiIiFgHdeJErMzDTsSSHc0cJSIiImIddE+ciIiIiIiIFdGVOBERKXIK23MlRURE8pK+jUREpMjJfK6kiIhIUaThlCIiIiIiIlZEV+JEREREpFjTEGyxNvqfJSIiIiLFmoZgi7XRcEoRERERERErok6ciIiIiIiIFVEnTkRERERExIqoEyciIiIiImJF1IkTERERERGxIurEiYiIiIiIWBE9YkDyVH49ZwX0rBUREREREVAnTvKYnrMiIiIiIneyt7fn66+/zpd1Ozg4EBAQkC/rLszUiRMRERERkXzj7e2db+uOiorKt3UXZronTkRERERExIroSpyIiIjkWn4NjyquQ6NERHJDnTgREbGY3E2G1Mzs1YM6EJoIKf/l1/Co4jo0SkQkN/QtJyIiFpObyZCWR9w0e61JlEREpLhSJ05ERERECj09xkjkH/ofKyIiIiKFnh5jJPIPzU4pIiIiIiJiRdSJExERERERsSLqxImIiIiIiFgRdeJERERERESsiDpxIiIiIiIiVkSzUxZT+TVNr6botV729vYPfHjyo3BwcCAgICBf1i0iIiJS3KjFXUxpml65m7e3d76tOyoqKt/WLSIiIlLcqBMnIiIiIiJWKb9GEhX2UUTqxImIiIiIiFXKr5FEhX0UkSY2ERERERERsSK6EiciItnSZEgiklPKC5H8p6NBRESypcmQpKAU1/tbihLlhUj+UydORERECo3ien+LiEhu6J44ERERERERK6IrcSKS7/J2eJRfHq1HRERE5P4K+9BudeIKOd0cLEVBXg6PWvhLnq1KRKRYUxtDJGuFfWi3jrJCTjcHi0huqFEmIjmlNoaI9dK3sohYtcI81MES1CgTub/CPjRKRCQ3bDIyMjIsXYnipmnTpri4uOSobEJCAvoTiRQsGxsbypUrl6Oy5cuXZ9GiRflWl9zkBSgzRApabvICCldmKC9ECl5etTHUiRMREREREbEiesSAiIiIiIiIFVEnTkRERERExIqoEyciIiIiImJF1IkTERERERGxIurEiYiIiIiIWBF14kRERERERKyIHvZdBN24cYMRI0Zw8eJFHB0dmT59Ok5OTmZlli5dyrp16wDw9fUlODjYElXNsfT0dCZMmMCRI0cwGAx89NFH1KhRw/T+5s2bmTdvHvb29rz44ov07NnTgrXNney2be3atXzxxRfY2dnx9NNPM2HCBGxtreP8S3bblumDDz6gbNmyvPvuuxaopXXKyXEOt/8GAwcOpE2bNvTq1csCNc2etWdWUcmnopRFxTV7ilIu3MnaM+J+ikpu3K0o5cjdCluuWMdek1z56quvePrpp/nyyy/p2rUr8+fPN3s/NjaWb7/9lhUrVrBy5Uq2bt3K4cOHLVTbnNm4cSM3b95k5cqVDB8+nGnTppneS01NZerUqSxevJjw8HBWrlzJ+fPnLVjb3HnQtt24cYNPPvmEZcuWsWLFChITE/npp58sWNvcedC2ZVqxYgVHjx61QO2sW3bHeaZPPvmEK1euFHDtcsfaM6uo5FNRyqLimj1FKRfuZO0ZcT9FJTfuVpRy5G6FLVfUiSuCdu/ejY+PDwAtW7Zk+/btZu9XqVKFhQsXYmdnh62tLWlpaZQoUcISVc2xO7epcePG7N+/3/Ten3/+iaurK2XLlsVgMODp6cmuXbssVdVce9C2GQwGVqxYQalSpQCs4m91pwdtG8CePXvYt28fL730kiWqZ9WyO84B1q9fj42NDS1btizo6uWKtWdWUcmnopRFxTV7ilIu3MnaM+J+ikpu3K0o5cjdCluuaDillYuIiOCLL74w+12FChV47LHHAHB0dOTatWtm7zs4OODk5ERGRgYzZszAzc2NJ598ssDq/DASExMpU6aM6bWdnR1paWnY29uTmJho2l64vc2JiYmWqOZDedC22draUrFiRQDCw8NJTk7mueees1RVc+1B23bu3Dnmzp3L3Llz+f777y1Yy8LvYY7zo0ePsnbtWj799FPmzZtXYHXNTlHMrKKST0Upi4pD9hSlXLhTUcyI+ykquXG3opQjdytsuaJOnJXr0aMHPXr0MPtdcHAwSUlJACQlJfH444/fs1xKSgrvv/8+jo6OjB8/vkDq+ijKlClj2ia4PS7Z3t7+vu8lJSWZhV9h96Bty3wdEhLCX3/9xZw5c7CxsbFENR/Kg7Zt/fr1XL58mYEDB3L+/Hlu3LhBrVq16Natm6WqW2g9zHG+evVqzp49y6uvvsrp06dxcHDAxcXF4mffi2JmFZV8KkpZVByypyjlwp2KYkbcT1HJjbsVpRy5W2HLFQ2nLIKaNGnCzz//DEBUVBSenp5m72dkZDBo0CCMRiOTJk3Czs7OEtXMlSZNmhAVFQXA3r17efrpp03vPfXUU5w8eZKEhARu3rzJrl278PDwsFRVc+1B2wYwbtw4UlJSmD9/vmkIgrV40Lb17duXyMhIwsPDGThwIC+88ILVNaIsKbvj/L333iMiIoLw8HACAwN57bXXClVD7U7WnllFJZ+KUhYV1+wpSrlwJ2vPiPspKrlxt6KUI3crbLlik5GRkZGvnyAF7vr164wcOZLz58/j4ODAzJkzqVSpEkuWLMHV1ZX09HSGDRtG48aNTcsMGzasUAdE5oxAR48eJSMjgylTpnDw4EGSk5N56aWXTLM4ZWRk8OKLL/LKK69Yuso59qBta9CgAS+++CJeXl6ms1V9+/albdu2Fq51zmT3d8sUGRnJ8ePHi8wMcQUhu+O8TZs2prJz5syhYsWKhXYWOmvPrKKST0Upi4pr9hSlXLiTtWfE/RSV3LhbUcqRuxW2XFEnTkRERERExIpoOKWIiIiIiIgVUSdORERERETEiqgTJyIiIiIiYkXUiRMREREREbEi6sSJiIiIiIhYET3sWySf7dmzh2nTpuHg4EDlypWZPn06Dg4Olq6WiBRSygwRySnlRfGlRwyI5LOzZ89StmxZSpYsSWhoKPXq1aN9+/aWrpaIFFLKDBHJKeVF8aUrcSL57IknnjD9bGdnh62tRjGLSNaUGSKSU8qL4kt/aZFHdOXKFYxGIx4eHri7u9OqVSsiIiLuKRcbG0tUVBR+fn4FX0kRKTSUGSKSU8oLyYquxIk8okOHDlG+fHl27NgBwHfffcfIkSNp06YNTk5OACQmJjJq1ChmzJiBwWCwZHVFxMKUGSKSU8oLyYquxIk8osOHD1O/fn3T62effZZbt25x9epVANLS0hg2bBjBwcHUqlXLUtUUkUJCmSEiOaW8kKyoEyfyiA4ePGgK2KtXrzJz5kzq169PjRo1AFi7di379u1j/vz5BAUF8Z///MeS1RURC1NmiEhOKS8kK5qdUuQRde7cmdjYWGxtbUlKSqJFixaEhISYhjmIiNxJmSEiOaW8kKzoSpzII7h58ybHjx/n22+/5bfffuPTTz9l3759ekaLiNyXMkNEckp5IQ+iTpzIIzh69CgGg4Hq1asD4O/vj7OzMz/88IOFayYihZEyQ0RySnkhD6JOnMgjOHToEHXq1MHGxsb0O19fXzZv3mzBWolIYaXMEJGcUl7Ig6gTJ/IIDh06hNFoNPudj48P27ZtIyUlxUK1EpHCSpkhIjmlvJAH0cQmIiIiIiIiVkRX4kRERERERKyIOnEiIiIiIiJWRJ04ERERERERK6JOnIiIiIiIiBVRJ05ERERERMSKqBMnIiIiIiJiRdSJExERERERsSLqxImIiIiIiFgRdeJERERERESsiDpxIiIiIiIiVuT/AeWYIFvqL2gVAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1008x90 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_fig1c(['growth_rate_stats/growth_stats_50it.json', \n",
    "            'growth_rate_stats/growth_stats_50itmore.json', \n",
    "            'growth_rate_stats/growth_stats_200its.json', \n",
    "            'growth_rate_stats/growth_stats_400its.json', \n",
    "            'growth_rate_stats/growth_stats_300its.json'], filename='plots/fig1c_r_stats.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 438,
   "id": "6311ad25",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fig1_supp_b(data_files_main, data_files_supp, filename=None):\n",
    "\n",
    "    # dataframe for all wait time information, \n",
    "    # each row is an iteration for expected 'S2', 'Nsp6', 'Nsp4', 'E', 'M', 'N'\n",
    "    to_plot = pd.DataFrame(columns = ['data', 'iteration', 'gene', 'nonsyn_syn', 'r_value'])\n",
    "    \n",
    "    for data_file in data_files_supp:\n",
    "        with open(data_file) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            df = pd.DataFrame(data['growth_rate_stats'])\n",
    "            to_plot = pd.concat([to_plot, df])\n",
    "\n",
    "    \n",
    "    for data_file in data_files_main:\n",
    "        with open(data_file) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            df = pd.DataFrame(data['growth_rate_stats'])\n",
    "            to_plot = pd.concat([to_plot, df])\n",
    "            \n",
    "    num_iterations = len(to_plot)/6\n",
    "    \n",
    "    with open('growth_rate_stats/growth_stats_observed.json') as observed_json:\n",
    "        data_observed = json.load(observed_json)\n",
    "        to_plot_observed = pd.DataFrame(data_observed['growth_rate_stats'])\n",
    "\n",
    "    \n",
    "    fig, ((ax1, ax2), (ax3, ax4), ( ax5, ax6), (ax7, ax8)) = plt.subplots(4,2, figsize=(8.5,3), sharey=True, sharex=True)\n",
    "    sns.set_style(\"white\")\n",
    "\n",
    "    \n",
    "    s1_expected = to_plot[(to_plot['gene']=='S1')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax1 = sns.histplot(data=s1_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax1)\n",
    "    s1_observed_r = to_plot_observed[(to_plot_observed['gene']=='S1'\n",
    "                                     )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "\n",
    "    ax1.axvline(x=s1_observed_r, color = \"#4770eb\", linewidth=4)\n",
    "    s1_pvalue= len(s1_expected[s1_expected['r_value']>s1_observed_r])/num_iterations\n",
    "    if s1_pvalue < 0.005:\n",
    "        s1_pvalue = '< 0.005'\n",
    "        ax1.annotate(f'p-value: {s1_pvalue}', xy=((s1_observed_r), 0.65), \n",
    "                     xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "    \n",
    "    \n",
    "    \n",
    "    s2_expected = to_plot[to_plot['gene']=='S2']\n",
    "    ax2 = sns.histplot(data=s2_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax2)\n",
    "    s2_observed_r = to_plot_observed[to_plot_observed['gene']=='S2']['r_value'].values[0]\n",
    "\n",
    "    ax2.axvline(x=s2_observed_r, color = '#1238aa', linewidth=4)\n",
    "    s2_pvalue= len(s2_expected[s2_expected['r_value']>s2_observed_r])/num_iterations\n",
    "    ax2.annotate(f'p-value: {round(s2_pvalue, 2)}', xy=((s2_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    \n",
    "    rdrp_expected = to_plot[(to_plot['gene']=='RdRp')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax3 = sns.histplot(data=rdrp_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax3)\n",
    "    rdrp_observed_r = to_plot_observed[(to_plot_observed['gene']=='RdRp'\n",
    "                                   )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "    ax3.axvline(x=rdrp_observed_r, color = \"#ff8c3a\", linewidth=4)\n",
    "    \n",
    "    rdrp_pvalue= len(rdrp_expected[rdrp_expected['r_value']>rdrp_observed_r])/num_iterations\n",
    "    ax3.annotate(f'p-value: {round(rdrp_pvalue, 2)}', xy=((rdrp_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "    \n",
    "    \n",
    "    \n",
    "    n_expected = to_plot[to_plot['gene']=='N']\n",
    "    ax4 = sns.histplot(data=n_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax4)\n",
    "    n_observed_r = to_plot_observed[to_plot_observed['gene']=='N']['r_value'].values[0]\n",
    "    ax4.axvline(x=n_observed_r, color = '#419D78', linewidth=4)\n",
    "    n_pvalue= len(n_expected[n_expected['r_value']>n_observed_r])/num_iterations\n",
    "    ax4.annotate(f'p-value: {round(n_pvalue, 2)}', xy=((n_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    \n",
    "    e_expected = to_plot[to_plot['gene']=='E']\n",
    "    ax5 = sns.histplot(data=e_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax5)\n",
    "    e_observed_r = to_plot_observed[to_plot_observed['gene']=='E']['r_value'].values[0]\n",
    "    ax5.axvline(x=e_observed_r, color = '#307358', linewidth=4)\n",
    "    \n",
    "    e_pvalue= len(e_expected[e_expected['r_value']>e_observed_r])/num_iterations\n",
    "    ax5.annotate(f'p-value: {round(e_pvalue, 2)}', xy=((e_observed_r+0.12) , 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    m_expected = to_plot[to_plot['gene']=='M']\n",
    "    ax6 = sns.histplot(data=m_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax6)\n",
    "    m_observed_r = to_plot_observed[to_plot_observed['gene']=='M']['r_value'].values[0]\n",
    "    ax6.axvline(x=m_observed_r, color = '#a2d9c3', linewidth=4)\n",
    "    \n",
    "    m_pvalue= len(m_expected[m_expected['r_value']>m_observed_r])/num_iterations\n",
    "    ax6.annotate(f'p-value: {round(m_pvalue, 2)}', xy=((m_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    nsp6_expected = to_plot[to_plot['gene']=='Nsp6']\n",
    "    ax7 = sns.histplot(data=nsp6_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax7)\n",
    "    nsp6_observed_r = to_plot_observed[to_plot_observed['gene']=='Nsp6']['r_value'].values[0]\n",
    "    ax7.axvline(x=nsp6_observed_r, color = \"#c35100\", linewidth=4)\n",
    "    \n",
    "    nsp6_pvalue= len(nsp6_expected[nsp6_expected['r_value']>nsp6_observed_r])/num_iterations\n",
    "    ax7.annotate(f'p-value: {round(nsp6_pvalue, 2)}', xy=((nsp6_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "    \n",
    "    nsp4_expected = to_plot[to_plot['gene']=='Nsp4']\n",
    "    ax8 = sns.histplot(data=nsp4_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax8)\n",
    "    nsp4_observed_r = to_plot_observed[to_plot_observed['gene']=='Nsp4']['r_value'].values[0]\n",
    "    ax8.axvline(x=nsp4_observed_r, color = \"#ffc59c\", linewidth=4)\n",
    "    \n",
    "    nsp4_pvalue= len(nsp4_expected[nsp4_expected['r_value']>nsp4_observed_r])/num_iterations\n",
    "    ax8.annotate(f'p-value: {round(nsp4_pvalue, 2)}', xy=((nsp4_observed_r+0.1), 0.65), \n",
    "                 xycoords='axes fraction', ha='left', size=11, fontweight='bold')\n",
    "\n",
    "\n",
    "\n",
    "#     ax1.set_title(f'S1 nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax2.set_title(f'S2 nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax3.set_title(f'RdRp nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax4.set_title(f'N nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax5.set_title(f'E nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax6.set_title(f'M nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax7.set_title(f'Nsp6 nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "#     ax8.set_title(f'Nsp4 nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "    \n",
    "\n",
    "#     ax7.set_xlabel('$R^2$', fontsize=14)\n",
    "#     ax8.set_xlabel('$R^2$', fontsize=14)\n",
    "    \n",
    "    for ax in [ax1, ax2, ax3, ax4, ax5, ax6, ax7, ax8]:\n",
    "        ax.set_ylabel('')\n",
    "        ax.set_xlabel('')\n",
    "        plt.setp(ax.get_yticklabels(), visible=False)\n",
    "        plt.setp(ax.get_xticklabels(), visible=False)\n",
    "#         ax.tick_params(axis='x', which='major', labelsize=12)\n",
    "    \n",
    "    \n",
    "    sns.despine(left=False, bottom=False)\n",
    "    plt.subplots_adjust(wspace=0.2, hspace=0.5)\n",
    "    \n",
    "\n",
    "    if filename:\n",
    "        fig.savefig(filename, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 439,
   "id": "7ca12d06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAegAAACxCAYAAAAPpi6HAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAgkUlEQVR4nO3df1RU953/8ecIjJ7QVgbF6tfIJrJxqKeIkLpkJqgJpydhc45r4Is91KxmrWbrcS3UGuwmbgzD+qOKjeHQNpxYJNumfs9Wv7qAjXZPgihEsfotS9VV96TGgySbEDtA4qzKIPP9I4dJR34zwNyZeT3O8Rxn7ufe++HOvD/vez+fez9j8ng8HkRERMRQJgS6AiIiItKbErSIiIgBKUGLiIgYkBK0iIiIASlBi4iIGJAStIiIiAFFDrRw9erVlJeXj1ddRAIqc8MNn9fH9kwNUE2MR22BjIYptgM+r/90elmAahIcBryCbmtrG696iIiBqS0QGX/q4hYRETEgJWihtLQUq9WKy+Ual/3dO3mdx+Nhx44d/NVf/RVPPvkk7777bp/r/fznP8dut/PYY49RVVXlfb+yspLFixdjt9v5+c9/7n1/z549WK1W77+ioqKx+YNExthYxmhLSwvLly8nJSWFtWvX9tlb0tbWxtq1a0lNTeWZZ56hpaUFgIyMDJ8Ys1qt/O53vxv1OoYrJWgZFx6Ph9raWp599lnOnj3rs+zo0aO88cYbfO973+OBBx5gw4YN3Llzx6dMU1MTxcXFfOtb38Jms/Hiiy/y0Ucf8dFHH7F582YWLVpETk4OxcXFNDU1AXD+/HkeeeQR9u3bR0VFBX/7t387bn+vSLAoLCzE6XRSVFREY2MjJSUlvcq8+uqrNDU14XA4aG1txeFwAFBcXExFRQUVFRWkpqZitVqZN2/eeP8JIUsJOkicOXMGq9VKSUkJNpuNjIwMTpw40W+5t99+G4CzZ89itVqpra3l8uXLZGdnk5SURHp6Om+++Wav9e89U8/IyCAvLw+A5uZmVqxYQUpKCjk5OZw/f95nn8ePH++1vTt37vDrX/+ap556iu9+97t89tlnTJ8+3adMXV0d06dPZ8WKFaxcuZKOjg5vkv3zMhEREaxbt47vfve7uN1uTp8+zenTp3G73Tz33HP8wz/8AxMmTKCurg6Px8OFCxe4ePEif//3f8/evXuJiYkZ/oEXGaJgjFG3201DQwOZmZksWbKE9PR0Tp482WufdXV1pKens2TJEp566ilOnz5NV1cXDz/8MHa7nfvuu4+mpia2bt3KpEmT/DuQ4qUEHWRqampwOBxYLBYKCgq4efMmnZ2duFwuXC4XqampxMfHe7uAq6uriYuLY+HChRw+fBiz2cyePXt46KGH+jxTHsimTZvo7u7G4XAwY8YM8vPzcbvdJCYmUlFRQXJysk/5P/3pTzz22GMUFxdjs9morKzk0KFDxMfH+5T7+OOPvckzNjYWgNbW1l5l7rvvPsxms0+Zjz/+GICYmBgmTpxIdHQ0ra2ttLW1cf/997N8+XJ2797Nf/zHf/DKK68M6+8VGYlgilGn04nb7faJv3tjDz6PtZ4yFosFt9uN0+n0Ln/ttddYvHixrp5H2YCPWYnxrFmzhieeeAKTycT69eu5cuUKBw4c4PDhwwDs2LGDrKwsysrKcDqdHDt2jGXLlhEREcHGjRt55513aGhooLm5mU8//XTQ/fWMF7tcLhobGwE4d+6cd/n169eZPXs2dru917omk4kJEybQ2dnJrVu3cLvd/e7DZDL57G+oZXr+f++y2NhYDh065C33m9/8RmNjMi6CKUZ7DCf+7vXhhx9y8uRJysrKBq2rDI8SdJDp6uoCvgioCRMmsG7dOnJzcwGIj4+ns7OT0tJStm/fTkdHB9nZ2QBs2LCB5uZm8vLymDRpEnv37u21/Z7t3r17F4Dbt2/77Dc3N5esrCxu3bpFVFQU06ZN67eusbGxHD9+nMrKSioqKjh06BBf//rXefXVV5k1a5a3XFxcHO+//z4AHR0dAL22GxcXh8vlwu12+5TpqWd7ezsTJ07k1q1bTJs2jevXr3PkyBGefvppZsyYQVdXF5GR+rrL2AumGLVYLERERNDe3g58Hn99lY+Li/OWaW9vJyoqytuTVV9fz8SJE7HZbEM/SDIkarGCTHl5OdHR0ezbt4+YmBjmzJlDdHR0r25ju91OdXU1ycnJJCQkANDQ0MCsWbO4c+cOtbW1wBdB3sNisQBQVVVFd3e3txtr8uTJzJ07l/r6etLS0ti/fz/Xrl2jpqaGjo4OLl68SGJiojdoe5jNZpYtW0ZOTg4nT56kvLyc//7v//ZJ0Ha7ncrKSn75y19y6tQpYmJiSE5OprW1lffee4/58+fz6KOP8rOf/Yyf/vSn3Lhxg6ioKOx2O263m4iICF5//XViY2O5e/cuixYtIjIykp/+9KdcvHiRjIwMTp06xZo1a0b1sxDpSzDFqNlsZsGCBbz11lskJCRQV1dHZmYmABcuXKC7u5t58+Zht9upqanhyJEjHD16FJvN5j3hbWpqIjExEbPZPPoHM8xpDDrIJCUlUVRURGtrK8XFxURHR/dZLicnB4CsrCzve3l5ebS0tLBt2zYefPBBAK5eveqz3tKlS7HZbOzevZtLly6RlJTkXVZSUsLMmTPZvHkzN27cYNeuXZjNZi5fvsyqVat63dj150wmE4sXL+YXv/gFCxYs8Fn2N3/zN3znO9/hJz/5CX/84x955ZVXmDhxIvX19axatYqWlha+8Y1vUFBQwIEDB6irq2Pbtm189atf5f7772fHjh3U19dz4MABCgoKmDdvHjNmzKC4uJhLly7hcDj467/+a9auXTu8gy0yAsEWo1u3bmXq1Km89NJLJCcnk5+fD8DOnTspLCwEoKCggPnz57NlyxamTJnCyy+/7F3/k08+6XXjp4wOk6e/QQcgOzvbZxxPAufMmTOsXLmSsrIyHn/88UBXJyRpqs/+qS0YnGJ0cJrqc3h0BS0iImJAGoMOEmlpaVy5ciXQ1RCRfihGZbQpQcuAKisr+308ajBRUVEsXbp0lGskIsPlTxyDYjlQlKBDnL+BGRkZyaJFi0a0bl8zEonI+HO73SOOY1AsB4oSdIjzNzBFRCIjIzl48OCI1tXV98gpQYuIGNxo9IT5Y6BZyAajq++RU4IOAv4Ep2bPEgl+6gkLT2q9g0CwBqc/3WKgrjERCW9K0DJm/OkWA3WNSegIdBe1BCd96iIiYyxYe8EksDSTmIiIiAHpCnocqHtLRMLVQPeiDOUelXC+F0Ut/zhQ95aIhCufe1F+fMJn2VDaxXC+F0UJWkRkEOoFk0DQt2aI9Czy+NPsRWIU6gWTQFDmGCIF6PjT7EUiEs7zKShBS0gaWVA/NhZVEYNQL1hwCuf5FEwej8fT38K0tDRmzpw5nvUZM21tbYGugoQRi8UypDLl5eXjUBv/GaEtUAxLoAwlnv3Zdn/twIAJWkRERAJDE5WIiIgYkBK0iIiIASlBi4iIGJAStIiIiAEpQYuIiBiQErSIiIgBKUGLiIgYkBK0iIiIASlBi4iIGJAStIiIiAEpQYuIiBiQErSIiIgBKUGLiIgY0IA/crp69eqg+Tk8GSM/ftT39cZ3A1MPCSi1BTIcGfte8Hld850dAapJcBvwClq/vyoioLZAJBDUxS0iImJAStCjrLS0FKvVisvlGvVtt7S0sHz5clJSUli7dm2/VzXd3d383d/9HdnZ2d73MjIysFqtPv9+97vfjXodRWTkjNp+3Llzh69//es+7YfT6Rz1OoovJeggUlhYiNPppKioiMbGRkpKSnqV+eSTT/j+97/P6dOnfd4vLi6moqKCiooKUlNTsVqtzJs3b7yqLiIB5k/7cfHiRdxuNz/60Y+87ciXv/zl8ap62ArLBH3mzBmsVislJSXYbDYyMjI4ceJEv+XefvttAM6ePYvVaqW2tpbLly+TnZ1NUlIS6enpvPnmm73Wv/dsOCMjg7y8PACam5tZsWIFKSkp5OTkcP78eZ99Hj9+3GdbbrebhoYGMjMzWbJkCenp6Zw8ebLXPgsLC/nwww9JSEjwef/hhx/Gbrdz33330dTUxNatW5k0adIIjp5IeAvH9uMPf/gDADt27GDDhg189NFHREVFDffQyTCFZYLuUVNTg8PhwGKxUFBQwM2bN+ns7MTlcuFyuUhNTSU+Pp6qqioAqquriYuLY+HChRw+fBiz2cyePXt46KGH+jwbHcimTZvo7u7G4XAwY8YM8vPzcbvdJCYmUlFRQXJysk95p9OJ2+0mJiYGgNjYWFpbW3ttNz8/n1//+tdMmTKlz/2+9tprLF68WFfPIn4Kp/bD7XaTkpLCzp07sdlsbNmyhZaWlmHVWYZvwMesQt2aNWt44oknMJlMrF+/nitXrnDgwAEOHz4MfH62mJWVRVlZGU6nk2PHjrFs2TIiIiLYuHEj77zzDg0NDTQ3N/Ppp58Ouj+PxwOAy+WisbERgHPnznmXX79+ndmzZ2O32/vdhslk8tnWvebMmdPvuh9++CEnT56krKxs0LqKyMDCqf147rnneO655wCYMWMGR48e5fe//z3333//oPWWkQvrBN3V1QV88aWdMGEC69atIzc3F4D4+Hg6OzspLS1l+/btdHR0eG+c2LBhA83NzeTl5TFp0iT27t3ba/s927179y4At2/f9tlvbm4uWVlZ3Lp1i6ioKKZNm9ZvXS0WCxEREbS3twPQ0dExYPm+1NfXM3HiRGw227DWE5Hewqn9qK6u5u7duzz99NPe+kRGhnX6GBdhfYTLy8uJjo5m3759xMTEMGfOHKKjo4mPj/cpZ7fbqa6uJjk52Ts209DQwKxZs7hz5w61tbXAF4HUw2KxAFBVVUV3d7f3rsfJkyczd+5c6uvrSUtLY//+/Vy7do2amho6Ojq4ePEiiYmJxMbGerdlNptZsGABb731FgkJCdTV1ZGZmQnAhQsX6O7uHrTbuqmpicTERMxm88gPmogA4dV+nDt3jsrKSgB++9vf8qUvfYn09HQ/jp4MRViPQSclJVFUVERrayvFxcVER0f3WS4nJweArKws73t5eXm0tLSwbds2HnzwQQCuXr3qs97SpUux2Wzs3r2bS5cukZSU5F1WUlLCzJkz2bx5Mzdu3GDXrl2YzWYuX77MqlWraGpq6lWPrVu3MnXqVF566SWSk5PJz88HYOfOnRQWFg76937yySdMnz590HIiMrhwaj82btzIwoULcTgcvPfee5SWlvKVr3xl8IMkfjF5+huMALKzszl06NB41mdcnDlzhpUrV1JWVsbjjz8e6OoYm6b6FEK3LRgJtR+D01SfoyOsr6BFRESMKizHoNPS0rhy5UqgqyEiQSgc24/KykrcbveI1z948CBRUVEsXbp0FGsV+sIyQYcyfwJJASQSevxNrvD5HduLFi0acvmfHfl/Pq8XLVrU58QoMjAl6BDjdruHFUh/TgEkEnr8aRMksDQGLSIiYkC6gjYYf7ujNHmAiEhoUGtuMIHsjoqMjOTgwYM+7+XcU+be5X9OY9giIqNHCVq8+pzD95d7fF4OdPKgMWyR0TVaN3hJcNInJyJiUKF0g1dfPXTDFW69dErQY8Cfs16d7YpIKBroV7aGKtx66ZQNxkAonfWKiEhg6DErERERA9IVtIwaf8aYwm1sSUKfbvASf+nT74OeRR4Zf8aYwm1sSUKfhrrEX+GZSQahwBIRkUBTghZD8PcRDHWRy2hS97QxhdujWiH7DdKjTsHF30cw1EUuo0m9aMYUbo9qGTYTjcY4sAIsfOgKXHro6lcGEkxX4SaPx+Ppb2FaWhozZ84c80r0pb29nQGqJjKqTCYTMTEx47pPi8VCeXn5uO5zpMarLWhraxvzfYj4azTbi4HagQETtIiIiASGJioRERExICVoERERA1KCFhERMSAlaBEREQNSghYRETEgJWgREREDUoIWERExICVoERERA1KCFhERMSAlaBEREQNSghYRETEgJWgREREDUoIWERExoAF/9HT16tVB83N44Szl+W/7vG7c/X8CVBMJVWoLZLz88r/qfV6vmJMeoJoE3oAJWr/NGhj+/uB8ZWXluPyYuIQPtQUi42/ABC2B4Xa7WbRo0ZDLb2v4v73WFxGR4BbWY9ClpaVYrVZcLteob7ulpYXly5eTkpLC2rVr+7wCuXDhAjk5OSQnJ5OVlcUf/vAHADweD2+88QZPP/00S5cu5d///d9HvX4iEpoC3a61tbWxdu1aUlNTeeaZZ2hpafGu++yzz5KSkkJ2djaXLl0a9fqFmrBO0GOpsLAQp9NJUVERjY2NlJSU9Crz/PPP093dzY9//GPu3r3LD37wAwDOnTvHG2+8wbPPPsv8+fPZuXPnmASbiMhwDKVde/XVV2lqasLhcNDa2orD4QDA4XDwwQcf4HA46OjoYPPmzeNd/aBj+AR95swZrFYrJSUl2Gw2MjIyOHHiRL/l3n77bQDOnj2L1WqltraWy5cvk52dTVJSEunp6bz55pu91r/3rDMjI4O8vDwAmpubWbFiBSkpKeTk5HD+/HmffR4/ftxnW263m4aGBjIzM1myZAnp6emcPHnSp0xXVxfLli3jxRdf5Jvf/CY2m42PPvqI7u5uLly4wNy5c8nKyuL555+noqKCSZMm+X8wRcQQQrVdA6irqyM9PZ0lS5bw1FNPcfr0abq6unjllVf41a9+hd1u5ytf+QpRUVH+HcQwYPgE3aOmpgaHw4HFYqGgoICbN2/S2dmJy+XC5XKRmppKfHw8VVVVAFRXVxMXF8fChQs5fPgwZrOZPXv28NBDD/V51jeQTZs20d3djcPhYMaMGeTn5+N2u0lMTKSiooLk5GSf8k6nE7fbTUxMDACxsbG0trb6lImMjGT16tV84xvf4IMPPqCqqoq0tDQmTJhAe3s7//M//8O3v/1tcnJyOHHiBBERESM/eCJiSKHWrgG0trZ6y1gsFtxuN06nky9/+ct89atf5cknn+S9997jhz/84fAPWJgJmpvE1qxZwxNPPIHJZGL9+vVcuXKFAwcOcPjwYQB27NhBVlYWZWVlOJ1Ojh07xrJly4iIiGDjxo288847NDQ00NzczKeffjro/jweDwAul4vGxkbg867nHtevX2f27NnY7fZ+t2EymXy21ZcPPviAFStW4Ha7efHFF73lP/jgA/7pn/6J3//+97z++us8/PDDzJ07d9B6i0jwCMV2zePxeMvcq7u7m9LSUv7lX/6FvLw8jh07xpe+9KVB6x2ugiZBd3V1AV98OSZMmMC6devIzc0FID4+ns7OTkpLS9m+fTsdHR1kZ2cDsGHDBpqbm8nLy2PSpEns3bu31/Z7tnv37l0Abt++7bPf3NxcsrKyuHXrFlFRUUybNq3fulosFiIiImhvbwego6Ojz/Iff/wxK1euxOVyUVFRQUJCAgCTJ09mypQpLF68mAceeIBDhw5x7do1JWiREBOK7VpcXJy3THt7O1FRUcTGxlJXV8fkyZOx2+20t7dTW1vL+++/T1JS0rCOWTgJmgRdXl5OdHQ0+/btIyYmhjlz5hAdHU18fLxPObvdTnV1NcnJyd6E19DQwKxZs7hz5w61tbXAF1/YHhaLBYCqqiq6u7txOp3A58ly7ty51NfXk5aWxv79+7l27Ro1NTV0dHRw8eJFEhMTiY2N9W7LbDazYMEC3nrrLRISEqirqyMzMxP4/M7t7u5u5s2bxw9/+ENaWlr4/ve/z2effcapU6dIS0vja1/7GseOHaOyspKrV69iMpmYM2fOmBxXEQmcUGzX7HY7NTU1HDlyhKNHj2Kz2YiMjOT111+npaWFF154gf379xMTE8Ps2bPH5LiGiqAZg05KSqKoqIjW1laKi4uJjo7us1xOTg4AWVlZ3vfy8vJoaWlh27ZtPPjggwBcvXrVZ72lS5dis9nYvXs3ly5d8jmrKykpYebMmWzevJkbN26wa9cuzGYzly9fZtWqVTQ1NfWqx9atW5k6dSovvfQSycnJ5OfnA7Bz504KCwv54x//yOnTp4HP73pctWoVq1at4vbt2zzyyCN861vfYu/evdTV1fGDH/yAv/zLv/Tj6ImIEQ23XfuLv/gLDh48yMGDB1m8eDHvv/8+W7ZswWw2A1BWVsZ//ud/AvBv//ZveDweEhIS+NGPfsTRo0eZNWuWd5tj0a4BFBQUMH/+fLZs2cKUKVN4+eWXAdi+fTv3338/mzZt4ubNm7z22mv9/r3yOZNngAHS7OxsDh06NJ716eXMmTOsXLmSsrIyHn/88YDWZbwcPHhwWBOVPLkr3+f15kf+tzegRUaDEdqCUDKSdm247UJfTp48afi2QVN9fiForqBFRETCieHHoNPS0rhy5UqgqyEiMmrUrslQGD5By/BFRkZy8ODBEa0bFRWlH9oQETEAJegQNNAzjIPpa2YgEQkN/py899BJ/PhRghYRCRP+nLz30En8+FGCFhEZQ/7+vnuPyEg11+FGn7iIyBga7u+7i/RQgh4j/pw160xZRESUCcaIzppFRMQfmqhERETEgJSgRUREDEhd3CIi/RiNO7B1T4mMlL45IiL90L0kvY3GZCegCU+GQglaRESGbDQmOwFNeDIUStD98LdrK1i7tTSPt4iIMQRnFhkH4dq1pXm8JRRo9i7j6/diYN50n5cDXTCE+kVBSH/7NFnI+PJ3bMqfYPO3QQ71QA834XqCHUz6uxj4bft/+bwe6HM8depUSI+Hmzwej6e/hWlpacycOXM869NLW1tbQPcv4cFkMhETEzOu+7RYLJSXl4/rPkdqqG2B4lWCUSDiv8dA7cCACVpEREQCQxOViIiIGJAStIiIiAEpQYuIiBiQErSIiIgBKUGLiIgYkBK0iIiIASlBi4iIGJAStIiIiAEpQYuIiBiQErSIiIgBKUGLiIgYkBK0iIiIASlBi4iIGNCAP3q8evXqoPk5PAkv9dkmn9fph/SjbGNJbYGMuVP3fL/sqwNTDwMZ8Apav+0qIqC2QCQQ1MUtIiJiQErQYay0tBSr1YrL5Rr1bbe0tLB8+XJSUlJYu3Ztn1dgbW1trF27ltTUVJ555hlaWlp8lr/wwgukpaWNet1EQpHR4/mTTz4hLS2Nf/zHfxz1+oUqJWgZE4WFhTidToqKimhsbKSkpKRXmVdffZWmpiYcDgetra04HA4APvvsM4qKijh06NB4V1tE+uBPPPfYvHkz7e3t41Tj0KAEbXBnzpzBarVSUlKCzWYjIyODEydO9Fvu7bffBuDs2bNYrVZqa2u5fPky2dnZJCUlkZ6ezptvvtlr/XvPvjMyMsjLywOgubmZFStWkJKSQk5ODufPn/fZ5/Hjx3225Xa7aWhoIDMzkyVLlpCens7Jkyd77bOuro709HSWLFnCU089xenTp+nq6qK0tJR3332XlJQU/w6eiMGEYzwD7N+/n4sXLxITEzPygxeGlKCDRE1NDQ6HA4vFQkFBATdv3qSzsxOXy4XL5SI1NZX4+HiqqqoAqK6uJi4ujoULF3L48GHMZjN79uzhoYce6vPsdyCbNm2iu7sbh8PBjBkzyM/Px+12k5iYSEVFBcnJyT7lnU4nbrfbG4yxsbG0trb22m5ra6u3jMViwe1243Q6yc3Npbq6mgceeGDYx0kkGIRTPF+7do3i4mL++Z//mejo6OEfrDA24GNWYhxr1qzhiSeewGQysX79eq5cucKBAwc4fPgwADt27CArK4uysjKcTifHjh1j2bJlREREsHHjRt555x0aGhpobm7m008/HXR/Hs/njy25XC4aGxsBOHfunHf59evXmT17Nna7vd9tmEwmn231tY+eMn9u9uzZg9ZPJJiFSzx7PB42bdrEN7/5TR599FE8Hg93796lq6uLyEiln8HoCAWJnq6ingCYMGEC69atIzc3F4D4+Hg6OzspLS1l+/btdHR0kJ2dDcCGDRtobm4mLy+PSZMmsXfv3l7b79nu3bt3Abh9+7bPfnNzc8nKyuLWrVtERUUxbdq0futqsViIiIjwjjd1dHT0WT4uLs5bpr29naioKGJjY4d1XESCUbjE8+3bt2lqaqKpqcnbG1BVVUV8fDzf+973hn7AwpQSdJAoLy8nOjqaffv2ERMTw5w5c4iOjiY+Pt6nnN1up7q6muTkZBISEgBoaGhg1qxZ3Llzh9raWuCLwO1hsViAz4Onu7sbp9MJwOTJk5k7dy719fWkpaWxf/9+rl27Rk1NDR0dHVy8eJHExESfxGo2m1mwYAFvvfUWCQkJ1NXVkZmZCcCFCxfo7u5m3rx52O12ampqOHLkCEePHsVms+msWsJCuMTzjBkz+Nd//VfvttavX8/cuXNZtmzZ6B7QEKUx6CCRlJREUVERra2tFBcX9zuWk5OTA0BWVpb3vby8PFpaWti2bRsPPvggAFevXvVZb+nSpdhsNnbv3s2lS5dISkryLispKWHmzJls3ryZGzdusGvXLsxmM5cvX2bVqlU0NTX1qsfWrVuZOnUqL730EsnJyeTn5wOwc+dOCgsLASgoKGD+/Pls2bKFKVOm8PLLL4/8AIkEkXCJZ7PZzPz5873/zGYzsbGxTJ8+feQHL4yYPP0NKADZ2dl61CXAzpw5w8qVKykrK+Pxxx8PdHUMQ1N9ji+1BaND8TwATfXZi66gRUREDEgDfgaXlpbGlStXAl0NERkFimcZDiVoEZEgVFlZidvtHvH6UVFRLF26dBRrJKNNCdrg/A3C0aJgFjEWt9vNokWLRrx+X7OBibEoQRucv0E4WhTMIqElMjKSgwcPjnh9nbSPPSVoEZEwNNCsYUOhk/axpwQtQ6KzbZHR5e/wlSb1CX36hMfQaIwfGyUIdbYtMrqMMnwlxmWM1j9EKQBFRGSklKBFRGTY/Bn20pDX0ChBi4jIsPkz7KUhr6FRghYRGQHd5CVjTd8QEZER0D0mMtaUoGVc+PuYFmjcSiRU9NUe5Pwv3zIDtRfh0hYoQQ9AXVijx9/HtEDjViKhos/24L1qn5cD9U6ES1ugDDIAdWGJhDZ/TsJ1Ai5jLWS/YaE0SYiI9G00erl0Eh58wmVmQ0NmoNFKrgq80PLnQTn9nmVDDdZgCUwZGvVyhadwmdnQ5PF4PP0tTEtLY+bMmcPeaFtbm1+VEhkrJpOJmJiYQFcDAIvFQnl5eaCrMSQjbQsGo7ZCAiFY2oEBE7SIiIgExoRAV0BERER6U4IWERExICVoERERA1KCFhERMSAlaBEREQP6/1mAEj65qZxYAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 612x216 with 8 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_fig1_supp_b(['growth_rate_stats/growth_stats_50it.json', \n",
    "            'growth_rate_stats/growth_stats_50itmore.json', \n",
    "            'growth_rate_stats/growth_stats_200its.json', \n",
    "            'growth_rate_stats/growth_stats_400its.json', \n",
    "            'growth_rate_stats/growth_stats_300its.json'], \n",
    "                 ['growth_rate_stats/growth_stats-supp_100its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_150its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_250its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_300its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_200itsmore.json'], \n",
    "                 filename = 'plots/fig1_supp2stats_notitles.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb643d32",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-kernel",
   "language": "python",
   "name": "ncov-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
