{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "33194542",
   "metadata": {},
   "outputs": [],
   "source": [
    "from augur.utils import json_to_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.gridspec as gridspec\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Seq import MutableSeq\n",
    "from collections import Counter\n",
    "import requests\n",
    "import random\n",
    "import math\n",
    "import json\n",
    "from scipy import stats"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71e410d0",
   "metadata": {},
   "source": [
    "Make an expectation for wait times by randomly putting mutations on tree and then calculating wait times"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "8f4a46bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_url = 'https://nextstrain-blab.s3.amazonaws.com/ncov_adaptive-evolution_2021-05-15.json'\n",
    "\n",
    "tree_json = requests.get(tree_url).json()\n",
    "\n",
    "#Put tree in Bio.Phylo format\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30469c3d",
   "metadata": {},
   "source": [
    "Find the actual number of S1 and RdRp mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c874c9be",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to find path from root to clade\n",
    "def get_parent(tree, child_clade):\n",
    "    node_path = tree.get_path(child_clade)\n",
    "    return node_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "13453d9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def consolidate_deletions(mutation_list):\n",
    "    \"\"\"\n",
    "    For deletion mutations, consider adjacent sites as part of the same deletion\n",
    "    \"\"\"\n",
    "    \n",
    "    without_deletions = [x for x in mutation_list if x[-1]!='-' and x[0]!='-']\n",
    "    #consolidate deletions and reversions\n",
    "    deletions_only = [x for x in mutation_list if x[-1]=='-' or x[0]=='-']\n",
    "    deletions_only.sort(key=lambda x:x[1:-1])\n",
    "    \n",
    "    \n",
    "    #keep track of start of separate deletions\n",
    "    separate_deletions = []\n",
    "\n",
    "    # if there are deletions, count a run of consecutive sites as a single deletion/mutation\n",
    "    if len(deletions_only) != 0:\n",
    "        separate_deletions.append(deletions_only[0])\n",
    "\n",
    "        deletion_tracker = int(deletions_only[0][1:-1])\n",
    "        \n",
    "        for deletion in deletions_only[1:]:\n",
    "\n",
    "            deleted_pos = int(deletion[1:-1])\n",
    "            if deleted_pos == deletion_tracker+1:\n",
    "                pass\n",
    "            else:\n",
    "                separate_deletions.append(deletion)\n",
    "            deletion_tracker = deleted_pos\n",
    "    \n",
    "    consolidated_mutation_list = separate_deletions + without_deletions\n",
    "            \n",
    "    return consolidated_mutation_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "ab10b892",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "# make dictionary with gene name as key and reference sequence of that gene as value\n",
    "reference_sequence_aa = {}\n",
    "reference_sequence_nt = {}\n",
    "\n",
    "\n",
    "# make dictionary giving gene by genomic location \n",
    "reference_gene_locations = {}\n",
    "\n",
    "# make dictionary saying what codon within the gene a certain genomic location falls within\n",
    "# and whether the mutation is at pos 0, 1 or 2 within codon\n",
    "reference_gene_codon = {}\n",
    "\n",
    "for record in SeqIO.parse(open(\"reference_seq_edited.gb\",\"r\"), \"genbank\"):\n",
    "    genome_seq = record.seq\n",
    "    for feature in record.features:\n",
    "        if feature.type == 'CDS':\n",
    "            # allow RdRp to overwrite Orf1a and Orf1b, \n",
    "            # to take care of changed reading frame due to  ribosome slippage\n",
    "            # S1 and S2 will also overwrite spike\n",
    "            for pos in range(int(feature.location.start), int(feature.location.end)):\n",
    "                reference_gene_locations[pos] = feature.qualifiers['gene'][0]\n",
    "                codon_num = math.floor((pos-feature.location.start)/3)\n",
    "                pos_in_codon = ((pos-feature.location.start)-codon_num*3)\n",
    "                reference_gene_codon[pos] = (codon_num, pos_in_codon)\n",
    "                \n",
    "            gene_seq = feature.location.extract(record.seq)\n",
    "            reference_sequence_nt[feature.qualifiers['gene'][0]] = gene_seq\n",
    "            gene_seq_aa = gene_seq.translate()\n",
    "            reference_sequence_aa[feature.qualifiers['gene'][0]] = gene_seq_aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "7b49fff7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# make dictionary with length of each gene\n",
    "gene_lengths_aa = {k:len(v) for k,v in reference_sequence_aa.items()}\n",
    "gene_lengths_aa['Nsp6'] = 290\n",
    "gene_lengths_aa['Nsp4'] = 485"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "239fd561",
   "metadata": {},
   "outputs": [],
   "source": [
    "# dictionary of list of colors for emerging lineages, depending on total number of lineages\n",
    "color_palettes = {8: ['#4068CF','#5098B9','#6CB28C','#94BD62','#BFBB47','#DFA53B','#E67131','#DB2823'], \n",
    "                  9: ['#3E5DD0','#4A8CC2','#60AA9E','#80B974','#A6BE55','#CBB742','#E29D39','#E56A2F','#DB2823'], \n",
    "                  10: ['#3F52CD','#4681CA','#57A1AD','#70B487','#90BC65','#B4BD4C','#D3B240','#E59638','#E4642E','#DB2823'], \n",
    "                  11: ['#3F47C9','#4274CE','#4F97BB','#64AC99','#7EB976','#9EBE5A','#BEBB48','#D9AE3E','#E69036','#E35F2D','#DB2823'], \n",
    "                  12: ['#403CC5','#4067CF','#4A8BC3','#5AA4A9','#6FB488','#8BBB69','#A9BD53','#C7B944','#DDA93C','#E68A35','#E35C2C','#DB2823'], \n",
    "                  13: ['#4433BE','#3E5ACF','#457FCB','#529AB6','#64AD98','#7BB77A','#96BD60','#B3BD4D','#CDB642','#DFA43B','#E68434','#E2582C','#DB2823'], \n",
    "                  14: ['#492AB5','#3F4CCB','#4271CE,''#4C8FC0','#5AA5A8','#6DB38A','#85BA6F','#A0BE59','#BBBC49','#D2B340','#E19F3A','#E68033','#E2562B','#DB2823'], \n",
    "                  15: ['#4D21AD','#403FC6','#3F63CF','#4783C8','#539BB5','#63AC9B','#77B67F','#8EBC66','#A8BD53','#C1BA47','#D6B03F','#E39C39','#E67C33','#E1532B','#DB2823'], \n",
    "                  16: ['#571EA2','#4334BF','#3F55CE','#4376CD','#4C91C0','#59A4A9','#6AB18F','#7FB975','#97BD5F','#AFBD4F','#C7B944','#D9AD3D','#E49838','#E67932','#E1512A','#DB2823'], \n",
    "                  17: ['#5E1D9D','#462EB9','#3F4CCB','#416CCE','#4887C6','#539CB3','#62AB9C','#74B582','#89BB6B','#A0BE59','#B7BD4B','#CCB742','#DDAA3C','#E69537','#E67631','#E14F2A','#DB2823']}\n",
    "                      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "92b12cb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def nuc_changes_from_reference(muts_on_path):\n",
    "    \"\"\"\n",
    "    From all the of the nucleotide changes that have occurred on the path from root to branch, \n",
    "    find the most recent nuc mutation at each site (giving the genotype at the branch)\n",
    "    \"\"\"\n",
    "    \n",
    "    final_muts_from_ref = {}\n",
    "\n",
    "    # overwrites genotypes at pos in historical order\n",
    "    for x in muts_on_path:\n",
    "        x_pos = int(x[1:-1])\n",
    "        final_muts_from_ref[x_pos] = x[-1]\n",
    "        \n",
    "\n",
    "    return final_muts_from_ref"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "fbd74bfd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def determine_synonymous(nuc_muts_on_branch, parent_diffs_from_ref):\n",
    "    \n",
    "    parent_diffs_pos = [int(k) for k,v in parent_diffs_from_ref.items()]\n",
    "\n",
    "    \n",
    "    # make dictionary of synonymous (and noncoding) mutations to add to tree\n",
    "    syn_muts = {}\n",
    "    \n",
    "    # don't care about deletions because they are obviously not synonymous\n",
    "    for mut in nuc_muts_on_branch:\n",
    "        if mut[-1]!= '-' and mut[0]!='-':\n",
    "            mut_pos = int(mut[1:-1])\n",
    "            # find what gene this mut happens in\n",
    "            if (mut_pos-1) in reference_gene_locations.keys():\n",
    "                mut_gene = reference_gene_locations[mut_pos-1]\n",
    "                mut_codon_num = reference_gene_codon[mut_pos-1][0]\n",
    "                mut_codon_pos = reference_gene_codon[mut_pos-1][1]\n",
    "                \n",
    "                # find the reference sequence of the codon this mutation occurs in\n",
    "                codon_ref_aa = reference_sequence_aa[mut_gene][mut_codon_num]\n",
    "                \n",
    "                codon_ref_nt = reference_sequence_nt[mut_gene][(mut_codon_num*3):(mut_codon_num*3+3)]\n",
    "                \n",
    "                # check if a mutation occurred within the same codon in a parent\n",
    "                # and if so, change the reference codon sequence accordingly, \n",
    "                # to tell whether the mutation at this branch is synonymous or not\n",
    "                codon_genome_pos = list(range((mut_pos-1-mut_codon_pos),(mut_pos-1-mut_codon_pos+3)))\n",
    "                \n",
    "                parent_codon = codon_ref_nt\n",
    "                for parent_diff in parent_diffs_pos:\n",
    "                    parent_diff_zero_based = parent_diff-1\n",
    "                    if parent_diff_zero_based in codon_genome_pos:\n",
    "                        parent_diff_pos = codon_genome_pos.index(parent_diff_zero_based)\n",
    "                        parent_codon = MutableSeq(str(codon_ref_nt))\n",
    "                        parent_codon[parent_diff_pos] = parent_diffs_from_ref[parent_diff]\n",
    "                        parent_codon = parent_codon.toseq()\n",
    "                \n",
    "                \n",
    "                codon_mutated = MutableSeq(str(parent_codon))\n",
    "                #if deletion (or seq error) has happened at neighboring nucleotide\n",
    "                if '-' in codon_mutated:\n",
    "                    pass\n",
    "                else:\n",
    "                    codon_mutated[mut_codon_pos] = mut[-1]\n",
    "                    codon_mutated = codon_mutated.toseq()\n",
    "                    codon_mutated_translation = codon_mutated.translate()\n",
    "\n",
    "                    if str(codon_ref_aa) == str(codon_mutated_translation):\n",
    "                        if mut_gene in syn_muts.keys():\n",
    "                            syn_muts[mut_gene] += [mut]\n",
    "                        else:\n",
    "                            syn_muts[mut_gene] = [mut]\n",
    "                        \n",
    "                \n",
    "\n",
    "            else:\n",
    "                if 'noncoding' in syn_muts.keys():\n",
    "                    syn_muts['noncoding'] += [mut]\n",
    "                else:\n",
    "                    syn_muts['noncoding'] = [mut]\n",
    "                    \n",
    "    return syn_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2c44c1af",
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_syn_mut_attribute(tree):\n",
    "    \n",
    "    for node in tree.find_clades():\n",
    "\n",
    "        node.node_attrs['syn_muts'] = {}\n",
    "\n",
    "        # only care if this branch has some nucleotide mutations\n",
    "        if hasattr(node, 'branch_attrs'):\n",
    "            if 'nuc' in node.branch_attrs['mutations']:\n",
    "\n",
    "                nuc_muts_on_branch = node.branch_attrs['mutations']['nuc']\n",
    "\n",
    "                node_path = get_parent(tree, node)\n",
    "\n",
    "                nucleotide_mut_path = []\n",
    "\n",
    "                # find all nucleotide mutations that happened in parents, \n",
    "                # in case they affect codons mutated on this branch\n",
    "                for parent in node_path[-1]:\n",
    "                    if hasattr(parent, 'branch_attrs'):\n",
    "                        if 'nuc' in parent.branch_attrs['mutations']:\n",
    "                            nucleotide_mut_path+=parent.branch_attrs['mutations']['nuc']\n",
    "\n",
    "                parent_diffs_from_ref = nuc_changes_from_reference(nucleotide_mut_path)\n",
    "\n",
    "                syn_muts_dict = determine_synonymous(nuc_muts_on_branch, parent_diffs_from_ref)\n",
    "\n",
    "                node.node_attrs['syn_muts'] = syn_muts_dict\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a26817cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "add_syn_mut_attribute(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "f202af18",
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in tree.find_clades(terminal=False):\n",
    "\n",
    "    node.s1_nonsyn_at_node = 0\n",
    "    node.s1_syn_at_node = 0\n",
    "    node.rdrp_nonsyn_at_node = 0\n",
    "\n",
    "    if hasattr(node, \"node_attrs\") and 'S1' in node.node_attrs['syn_muts']:\n",
    "        node.s1_syn_at_node = len(node.node_attrs['syn_muts']['S1'])\n",
    "\n",
    "\n",
    "    if hasattr(node, 'branch_attrs'):\n",
    "\n",
    "        s1_nonsyn_at_this_node = []\n",
    "        if \"S\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"S\"]:\n",
    "                if int(mut[1:-1]) in range(14,686):\n",
    "                    s1_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "        s1_consolidated = consolidate_deletions(s1_nonsyn_at_this_node)\n",
    "        node.s1_nonsyn_at_node = len(s1_consolidated)\n",
    "\n",
    "\n",
    "\n",
    "        rdrp_nonsyn_at_this_node = []\n",
    "        if \"ORF1a\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"ORF1a\"]:\n",
    "                if int(mut[1:-1]) in range(4492,4401):\n",
    "                    rdrp_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "\n",
    "        if \"ORF1b\" in node.branch_attrs[\"mutations\"]:\n",
    "            for mut in node.branch_attrs[\"mutations\"][\"ORF1b\"]:\n",
    "                if int(mut[1:-1]) in range(1,923):\n",
    "                    rdrp_nonsyn_at_this_node.append(mut)\n",
    "\n",
    "        rdrp_consolidated = consolidate_deletions(rdrp_nonsyn_at_this_node)\n",
    "        node.rdrp_nonsyn_at_node = len(rdrp_consolidated)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "f9ada0b7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nodes with S1 nonsyn mutations: 277\n",
      "Total number of S1 nonsyn mutations: 353\n",
      "Nodes with S1 syn mutations: 100\n",
      "Total number of S1 syn mutations: 104\n",
      "Nodes with RdRp mutations: 98\n",
      "Total number of RdRp mutations: 101\n",
      "Time window: 2019.982929101653-2021.3404645014461\n"
     ]
    }
   ],
   "source": [
    "# find number of nodes that have S1 mutations\n",
    "# consider only internal branches\n",
    "nodes_with_s1_nonsyn = 0\n",
    "total_s1_nonsyn = 0\n",
    "nodes_with_s1_syn = 0\n",
    "total_s1_syn = 0\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    # look only at nodes with at least 5 descendents, to exclude weird stuff at tips and sequencing errors\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        if node.s1_nonsyn_at_node!= 0:\n",
    "            nodes_with_s1_nonsyn+=1\n",
    "            total_s1_nonsyn += node.s1_nonsyn_at_node\n",
    "        if node.s1_syn_at_node!=0:\n",
    "            nodes_with_s1_syn+=1\n",
    "            total_s1_syn += node.s1_syn_at_node\n",
    "        \n",
    "        \n",
    "print(f'Nodes with S1 nonsyn mutations: {nodes_with_s1_nonsyn}')\n",
    "print(f'Total number of S1 nonsyn mutations: {total_s1_nonsyn}')\n",
    "print(f'Nodes with S1 syn mutations: {nodes_with_s1_syn}')\n",
    "print(f'Total number of S1 syn mutations: {total_s1_syn}')\n",
    "\n",
    "\n",
    "# find number of nodes that have RdRp mutations\n",
    "nodes_with_rdrp_muts = 0\n",
    "total_rdrp_muts = 0\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    # look only at nodes with at least 5 descendents, to exclude sequencing errors\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        if node.rdrp_nonsyn_at_node!= 0:\n",
    "            nodes_with_rdrp_muts+=1\n",
    "            total_rdrp_muts += node.rdrp_nonsyn_at_node\n",
    "        \n",
    "print(f'Nodes with RdRp mutations: {nodes_with_rdrp_muts}')\n",
    "print(f'Total number of RdRp mutations: {total_rdrp_muts}')\n",
    "\n",
    "first_timepoint = tree.root.node_attrs['num_date']['value']\n",
    "last_timepoint = 0\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    if node.node_attrs['num_date']['value'] > last_timepoint:\n",
    "        last_timepoint = node.node_attrs['num_date']['value']\n",
    "        \n",
    "print(f'Time window: {first_timepoint}-{last_timepoint}')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ed5c9f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "total_branchs = 0\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    if len(node.get_terminals()) >=3:\n",
    "        total_branchs+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a4da7f8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# find all branch lengths\n",
    "root_date = tree.root.node_attrs['num_date']['value']\n",
    "\n",
    "# keep track of parts of paths that have already been considered\n",
    "paths_segments_already_used_all = []\n",
    "\n",
    "# store all wait times between branchs\n",
    "branch_lengths_all = []\n",
    "branch_names_all = []\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "    \n",
    "    if len(node.get_terminals()) >=3:\n",
    "\n",
    "\n",
    "        node_path = get_parent(tree, node)\n",
    "\n",
    "        # find the branch length (in decimal years) between node and previous node\n",
    "        path_dates = [n.node_attrs['num_date']['value'] for n in node_path]\n",
    "\n",
    "        if len(path_dates) > 1:\n",
    "            path_dates_with_root = [root_date] + path_dates\n",
    "            branch_lengths = [j-i for i, j in zip(path_dates_with_root[:-1], path_dates_with_root[1:])]\n",
    "\n",
    "    #         branch_lengths = [j-i for i, j in zip(path_dates[:-1], path_dates[1:])]\n",
    "\n",
    "\n",
    "            for n_index in range(len(node_path)):\n",
    "\n",
    "                # name of the node at the end of the branch\n",
    "                # branch length will be assigned to this node because \n",
    "                # it is the length of the branch leading to this node\n",
    "                branch_name = node_path[n_index].name\n",
    "                if n_index==0:\n",
    "                    preceding_node_name = tree.root.name\n",
    "                else:\n",
    "                    preceding_node_name = node_path[n_index-1].name\n",
    "\n",
    "\n",
    "                if [preceding_node_name, branch_name] not in paths_segments_already_used_all:\n",
    "\n",
    "                    branch_length = branch_lengths[n_index]\n",
    "\n",
    "                    if branch_length<0:\n",
    "                        branch_length = 0.0\n",
    "\n",
    "                    branch_lengths_all.append(branch_length)\n",
    "                    branch_names_all.append(branch_name)\n",
    "    #                 branch_names_all.append([root_name, branch_name])\n",
    "                    paths_segments_already_used_all.append([preceding_node_name, branch_name])\n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfea755f",
   "metadata": {},
   "source": [
    "Throw each of the S1 mutations onto a random branch of the tree by picking a random branch X times (where X= total number of S1 mutations). Can result in multiple hits on same branch (as with empirical data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "0fbbc586",
   "metadata": {},
   "outputs": [],
   "source": [
    "def randomize_mutations_on_tree_multinomial(tree, num_mutations, branch_lengths, branch_names):\n",
    "    \n",
    "    # multinomial draw to randomize mutations on tree\n",
    "    branch_lengths_normalized = [x/sum(branch_lengths) for x in branch_lengths]\n",
    "\n",
    "    num_branches = len(branch_lengths_normalized)\n",
    "\n",
    "    hits = list(np.random.multinomial(num_mutations, branch_lengths_normalized, size=1))[0]\n",
    "    \n",
    "    muts_on_branch = dict(zip(branch_names, hits))\n",
    "\n",
    "    for node in tree.find_clades(terminal=False):\n",
    "        if len(node.get_terminals()) >=3:\n",
    "        # no branch length at root\n",
    "            if node.name == 'NODE_0000000':\n",
    "                node.random_muts = 0\n",
    "\n",
    "            else:\n",
    "                node.random_muts = muts_on_branch[node.name] \n",
    "    return tree\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "9f00585f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_linege_names(old_name):\n",
    "    \n",
    "#     old_to_new = {'A.23.1': 'A.23.1', 'B.1.1.7': 'B.1.1.7 (Alpha)', 'B.1.351': 'B.1.351 (Beta)', \n",
    "#                   'B.1.427+B.1.429': 'B.1.427/429 (Epsilon)', 'B.1.525': 'B.1.525 (Eta)', 'B.1.526': 'B.1.526 (Iota)', \n",
    "#                   'B.1.617': 'B.1.617.1/2 (Delta/Kappa)', \n",
    "#                   'C.37': 'C.37 (Lambda)', 'P.1': 'P.1 (Gamma)', 'P.3': 'P.3', \n",
    "#                   'unassigned': 'basal'}\n",
    "    \n",
    "    old_to_new = {'unassigned': 'basal'}\n",
    "    if old_name in old_to_new.keys():\n",
    "        new_name = old_to_new[old_name]\n",
    "    else:\n",
    "        new_name = old_name\n",
    "    \n",
    "    return new_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "88e41034",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_linear_reg_stats(mut_location, nonsyn_syn, x_axis):\n",
    "    \n",
    "    df_subset = muts_information_df.copy()\n",
    "    df_subset = df_subset[(df_subset['mut_location']==mut_location)&(df_subset['nonsyn_syn']==nonsyn_syn)]\n",
    "    \n",
    "    if x_axis == 'logistic_growth':\n",
    "        df_subset = df_subset.dropna()\n",
    "    \n",
    "    \n",
    "    slope, intercept, r_value, p_value, std_err = stats.linregress(df_subset[x_axis], df_subset['muts_per_codon'])\n",
    "    slope = round(slope, 4)\n",
    "    r_value = round(r_value, 2)\n",
    "    \n",
    "    return slope, r_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "11053433",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_growth_randomizations(iterations, save_filename):\n",
    "\n",
    "    genes = ['S1','RdRp']\n",
    "\n",
    "    growth_stats_dict = {'growth_rate_stats':[]}\n",
    "\n",
    "    for gene in genes:\n",
    "        if gene == 'RdRp':\n",
    "            nonsyn_syn_options=['nonsyn']\n",
    "        elif gene == 'S1':\n",
    "            nonsyn_syn_options=['nonsyn', 'syn']\n",
    "\n",
    "        for nonsyn_syn in nonsyn_syn_options:\n",
    "\n",
    "            for x in range(iterations):\n",
    "                #Download tree json\n",
    "                tree_json = requests.get(tree_url).json()\n",
    "\n",
    "                #Put tree in Bio.Phylo format\n",
    "                tree = json_to_tree(tree_json)\n",
    "                \n",
    "                # randomize mutation positions on tree\n",
    "                if gene == 'RdRp':\n",
    "                    tree = randomize_mutations_on_tree_multinomial(tree, total_rdrp_muts, branch_lengths_all, branch_names_all)\n",
    "\n",
    "                elif gene == 'S1':\n",
    "                    if nonsyn_syn == 'nonsyn':\n",
    "                        tree = randomize_mutations_on_tree_multinomial(tree, total_s1_nonsyn, branch_lengths_all, branch_names_all)\n",
    "                    elif nonsyn_syn == 'syn':\n",
    "                        tree = randomize_mutations_on_tree_multinomial(tree, total_s1_syn, branch_lengths_all, branch_names_all)\n",
    "                \n",
    "                \n",
    "                # keep track of logistic growth rate for this iteration\n",
    "                muts_information = []\n",
    "                # count accumulation of mutations on path\n",
    "                for node in tree.find_clades(terminal=False):\n",
    "                    \n",
    "                    # only care if it has a logistic growth rate\n",
    "                    if \"logistic_growth\" in node.node_attrs:\n",
    "                        logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "\n",
    "                        #Find all parents of the node\n",
    "                        parents = get_parent(tree, node)\n",
    "\n",
    "                        #Find mutations that occur in the parents, and at node\n",
    "                        parents_random_muts = 0\n",
    "\n",
    "                        for parent in parents:\n",
    "                            if hasattr(parent, \"random_muts\"):\n",
    "                                parents_random_muts+=parent.random_muts\n",
    "                                \n",
    "                        muts_information.append({'muts_per_codon': parents_random_muts/gene_lengths_aa[gene],\n",
    "                                                 'logistic_growth': logistic_growth})\n",
    "                        \n",
    "\n",
    "\n",
    "#                         node.node_attrs[\"random_mut_total\"] = parents_random_muts\n",
    "\n",
    "\n",
    "\n",
    "#                 for node in tree.find_clades(terminal=False):\n",
    "\n",
    "\n",
    "#                     logistic_growth = None\n",
    "#                     # only care if it has a logistic growth rate\n",
    "#                     if \"logistic_growth\" in node.node_attrs:\n",
    "#                         logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "\n",
    "#                         # add information from this iteration of gene, nonsyn_syn randomization\n",
    "#                         # for these stats, only need the growth rate and number of muts\n",
    "#                         muts_information.append({'muts_per_codon': node.node_attrs[\"random_mut_total\"]/gene_lengths_aa[gene],\n",
    "#                                                  'logistic_growth': logistic_growth})\n",
    "    \n",
    "                \n",
    "\n",
    "                muts_information_df = pd.DataFrame(muts_information)\n",
    "                slope, intercept, r_value, p_value, std_err = stats.linregress(muts_information_df['logistic_growth'], \n",
    "                                                                               muts_information_df['muts_per_codon'])\n",
    "\n",
    "                growth_stats_dict['growth_rate_stats'].append({'data': 'randomized', 'iteration':x,\n",
    "                                                               'gene': gene, 'nonsyn_syn': nonsyn_syn, \n",
    "                                                               'r_value':r_value})\n",
    "\n",
    "    with open(save_filename, 'w') as outfile:\n",
    "        json.dump(growth_stats_dict, outfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "d2084dd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "run_growth_randomizations(300, 'growth_rate_stats/growth_stats_300its.json')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "id": "af1218b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_fig1c(data_files, filename=None):\n",
    "\n",
    "    # dataframe for all wait time information, \n",
    "    # each row is an iteration for expected s1 or rdrp, syn or nonsyn\n",
    "    to_plot = pd.DataFrame(columns = ['data', 'iteration', 'gene', 'nonsyn_syn', 'r_value'])\n",
    "    \n",
    "    for data_file in data_files:\n",
    "        with open(data_file) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            df = pd.DataFrame(data['growth_rate_stats'])\n",
    "            to_plot = pd.concat([to_plot, df])\n",
    "            \n",
    "    num_iterations = len(to_plot)/3\n",
    "    \n",
    "    with open('growth_rate_stats/growth_stats_observed.json') as observed_json:\n",
    "        data_observed = json.load(observed_json)\n",
    "        to_plot_observed = pd.DataFrame(data_observed['growth_rate_stats'])\n",
    "\n",
    "    \n",
    "    fig, (ax1, ax2, ax3) = plt.subplots(1,3, figsize=(12,1.25), sharey=True)\n",
    "    sns.set_style(\"white\")\n",
    "\n",
    "    s1_expected = to_plot[(to_plot['gene']=='S1')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax1 = sns.histplot(data=s1_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax1)\n",
    "    s1_observed_r = to_plot_observed[(to_plot_observed['gene']=='S1'\n",
    "                                     )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "\n",
    "    ax1.axvline(x=s1_observed_r, color = \"#4770eb\", linewidth=4)\n",
    "    s1_pvalue= len(s1_expected[s1_expected['r_value']>s1_observed_r])/num_iterations\n",
    "    if s1_pvalue < 0.005:\n",
    "        s1_pvalue = '< 0.005'\n",
    "        ax1.annotate(f'p-value: {s1_pvalue}', xy=(0.38, 0.65), \n",
    "                     xycoords='axes fraction', size=13, fontweight='bold')\n",
    "    else:\n",
    "        ax1.annotate(f'p-value: {round(s1_pvalue, 2)}', xy=(0.38, 0.65), \n",
    "                     xycoords='axes fraction', size=13, fontweight='bold')\n",
    "    \n",
    "    \n",
    "    s1_syn_expected = to_plot[(to_plot['gene']=='S1')&(to_plot['nonsyn_syn']=='syn')]\n",
    "    ax2 = sns.histplot(data=s1_syn_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax2)\n",
    "    s1_syn_observed_r = to_plot_observed[(to_plot_observed['gene']=='S1'\n",
    "                                   )&(to_plot_observed['nonsyn_syn']=='syn')]['r_value'].values[0]\n",
    "    ax2.axvline(x=s1_syn_observed_r, color = \"#5094e3\", linewidth=4)\n",
    "    s1_syn_pvalue= len(s1_syn_expected[s1_syn_expected['r_value']>s1_syn_observed_r])/num_iterations\n",
    "    ax2.annotate(f'p-value: {round(s1_syn_pvalue, 2)}', xy=(0.3, 0.65), \n",
    "                 xycoords='axes fraction', size=13, fontweight='bold')\n",
    "\n",
    "    \n",
    "    rdrp_expected = to_plot[(to_plot['gene']=='RdRp')&(to_plot['nonsyn_syn']=='nonsyn')]\n",
    "    ax3 = sns.histplot(data=rdrp_expected, \n",
    "                       x='r_value', color = \"#ABABAB\",\n",
    "                      element = 'step',alpha=0.25,\n",
    "                      bins=10, stat='count', ax=ax3)\n",
    "    rdrp_observed_r = to_plot_observed[(to_plot_observed['gene']=='RdRp'\n",
    "                                   )&(to_plot_observed['nonsyn_syn']=='nonsyn')]['r_value'].values[0]\n",
    "    ax3.axvline(x=rdrp_observed_r, color = \"#ff8c3a\", linewidth=4)\n",
    "    \n",
    "    rdrp_pvalue= len(rdrp_expected[rdrp_expected['r_value']>rdrp_observed_r])/num_iterations\n",
    "    ax3.annotate(f'p-value: {round(rdrp_pvalue, 2)}', xy=(0.7, 0.65), \n",
    "                 xycoords='axes fraction', size=13, fontweight='bold')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "    ax1.set_title(f'S1 nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "    ax2.set_title(f'S1 synonymous', fontsize = 12, fontweight='bold')\n",
    "    ax3.set_title(f'RdRp nonsynonymous', fontsize = 12, fontweight='bold')\n",
    "    \n",
    "    ax1.set_xlabel('$R^2$', fontsize=14)\n",
    "    ax2.set_xlabel('$R^2$', fontsize=14)\n",
    "    ax3.set_xlabel('$R^2$', fontsize=14)\n",
    "    \n",
    "    for ax in [ax1, ax2, ax3]:\n",
    "        ax.set_ylabel('')\n",
    "        plt.setp(ax.get_yticklabels(), visible=False)\n",
    "        ax.tick_params(axis='x', which='major', labelsize=12)\n",
    "    \n",
    "    \n",
    "    sns.despine(left=False, bottom=False)\n",
    "    plt.subplots_adjust(wspace=0.1)\n",
    "    \n",
    "    plt.figtext(0.04, 0.94, 'C', fontsize=20, fontweight='bold')\n",
    "\n",
    "    if filename:\n",
    "        fig.savefig(filename, dpi=300, bbox_inches='tight')\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "id": "ec2b79ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAwUAAACOCAYAAABg+qT/AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/d3fzzAAAACXBIWXMAAAsTAAALEwEAmpwYAAAs/0lEQVR4nO3deVxUdfv/8RfCuO+GopJaLmOiIagh7porKhBu5NJtbrmbWu7pNxVNMr1VKs0ly0otw3DBtcUtb3fLLRe+aq6QpJniCvP7wx/nCwIy3rHMMO/n4+FDZjjLdQ5zfc65zvmczzhZLBYLIiIiIiLisHJldwAiIiIiIpK9VBSIiIiIiDg4FQUiIiIiIg5ORYGIiIiIiINTUSAiIiIi4uBUFIiIiIiIODiXp5l4//79LF++nAMHDnDt2jXy5s1L5cqVadu2LZ07dyZ37tyZFaeIiIiIiGQSJ2u+pyAhIYFp06axbNmyNKepVq0aS5YsoVixYhkaoIiIiIiIZC6rioJZs2axYMECAIoWLUr37t15/vnnOXfuHMuWLeP69esANGrUiIULF2ZuxCIiIiIikqHSLQrOnTtHmzZtSEhIoFixYqxatQp3d3fj9xcuXKBjx47cuHEDgPDwcDw8PDI1aBERERERyTjpPlOwcuVKEhISAOjbt2+yggDg2WefpV+/fpw5cwYfHx9Kly6dOZGKiIiIiEimSLco2Ldvn/Fzw4YNU52md+/eGReRiIiIiIhkqXSHJL1w4YLx87PPPpupwYiIiIiISNZLtyi4deuW8XOePHkyNRjJHps3b8bf358aNWpQt25dBg4cyPnz55NNExsbi6+vLwMHDsymKEUkLcphkX/u4sWLmM1m41/VqlXx8vKiW7dunDx5MtV5li5ditlsZs+ePf/V/CK2JN2iIG/evMbPcXFxmRqMZL3z588zbNgwXF1dmTNnDiNGjGD37t28+eabxjS7d+8mODiYP//8M/sCFZFUKYdFMlbLli359NNPWbx4MW+99RYHDx5kypQpWTa/SHZJtygoU6aM8fPFixdTnea3337jwIEDWDG6qdiYa9eukZCQQL58+ShfvjydO3dmzpw5dOnSBYvFwqVLl3j99dfx9fV94nL27NmD2Wxm1qxZtG/fHi8vL4YMGcLdu3cBuHLlCoMHD8bb25v69eszfvx4Y8SqefPmYTabWbp0KQ0bNsTHx4d58+YB8PfffzNkyBDq1KmDp6cnwcHBnD59mujoaDw8PHj77beNGLp06UKjRo2Ij4+nWbNm9OrVi5EjR1KzZk3atm3LoUOHgEffu/HRRx/RtGlTvLy8CA4O5sCBA8D/XSl69913CQ4OpmbNmvTs2ZPY2FgiIyMxm8188803Rmwvvvgiw4cPN7Z/5syZ+Pv74+Pjw4wZMwgLC6N+/fo0aNCADRs2GLGuWLGCVq1a4enpSWBgIN9//73xO7PZnOxqbtLXkZGRtGrViho1atCgQQPmzp37X/3dJefIqBy+cuUKr7/+Ot7e3nh5edG7d2+io6P55ZdfMJvN/Pvf/wYe5U+jRo3o3Lkz8OjzOXLkSPr27YunpyedOnXi7NmzANy7d4/p06dTv359vL296dWrF6dPnwae3GZ88sknmM1mdu/eDcDZs2eN/AoPD8dsNvPhhx/SsmVL6tevz6JFi5g8eTI+Pj40a9bMeBbOmlwPCQlJ9fWyZcto2rQp1atXp1mzZnz11VcZ+FcTW1ayZElq1aqFt7c33t7emEwmHj58CDzqPTFy5Ehq165NYGAgJ06ceKr5E493ixYtwtfXlwYNGvDFF1+kGkePHj1o3749ISEheHt706xZs2THivSOI2nlZVrHkUmTJmE2m427jDt27MBsNrN8+XIj7+bPn0/z5s2pXbs277zzjnHed+rUKXr27EnNmjVp3LgxM2bMMI7/Y8aMoXbt2syfPx8fHx8aNGjAypUrAWjbti1NmzY1lvPBBx9gNpuJioqiR48etG3blqlTp+Lj40NgYCCbNm2ia9eueHp60rdvX+NitTXnGIl/q6Sv0zrHcFTpFgV16tQxft65c2eq0yxcuJCuXbvSqFEjfv7554yLTjJdzZo1adGiBVu2bMHPz884eX3ppZdwcnKiaNGibNiwgcmTJ1u1vPDwcPr06UP9+vXZvHkzW7ZswWKx8MYbb7Bnzx7Gjh3LoEGDiIyMZMSIEcnm3bRpE2PHjqV48eKEhYURHR3N6tWr2bx5MyNHjiQkJIT79++zbt06SpUqxcsvv8yWLVu4desWv//+O4cPH6Zjx444OzsDsGvXLsqUKcNbb71FVFSUUWgsXbqUOXPm0LhxY0JDQ4mPj6d3796cO3fOiOW7776jQ4cO+Pv7s3v3br755htatGiBq6srq1evBmDjxo3cu3ePLl26GPOtXbuWfv368eyzz7JkyRL27NnD+PHjuXv3Lh988IGxnYmN7wcffEDJkiUZPHgw+/fvT3f//s///A/PPPMM8+bNo3nz5uzYsSNFNxFxLBmVw0uXLmXfvn2EhIQwbtw4YmJi2LJlC56enlSvXp2IiAgsFgt79uwhOjo62ed+w4YN+Pr60rt3b3799VeWLFkCQGhoKEuXLqVLly5MnTrVKFASD9aQepvRqVMn8uTJQ3h4OAARERE4OTkZhQg8yqM333yTvHnz8v7773Pt2jXGjh3LlStXCAsLM7YpvVxPzd9//83UqVOpUaMGc+fOxcvLi3Xr1nHz5s2n+dOInfriiy948cUXqVmzJq+88gqVK1dm0qRJwKMTynXr1tGtWzf69OmT6nnRk+ZPtHXrViZPnszzzz/PlClTOHjwYKqxnDp1iri4OCZOnEhsbCyhoaGAdceRtPIyreNI9+7dAYxj3Jo1a8ifPz/t27c3lrlmzRqGDx9OlSpV+Prrrzl06BC3b9+mV69enDt3jqlTp9KtWzeWLl1qFNjwKKd++eUXJk2aZHwhbnx8PF27duXy5cv85z//wWKxsG7dOurUqUPFihUBOHPmDPHx8QwYMIATJ04wYsQI/Pz88PPzY/v27WzevNnqc4zUpHWO4ajSLQqSNsJLlizhypUryX5/+PBhNm7cCMCff/5JhQoVMjZCyVTOzs6EhYXx1Vdf8cYbb+Du7s7q1at59dVXuXHjBgUKFOC5556zennBwcEEBAQQHBwMPPpM/Pbbb5w8eZJXX32VTp060bVrVwICAti1axfXrl0z5k1M9hYtWgBw/fp1vL29yZcvHx9//DHbtm2jW7duvPHGGwB069aNO3fusHHjRiIiIsiVKxedOnUylufu7s7IkSPp3r07zzzzjNF1IiIigpIlSzJp0iRatGjBO++8w507d9i8ebMxb+vWrenUqROvv/66sR0mk4kuXbpw4MABfv/9d9asWUOFChWoW7euMV+XLl1o164dDRo0AGDYsGH4+flRpUoV/vrrL2P9zs7OzJgxg+bNmzN9+nQSEhJYu3Ztiv35+N23evXqcfjwYRYvXkyRIkWYNGkS5cuXt/rvIzlPRuWwj48PCQkJzJ49m19++YUBAwYYJ/7dunUzDtwREREUKlQIPz8/Y97atWvTq1cv+vfvD5As17y8vBg6dCh+fn4MGzaMP/74I9nFo9TajGLFiuHn52cU/WvXrqVevXqUK1fOmK937974+fnh7e0NwNixYwkMDKRkyZLJci29XE+UNNcKFiyIp6cn27Zt48svv6RixYpMnTqVwoULW/dHEbvm5+fHJ598QuXKlSlUqBBvvfUWL7zwAgB79+6lbNmyDB8+nHbt2iU75lgzf6Lhw4fTokULRo0aBWDcFXucyWTi3XffJTAwkBdeeCFZbqV3HEkrL9M6jlSuXJmXXnqJiIgIbt26xdatW2nXrh0FCxY0ltmvXz/atm1LQECAsczdu3fzxx9/MGDAANq1a0e/fv1o0KABa9euNYa0B3jnnXfw8/PD19eXu3fvEhcXR0BAAAUKFGD16tXs3buXy5cvJ7vgYDKZeOedd+jRo4exTd27d8ff3x+Av/76y+pzjERJc/1J5xiOKN2ioGrVqsYfIzY2lqCgID766CPWr1/PrFmzeP31143bYl27dk3W3UhsX2RkJP3796dEiRKMGDGCFStWMHjwYG7cuMHRo0efenmJjYfJZAIeJZ+Tk1OK6RKTMunvUpu3evXqrFu3jiFDhpAnTx5CQ0ONRtjHx4fKlSuzevVq1q5dS+PGjZN9T0bShix37tzGOnPlypVsvdbGAo+KZBcXFz766CP27duXrPECyJ8/P4Bxt6JAgQLG66TrT7q+1NYfHx8PpHyOJ/HbxV966SUOHjxIhw4diIyMRBxXRuVws2bNWLt2LT179uTBgweMGzfO6LbWtm1bihYtyvLly9m8eTMBAQHky5fPmDcxX3Lnzg2Q4rP+JGnlWteuXblz5w4hISFcvHgxRa4lza3HX1uT64nvJ+banTt3jOmcnJz44osvmDVrFh4eHsbJUWIXRMnZnnnmGRo3bsxHH33Ew4cPGTRokHFB1MnJKdmJbmrHtyfNn+jBgwfJ5k8rV/LmzYuLy6PR400m01MdR9LKyycdRxKv3IeEhBAXF5ci7x7P17T2QWrH/tRyvWDBggQEBLB582aWL19O0aJFadWqVbLtz5UrV4o8T9x+a84xEn+feK6aNNefdI7hiNJvsXnUH6xDhw7Ao6ow8WG2BQsWGCctjRs3Tta/W+xDqVKl2LZtGyNGjGDjxo1ERkayadMm8uTJQ+XKlTNkHVWqVOG5555jxYoVrFq1ihUrVrBmzRoaNWpEiRIlnjhvWFgYzZs35/LlyzRr1gw3NzeuXLliHMhfffVV9u/fz/nz51M0Xmlp2bIl0dHRTJkyha1btzJt2jTy58+frCFKS6lSpWjevDmrV6/GZDIRGBho1TofX398fDxjxoxh69atjB8/HmdnZ+PKR4kSJTh69Ch79uxJ9szAzZs38fX1JTQ0lMqVKxt3Iy5duvTUMUjOkVE5PG7cOPz9/Xnw4AHNmzenSJEixmcrT548dOjQgU2bNnH79u1kd5CfpGXLlhw6dIiwsDA2bNjAnDlzKFmyJPXr10933hdffJEaNWoQHh6Oq6srzZo1s3pbkq4/rVwvWrQozs7O7N27l/3797NgwQJjvnPnzlG7dm0+++wzXnzxRWrVqkV8fHyKEzvJ2cqVK8ebb77J7du3mTZtGvDo+5quXLnCzJkzWbt2rdE33tr5E82ePZstW7Ywe/ZsnJycqFev3lPFlt5xJC3pHUdatGhByZIlCQ8Px8PDg+rVq6cbi4+PD0WLFmXBggWsX7+ehQsXsmvXLgICAqy6MJB413/Dhg0EBQUZRYy10jvHKF68OADr1q1j165dyS6kpXeO4WjS/fIyABcXF6ZNm0bbtm1Zvnw5hw8f5vr16xQoUIBq1arxyiuv4O/vn2q1JratVq1a/Pvf/2bBggWMHTsWi8VC1apVWbBgAaVKlcqQdeTKlYuFCxcyffp0QkJCyJcvH23btrWqiOzTpw8xMTEsX76c27dv8/zzzzN37lzjqoG/vz8hISG4urrSqFEjq+Lp168fDx8+5NtvvyU8PJyqVauyZMkSypUrl+bD9EkFBgayceNGWrZsaTQ2T8Pf35+bN2+ybNkyfvjhB55//nk+/PBDoxvEqFGjmDt3LoMHD+Zf//qXcfetcOHCzJw5k9mzZzNq1CgKFChAp06djDt54pgyKoffeust4uLiCAsL4+HDh5jNZiZMmGD8PjAwkMWLF+Pl5YXZbLZqmePHjydv3rwsX76cO3fu4OXlxbhx4yhSpIhV8wcGBnLkyBGCgoKSXZm01pNyHR514fjss88YOnQow4YNM/oSV6hQgalTpzJ//nyGDx9OkSJF6Nu3L61bt37qGMS+9ejRg8jISDZv3syOHTvo378/MTExrFixgjJlyhAQEGD01bdm/kSenp68++67ODk58c477+Dp6flUcaV3HElLescRFxcX2rVrx5IlS6y+0FawYEE+/fRTZsyYYeT366+/zrBhw6yav1KlSnh4eHDs2LH/6ip9eucY7du354cffuDrr7/mwIEDtG/fnoULFwLpn2M4GieLhgwSO3Xw4EF+/PFHPvnkE958800GDBiQqeu7e/cuhw4d4tNPP2Xbtm189dVX1KpVK1PXKWILdu/ezZo1awgPD+f9999P92rkP3Xjxg2OHz/OrFmz+O2339i4cSPu7u6Zuk6RrDBv3jzCwsL47rvvUjxnkN2io6M5efIk7777Ln///Tc//vij0V0ns5w5c4Zjx44xfvx46tSpw6effpqp65Mns+pOgYgt+uCDDzh69CgtW7akZ8+emb6++Ph4Bg4ciMlkYtCgQSoIxGGMGzeOv/76y3iQPrPFxsbSv39/ChcuzIQJE1QQiGSBU6dOMWjQIFxdXZk2bVqmFwTwqEvPokWLqFKlChMnTsz09cmT6U6BiIiIiIiDs+pBYxERERERyblUFIiIiIiIODg9U5ABevfuzeLFi7M7DJF/pPXw5F/ysnH2M9kUScbI6Xn5yofJR8paPUj97sX25fS8lEzwwWNDGI/clT1xOADdKcgA169fz+4QROQxyksR26O8FLFdKgpERERERBycug+JiEiWi4iI4MGDBxmyLJPJREBAQIYsS0TEUakokBxPJx8itufBgwdWfwt5erZv354hyxERHTMdmYoCyfF08iEiImIdHTMdl54pEBERERFxcLpTIDYpI29furjoYy4iIiLyJDpbEpuUkbcvRUREROTJVBSIiIhVdAdPRCTnUqssIiJW0R08EZGcSw8ai4iIiIg4OBUFIiIiIiIOTkWBiIiIiIiD0zMFIiIiInZMgwBIRtBfXhxGZGQk06dPZ+DAgbz66qvZEsP169cpVqxYivfXrVvHvHnziI2NpV69ekyePJmiRYumuoxPP/2Uzz//nLi4OFq1asW4cePImzcvALt372batGlcunSJmjVrMmXKFMqWLQvArl276NWrV7Jl9erVi9GjR2fsRopDycq8io+PZ+HChWzYsAGAtm3b0rt371SnPXXqFCEhIfz6668UKVKEvn370q1bNwDu3bvHe++9x08//cRff/2Fl5cXEyZM4LnnnsvU+EUyiwYByDzh4eGMHTuWUaNGpdneZJT4+Hhmz57Nd999B0CHDh0YOnQozs7Oac5z5MgROnXqRNeuXZk4caLxfnBwMIcOHUo27cGDBylQoECay1JRIJLJ7t+/z9q1a1m6dCktW7ZkyJAhyX5//Phx3n77bSpVqoSfnx9ff/01AHPnzk2xrK1bt/Lee+/h7e1NmTJlWLlyJQUKFGD06NHExMQwYMAAihcvTlBQECtXrmTIkCGEh4cDcOLECQB69uxJkSJFAPDy8srMTRfJUOHh4Xz55Zf4+vpisVhYtmwZJUqUwNXVNdl09+7do2/fvty8eZPg4GB++eUXJk+ezDPPPEOrVq345JNP+Oqrr2jWrBmlSpXi66+/Zvjw4caBWEQkO3zxxRcsXLiQJk2akJCQwPz583F1daV79+6pTr97925GjhyJxWJJ9r7FYuHkyZNUrVqVVq1aGe/nzp37ievXMwWSrS5evIjZbGbChAmMGjWKmjVr8sorr3DhwoVUp//8889p2LAhGzduNN4bPXo0jRo14tKlS9y8eZPx48fTpk0bXn75Zfr06cPx48dTXdaQIUNo2LAhN27cAGD58uU0bNiQyMhIAKKjoxk9ejStWrWiQ4cOfPPNN8a8e/bswWw2M3ny5DS3LTY2lrCwMJo0acK4ceO4evUqFStWTDHd+vXrSUhIYPz48UyePJlatWrx/fffc/v27RTTrlu3DoDQ0FA++OADypQpw5o1a4BHBcOdO3cYNGgQEyZMoE2bNhw7doyoqCjgUVHg7OzMG2+8Qbdu3Rg4cCC+vr5pxi/2K628+uWXX1Kd/uOPP8ZsNic7Ke7fvz9Vq1bl999/58aNGwwePJgpU6Zkal4dOnSIhg0bMnv27FSXvWXLFkwmE1OnTmXq1KmYTCY2b96cYrqoqCiuXr1K69atGT16NIsWLcLJyckouLt168a3337LjBkzaNq0KSaTCZPJlO5+FRHbkFltXJ06dXjxxRcJCgri119/TXVZPXr0wGw28+effwKwePFizGazcQHu8uXL9O/fH29vbxo3bsxnn31mzJveucO6deswmUzMmzePefPmYTKZjGN8atvUs2dPChcunOJ3586dIy4ujiZNmtChQwfeeOMNBg4cmG47p6JAbEJ4eDixsbEEBQVx+vRpli1blupJcZs2bXB2djZOBP766y/27t2Lp6cnZcuW5aOPPmLHjh00bdqUdu3acerUKT788MOnjic+Pp7Ro0dz4MAB2rZtS/ny5Zk7dy5HjhwBwN3dnWHDhtG0adNU5w8JCaFJkyaEhYVRsWJFQkND2blzJ35+fimmTTxpL1++vPH/w4cP+f3331Od1tnZmWeffRaAcuXKce3aNW7evMn//u//plgOwNmzZwGMk7hGjRrx0ksv0atXL65fv/7U+0bsx+N51b9/f27dupViuqCgIJydnY2Dz/Xr19m5cyd16tShXLlyvP/++2zdupXq1atnSl798MMPAJQuXZo+ffpQr169VOc/f/48rq6u5M6dmzx58uDq6ppqnri6upIrVy4OHjzIr7/+yvr167FYLJw+fRqA4sWLU716dcLDw+nXrx/58+dnxowZT709IpK9MrqNa926NR07duT48eO89957Tx1PfHw8/fv3Z/fu3XTo0IGKFSsybdo046JIeucOUVFRuLm5kTt3bvLmzYubm5txbH9ciRIleP/993n33XdT/C6xZ8DSpUuNY35ERES68av7kNiE0qVLs3DhQnLlyoWzszOff/45Bw4cwNvbm9jYWGO6smXLUqdOHfbt28e1a9fYvn07Dx8+pG3btgAMHTqUjh074urqytGjR9m0aVOy+a3122+/ERUVRevWrQkKCiI+Pp6+ffuyb98+I46BAwemOf/333/P/fv3adeuHW+99RalS5dOc9o7d+4AkCdPHuD/bu+lVhTduXMn2e2/xJ/j4uKIi4tL9l7S5Tx8+BCTyYSHhwedO3fm8OHDrFq1itDQUKZPn27dThG7k1pe7d69m7p16xITE2NMV65cORo0aMDOnTuJiYlhy5YtPHjwgA4dOgAwbtw4evTowd69eylUqFCG59XatWtp1qwZbm5u/Otf/0pz/rt37yb7/JtMJiN/knJ1dWX48OHMnj2bTp064ebmRsGCBY0cSeTp6cmkSZOYNWsWQ4YMYfXq1eneXhcR25HRbVypUqU4fPgwERER/PHHH08dz5EjRzh58iSBgYF0796dhw8f0rFjR7755hv8/PzSPXe4c+eOcS4Aj47jj7dbiTp37gw8uvvwuJs3b1K6dGnatGnDc889R1hYGOPHj8fT05MKFSqkuX4VBWITypUrR65cj25cJXaxiY2NZfv27clOWr/++mvatWvHf/7zH3744Qd+/PFHChQoQJMmTYBHt8xmzpzJpUuXqFatGgkJCcTHx1sVQ0JCgvHz1atXAdi4cWOyrkpJG5knCQkJYfHixaxfv57IyEgaNWpE586dadKkSYoHhvLlywc8evYAHvWHBsifP3+K5ebLly/ZCBOJ0+bLl++Jy3FxcUl2C7Jjx45s2bKFn3/+2artEfuUWl4lHhDHjh1rTPf999/TsWNHtm3bRmRkJBs3bqRgwYJGX9SoqCgmTZpEVFQUNWrUyPC8SryblZ48efIk+/zfv3+fvHnz4uLiwqpVq5JNW7x4cYYPH05sbCzu7u7MmjULk8mUYrrcuXNTo0YNfv75Z2bNmsULL7xAQECAVfGISPbK6Dbu/PnzeHp6PlUbl3S6y5cvA/Ddd98l66qUeJcyPXnz5jWO4fDoOJ54bH8awcHBBAcHG6/j4uKYPn06+/btU1Egtu/MmTM8ePAAk8nEuXPngEdX+ypVqpSs712xYsWoX78+xYoVY/369Zw9e5b27dsbo++MHz+ehIQEIiIicHFxeeLBPfHk/O7duwDJrnyWKFECgIYNG9KiRQvg//roWcPX1xdfX1+ioqL47LPPiIiI4KeffmLo0KEMGjQo2bSJ3XwuXLiAm5sbFy5cwNnZmXLlyqVYbvny5Tl16hSXL1+mTJkyXLhwgRIlSlCkSBFjORcvXqR27drGcxnPP/88V69eZcmSJVStWpWgoCDu3bvH/fv3KV68uFXbI/Yptbxyc3OjatWqzJkzx5iuRIkSNG3alBIlSvDtt99y+vRpOnfubByMBg0aREJCAmPGjKFJkyYZnldPOkgl5e7uzvnz57l//z4Wi4Vr165RqVKlVLsbzZ8/n4cPHzJ48GBiYmKIi4ujXr16NGrUiNmzZ7N//34WL15M3rx52bFjBwB169bl2rVrVsUiIulLrWB/Wh2f8LuMbuN27dqFyWSifv36aa4zsY1LvEuZtM1IHPSgefPmtG/fHovFQlRUVKrPE6amQoUKREVFGW1cdHQ0L7zwglXzJvXtt99y6NAhRo8eTaFChYwuValdbExKRYFkmP9mnOTEPu0xMTG0b98eNzc39u3bR6FChahVqxb58uXDzc0txXytWrVixYoVAMn66VssFm7cuEFYWBiXL1/m5s2baT5YU7ZsWQ4cOMCcOXOoUqWK8RAvQI0aNXB3d2f//v24ubkRExPDtm3beOmllwC4dOkSERER1KhRg4YNG6a5fRUrVmTy5MkMHz6cFStWpDrMqJ+fH0uXLmXy5Ml4eXmxf/9+Xn75ZQoWLMjff//NsmXLqFChAn5+fvj5+bFlyxZGjhxJmTJluHz5stHdonnz5rz//vvMnTuXo0ePsmHDBqpVq0bFihW5d+8ekZGRrFixghMnTvDbb79x586dZFcSJOeJiYmhV69emM1mVq5ciaurK76+vuTPn98YqjapgIAAlixZAmDcVodHefXnn38SGRlJZGRkhudVQEAATZs25erVq2zatImqVavi4+OTYtkvv/wy8+fPZ8KECVgsFh48eGAUF2fOnGHXrl3Url0bDw8PLl68yLZt27h37x7Hjh0DIDAwEHhUXIeHh/P222/z3HPPsWXLFipXrkyVKlVUFIhkoLSeD3oqy1IfeAAyvo2bPn06Fy9e5MaNG2m2ceXKlWP37t2EhIRQrVq1ZIMleHt7U758eX7++WfKli1rtGnBwcG0adMm3XMHPz8/Zs6cyZAhQ4w2rl27dsCj7pc//PAD9erVo2bNmk/cZbdu3eKbb77hzJkzeHl5sWLFCkqWLJnmswyJ9KCxZJjEcZKf5l/iSXatWrUwm80cOXKEihUrMnv27CfeMkt8hqBChQp4eHgY748aNYpSpUqxadMmHj58iLe3N9evX0+1b+Brr72Gh4cHe/fu5eDBg/Tr18/4nbOzM++99x7VqlVjzZo1HDp0CH9/f6MAuXjxInPmzOHHH3+0at8UK1aMAQMGpDqOu6enJ7NmzeL+/fusX7+eFi1aMGXKFODRg9Rz5swxHhDy8/Nj3LhxREdHs3PnTjp27Mjw4cMBKFOmDJ988gmFChXi22+/xdvb2xjWNE+ePCxatAhPT09WrVrF2bNnGTJkyBP7b4v98/X1pWzZsnz33XdUrlyZ+fPnP/FKUceOj67JVapUCU9PT+P9qVOnUrp0aQ4fPpwpeZU4TO+VK1dYtGhRmt3agoOD6dGjB6dOneLUqVN069aNoKAg4NH3EixatMgYMWTYsGHUrl2bDRs2cPv2bcaNG2eMtvXKK6/Qt29fLl++zJYtW6hfvz4zZszQlzaJ2JmMbuPWrFnDgwcP8PHxITY2lujo6BTLGDBgADVr1mTHjh3s2bPHOAbDozbu448/xtPTk5UrV7Jnzx66dOlidGVK79yhV69e9O/fn2PHjnHs2DH69u1rDEd6/Phx5syZw4EDB9LdL6+99hpDhgzhypUrfPHFF3h4eLBkyZJ07xQ4WR4f3FSeWlBQkDEUlSNbtWrVU395ypUrV+jcuTO+vr6EhoZmUmQZZ/v27UajktO0Hp78CunG2c9kUyQZI6fn5SsfXkz2evUgd+Pnixcv8vLLL9OkSRMWLFiQYev8b3Lc3uTkHLcFOT0vs0tOzs2Sy15J/sbIXZnWxjk63SkQEREREXFwKgpERERERBycOlBKtipdurQx8oc9yIiRFBKZTCYNfSiZwt3dnZMnT2Z3GCIimUJtXOZQUSDyFDJkJIX/b/v27Rm2LBEREZF/Qt2HREREREQcnO4UiIjkYP/N94ekRUN2iojkXGrhRURysMTvDxEREXkSdR8SEREREXFwulMgIiIiksXUtU9sjT5FIiIiIllMXfvE1qj7kIiIiIiIg1NRICIiIiLi4FQUiIiIiIg4OBUFIiIiIiIOTkWBiIiIiIiDU1EgIiIiIuLgNCSpg9M4ySIiIiKiszgHp3GSRURExF6sWrXqHy/DZDIREBCQAdHkLCoKRERERMQuZMSFzO3bt2dAJDmPnikQEREREXFwulMgIiLy/7m4uGRI94RE6qYgIvZCRYGIiI2xbgCAuslepXUiqwEAnk69evUydHnqpiAi9kJHCxERG2PNAABffnM/2WsNGCAiIv+EigIRERERK2gYb8nJ9IkUERERsYKG8ZacTKMPiYiIiIg4OBUFIiIiIiIOTkWBiIiIiIiDU1EgIiIiIuLgVBSIiIiIiDg4jT5khzQkWs6Qkd+cqm9NFRERkX9CZ4R2SEOi5QwZ+c2p+tZUERER+SdUFIiIiIiIw9Cd+tSpKBARERERh6E79anTg8YiIiIiIg5OdwpERDKABgAQsU3KTRHr6NMtIpIBNACApEZ9l7OfclPEOioKREREMon6LouIvdAzBSIiIiIiDk53CkRygIzpotAkI0IRERFxGDmpi6CKgiyiB50kM2VEF4VFuzIgEBGRDKBjptiLnNRFUJmSRfSgk4jt0YmHiG3SMVMk6+koJiKpyojbodl9KzQ9OvEQe5KTuimIiO1xslgsluwOwt75+PhQtmzZJ05z48YNtKvF0Tg5OVG0aNF0pytWrBiLFy/O0HUrL0XSZk1uZldegnJTHFN25aWxfhUFIiIiIiKOTUOSioiIiIg4OBUFIiIiIiIOTkWBiIiIiIiDU1EgIiIiIuLgVBSIiIiIiDg4FQUiIiIiIg5ORYEN+umnn2jfvj2tWrVi6NCh3Lp1K9XpIiIi8Pf3JyAggODgYI4cOZLFkf4fa2K2druymjVx2dK+Tupp9unWrVvx8vLKwuhsy9N+/rJ7f9l6O2BvOW9PeZ5T89recjCRrefi4+wtN5Oypzx9XI7IW4vYlNjYWEvdunUtZ8+etVgsFktoaKhl0qRJKaaLioqy1K9f3xIdHW2xWCyWn376ydK4ceOsCzQJa2K2druymjVx2dK+Tupp9unZs2ctzZs3t9SsWTPrArQhT/v5y+79ZevtgL3lvD3leU7Na3vLwUS2nouPs7fcTMqe8vRxOSVvVRTYmIiICEvfvn2N1xcuXLB4e3tbEhISkk134cIFy48//mi8vnbtmsXDw8Ny7969rArVYE3M1m5XVrMmLlva10lZu0/j4uIsHTt2tGzdutUmG6Gs8DSfP1vYX7beDthbzttTnufUvLa3HExk67n4OHvLzaTsKU8fl1Py1iW771Q4qm3btjFgwIAU7w8cOBA3NzfjtZubG7du3eL27dsULFjQeN/d3R13d3cALBYL06dPp1mzZuTOnTvzg3/M1atX043ZmmmygzVx2dK+TsrafTpx4kS6dOmC2WzOjjCz1D/NK8ja/WWv7YC95bw95bm957W95WAie83Fx9lbbiZlT3n6OHvP20QqCrJJ48aNOX78eIr358+fj5OTU4r3c+VK/fGPuLg4xowZw9WrV1m0aFGGx2mNhISEdGO2Zprs8DRx2cK+Tsqa2L/88ktcXFzo2LEjFy9ezMrwssU/zaus3l/22g7YW87bU57be17bWw4mstdcfJy95WZS9pSnj7P3vE2kB41tTOnSpYmJiTFeR0dHU6RIEfLnz59i2suXLxMcHIyzszOff/45hQsXzspQDdbE/DTblZWsjctW9nVS1sS+evVqjhw5QkBAAP369ePu3bsEBAQQHR2dHSFnG2v/zrayv2y9HbC3nLenPM+peW1vOZjI1nPxcfaWm0nZU54+LsfkbXb2XZKUrl27ZvH19TUeVpk5c6ZlzJgxKab7+++/Lc2aNbPMmzcviyNMyZqYrd2urGZNXLa0r5N62n164cIFm+zDmBX+m89fdu4vW28H7C3n7SnPc2pe21sOJrL1XHycveVmUvaUp4/LKXmrosAG/fTTT5b27dtbWrdubenXr5/l+vXrFovFYvn1118t/v7+FovFYpk/f76latWqFn9//2T//vzzT5uJOWm8aU1jC9KL3db2dVLW7PdEttoIZRVr8iqp7N5ftt4O2FvO21Oe59S8trccTGTruWhNvLacm0nZU54+LifkrZPFYrFk990KERERERHJPnqmQERERETEwakoEBERERFxcCoKREREREQcnIoCEREREREHp6JARERERMTBqSgQEREREXFwKgrEYVy5coUePXrg5+eHv78/mzdvzu6QRATlpogtUl46Hn1PgTiMmJgYYmNjeeGFF4iNjSUoKIiNGzeSL1++7A5NxKEpN0Vsj/LS8bhkdwAiWaVkyZKULFkSgBIlSlC4cGGuX7+uBk4kmyk3RWyP8tLxqCiQHOO1115jz549ALi4uFC2bFn69OlD586dU0x75MgRHj58SOnSpbM6TBGHo9wUsT3KS3mcnimQHOPEiROMGDGCnTt3snnzZvz8/Jg4cSLHjx9PNt3169cZPXo0ISEhODk5ZVO0Io5DuSlie5SX8jgVBZIj/P7779y8eZOGDRvi6upK2bJlCQ4OxmKxcPr0aWO6+/fvM3jwYPr164e3t3c2RiziGJSbIrZHeSmpUVEgOcLRo0cpWLAgZrMZgD/++IPQ0FBy5cqFh4cHABaLhTFjxlC3bl0CAwOzMVoRx6HcFLE9yktJjZ4pkBzh2LFjxMXFUbt2bRISErh79y4mk4nRo0dTqVIlAA4cOEBkZCRms5mtW7cCEBoaajSKIpLxlJsitkd5KanRkKSSI/Ts2ZNy5crRu3dvbt26xaxZsyhfvjwTJ07M7tBEHJpyU8T2KC8lNeo+JDnCiRMn8PLyonz58nh4eDB58mS++uorTp48md2hiTg05aaI7VFeSmpUFIjdu3DhAjdu3KBKlSrGe2XLlqVatWpERERkY2Qijk25KWJ7lJeSFhUFYveOHTtGrly5qFixYrL369WrZ/SDFJGsp9wUsT3KS0mLigKxe8eOHaNcuXLkzZs32fv169fn/PnzyYZXE5Gso9wUsT3KS0mLHjQWEREREXFwulMgIiIiIuLgVBSIiIiIiDg4FQUiIiIiIg5ORYGIiIiIiINTUSAiIiIi4uBUFIiIiIiIODgVBSIiIiIiDk5FgYiIiIiIg1NRICIiIiLi4FQUiIiIiIg4uP8H4hdm+imEYk4AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x90 with 3 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plot_fig1c(['growth_rate_stats/growth_stats_50it.json', \n",
    "            'growth_rate_stats/growth_stats_50itmore.json', \n",
    "            'growth_rate_stats/growth_stats_200its.json', \n",
    "            'growth_rate_stats/growth_stats_400its.json', \n",
    "            'growth_rate_stats/growth_stats_300its.json'], filename='plots/fig1c_r_stats.png')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9428500",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "475abc7b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "982fda9b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-kernel",
   "language": "python",
   "name": "ncov-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
