{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7bd2fb51",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/katekistler/anaconda3/envs/ncov/lib/python3.6/site-packages/Bio/Seq.py:2859: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n",
      "  BiopythonWarning,\n"
     ]
    }
   ],
   "source": [
    "from utils import sars2_genome_info, add_syn_mut_attribute, add_mut_at_node_attr, add_mut_accumulation_attr\n",
    "from utils import get_parent, prune_tree\n",
    "from augur.utils import json_to_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import matplotlib.gridspec as gridspec\n",
    "from collections import Counter\n",
    "import requests\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9237b7f7",
   "metadata": {},
   "source": [
    "## Investigate recombination as potential source of Nsp6 deletions in multiple lineages \n",
    "\n",
    "##### If the 8 occurrences of Nsp6 deletion are actually due to recombination, not convergent evolution, then the acceptor lineage should have recieved other mutations near Nsp6 along with the deletion during recombination.\n",
    "\n",
    "##### For each of the 8 occurrences of the Nsp6 deletion, find what other mutations were present in the lineage. Then, determine whether these mutations are shared by other lineages with the Nsp6 deletion.\n",
    "\n",
    "### There is no figure associated with this notebook- it is more of a sanity check. The results (no evidence of recombination) are summarized in the table at the bottom."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f22c5ef",
   "metadata": {},
   "source": [
    "Import the tree and convert to Bio Phylo format. This is a time-resolved phylogeny built from 9544 SARS-CoV-2 genomes sampled between December 2019 and May 15, 2021. The tree can be viewed at https://nextstrain.org/groups/blab/ncov/adaptive-evolution/2021-05-15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "533b6209",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_url = 'https://nextstrain-blab.s3.amazonaws.com/ncov_adaptive-evolution_2021-05-15.json'\n",
    "\n",
    "tree_json = requests.get(tree_url).json()\n",
    "\n",
    "#Put tree in Bio.Phylo format\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b5e1cd1",
   "metadata": {},
   "source": [
    "Get information about genome position and length of each gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "45bcacd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_gene_locations, reference_gene_codon, gene_lengths_aa = sars2_genome_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "25595ae6",
   "metadata": {},
   "source": [
    "Add information about synonymous mutations as an attribute of nodes on the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8449c4fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = add_syn_mut_attribute(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06474fcb",
   "metadata": {},
   "source": [
    "Add the number of mutations that occur at a node as an attribute of that node, for all nodes on the tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "53b19fce",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = add_mut_at_node_attr(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a37ea30",
   "metadata": {},
   "source": [
    "Add an attribute to each node that gives the total number of mutations accumulated between the tree root and that node (including mutations on the node). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "1548a13b",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = add_mut_accumulation_attr(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b10ad359",
   "metadata": {},
   "source": [
    "Find all occurrences of Nsp6 deletion and put them in chronological order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "cf380bff",
   "metadata": {},
   "outputs": [],
   "source": [
    "nsp6del_nodes = []\n",
    "\n",
    "for node in tree.find_clades(terminal=False):\n",
    "\n",
    "    if len(node.get_terminals()) >=15:\n",
    "\n",
    "        # finds branchs with the mutation\n",
    "        if hasattr(node, \"branch_attrs\") and \"mutations\" in node.branch_attrs:\n",
    "            if 'ORF1a' in node.branch_attrs[\"mutations\"]:\n",
    "                for m in node.branch_attrs[\"mutations\"]['ORF1a']:\n",
    "                    if '3676-' in m:\n",
    "                        # manually add emerging lineages that descend from these nodes:\n",
    "                        emerging_lineage = {'NODE_0000031': 'B.1.619/B.1.620', 'NODE_0004363': 'B.1.526 (Iota)', \n",
    "                                            'NODE_0004416': 'B.1.351 (Beta)', 'NODE_0008062': 'P.1 (Gamma)', \n",
    "                                            'NODE_0005404': 'B.1.1.7 (Alpha)', 'NODE_0000152': 'B.1.525 (Eta)', \n",
    "                                            'NODE_0005215': 'B.1.1.318', 'NODE_0006885': 'C.37 (Lambda)'}\n",
    "                        nsp6del_nodes.append({'node':node.name, 'date': node.node_attrs[\"num_date\"][\"value\"], \n",
    "                                              'emerging_lineage': emerging_lineage[node.name]})\n",
    "                        \n",
    "nsp6del_nodes = sorted(nsp6del_nodes, key = lambda i: i['date'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7c9dad6d",
   "metadata": {},
   "source": [
    "For every occurrence of a convergently-evolved mutation, find all nucleotide mutations that occurred before it and what nodes they occurred at"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "90074469",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_mut_history(gene, mutation):\n",
    "    \"\"\"\n",
    "    Find internal branches with the convergently-evolved mutation.\n",
    "    Return a dictionary with the name of this branch, the date, \n",
    "    and all mutations that occurred on the path to this branch\n",
    "    \"\"\"\n",
    "    \n",
    "    mut_history = {}\n",
    "    \n",
    "    for node in tree.find_clades(terminal=False):\n",
    "\n",
    "        if len(node.get_terminals()) >=15:\n",
    "            \n",
    "            # finds branchs with the mutation\n",
    "            if hasattr(node, \"branch_attrs\") and \"mutations\" in node.branch_attrs:\n",
    "                if gene in node.branch_attrs[\"mutations\"]:\n",
    "                    for m in node.branch_attrs[\"mutations\"][gene]:\n",
    "                        if mutation in m:\n",
    "\n",
    "                            mut_date = node.node_attrs[\"num_date\"][\"value\"]\n",
    "\n",
    "                            # find all mutations on the path leading to this mutation, including this branch\n",
    "                            muts_in_parents = {'nuc':[],**{k: [] for k,v in gene_lengths_aa.items()}}\n",
    "                            muts_in_parents = {}\n",
    "                            \n",
    "                            parents = get_parent(tree, node)\n",
    "                            for p in parents:\n",
    "                                if hasattr(p, \"branch_attrs\") and \"mutations\" in p.branch_attrs:\n",
    "                                    if 'nuc' in p.branch_attrs['mutations']:\n",
    "#                                     for k,v in p.branch_attrs['mutations'].items():\n",
    "                                        muts_in_parents[p.name]=p.branch_attrs['mutations']['nuc']\n",
    "                                        \n",
    "                            # remove entries for genes that don't have mutations\n",
    "                            muts_in_parents = {k:v for k,v in muts_in_parents.items() if len(v)!=0}\n",
    "\n",
    "\n",
    "                            mut_history[node.name] = {'branch_with_mut':node.name, \n",
    "                                                      'mut_date': mut_date, \n",
    "                                                      'mut_history': muts_in_parents}\n",
    "    return mut_history   "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7c2dd5b",
   "metadata": {},
   "source": [
    "Run this for the Nsp6 deletion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "4e5ffd43",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "mut_history = get_mut_history('ORF1a', '3675-')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e68ab8a8",
   "metadata": {},
   "source": [
    "Now consider all possible pairs of nodes with the nsp6 deletion as potential donors/acceptors of recombination"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "c9a040cd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_informative_muts(donor_dict, acceptor_dict, donor_node, acceptor_node, on_deletion_branch):\n",
    "    \"\"\"\n",
    "    Given two dictionaries of nodes with mutations, remove \n",
    "    common mutations (that occurred at the same node) \n",
    "    from those dictionaries. Also remove the nsp6 deletion \n",
    "    itself from the list of nucleotide mutations\n",
    "    \"\"\"\n",
    "    # list of nucleotide muts that make up deletion\n",
    "    nsp6del_nucs = ['T11288-', 'C11289-', 'T11290-', 'G11291-', 'G11292-', \n",
    "                    'T11293-', 'T11294-', 'T11295-', 'T11296-']\n",
    "    \n",
    "    # remove shared ancestral mutations\n",
    "    donor_informative_muts = {k:v for k,v in donor_dict.items() if k not in acceptor_dict}\n",
    "    acceptor_informative_muts = {k:v for k,v in acceptor_dict.items() if k not in donor_dict}\n",
    "    \n",
    "    # if on_deletion_branch is True, then consider nuc muts that happened on same branch as deletion as informative, \n",
    "    # otherwise only look at nuc muts that definitely occurred before\n",
    "    if on_deletion_branch == True:\n",
    "        # remove the nsp6 deletion\n",
    "        donor_informative_muts[donor_node] = [x for x in donor_informative_muts[donor['node']] \n",
    "                                                 if x not in nsp6del_nucs]\n",
    "        acceptor_informative_muts[acceptor_node] = [x for x in acceptor_informative_muts[acceptor['node']] \n",
    "                                                 if x not in nsp6del_nucs]\n",
    "    elif on_deletion_branch == False:\n",
    "        del donor_informative_muts[donor_node]\n",
    "        del acceptor_informative_muts[acceptor_node]\n",
    "    \n",
    "    return donor_informative_muts, acceptor_informative_muts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "id": "eba7c24e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_closest_muts(all_muts):\n",
    "    \"\"\"\n",
    "    Given a list of informative nucleotide mutations, \n",
    "    find the mutation that is closest to the deletion on \n",
    "    the either side, or return 1 (start of genome) \n",
    "    if there are no informative mutations to the left \n",
    "    or 29903 (end of genome) if no informative muts to right\n",
    "    \"\"\"\n",
    "    # find position of mutation\n",
    "    all_muts_pos = [int(x[1:-1]) for x in all_muts]\n",
    "    \n",
    "    # find only muts to left\n",
    "    muts_left = [x for x in all_muts_pos if x<11288]\n",
    "    if len(muts_left)==0:\n",
    "        closest_left = 1\n",
    "    else:\n",
    "        closest_left = min(muts_left, key=lambda x:(11288-x))\n",
    "    \n",
    "    # find all muts to the right\n",
    "    muts_right = [x for x in all_muts_pos if x>11296]\n",
    "    if len(muts_right)==0:\n",
    "        closest_right = 29903\n",
    "    else:\n",
    "        closest_right = min(muts_right, key=lambda x:(x-11296))\n",
    "    \n",
    "    return closest_left, closest_right"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a8ea6ec",
   "metadata": {},
   "source": [
    "Make a dataframe summarizing whether there is evidence of recombination (shared mutations) in the donor/acceptor pairs, and what the possible window of the genome where recombination could have occurred is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "7652f96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "on_deletion_branch =True\n",
    "\n",
    "# initialize list to keep track of info about possibility of recombination\n",
    "possible_recomb_info = []\n",
    "\n",
    "# potential recombination donor/acceptor pair\n",
    "for donor in nsp6del_nodes:\n",
    "    for acceptor in nsp6del_nodes:\n",
    "        if donor!= acceptor:\n",
    "            # donor had to have existed first\n",
    "            if donor['date']< acceptor['date']:\n",
    "                # find informative mutations in the donor and acceptor\n",
    "                # (mutations that didn't occur in a common ancestor of the donor and acceptor)\n",
    "                donor_muts = mut_history[donor['node']]['mut_history']\n",
    "                acceptor_muts = mut_history[acceptor['node']]['mut_history']\n",
    "                \n",
    "                donor_informative_muts, acceptor_informative_muts= find_informative_muts(donor_muts, acceptor_muts, \n",
    "                                                                                         donor['node'], acceptor['node'], \n",
    "                                                                                         on_deletion_branch)\n",
    "                \n",
    "                # find informative mutations near the nsp6del\n",
    "                # closest mutations to the deletion in the donor\n",
    "                all_donor_muts = [x for y in donor_informative_muts.values() for x in y]\n",
    "                closest_left_donor, closest_right_donor = find_closest_muts(all_donor_muts)\n",
    "                # closest mutations to the deletion in the acceptor\n",
    "                all_acceptor_muts = [x for y in acceptor_informative_muts.values() for x in y]\n",
    "                closest_left_acceptor, closest_right_acceptor = find_closest_muts(all_acceptor_muts)\n",
    "\n",
    "                # if closest mutation on either side is shared by both donor and acceptor, \n",
    "                # this would support the hypothesis of the different nsp6 deletion occurrences \n",
    "                # being a result of recombination\n",
    "                support_for_recomb_left = False\n",
    "                support_for_recomb_right = False\n",
    "                # just looking to see whether mutation occurs at same site. \n",
    "                # If it did, would need to check identity of mutation\n",
    "                if closest_left_donor == closest_left_acceptor:\n",
    "                    # don't consider lack of informative mutations as evidence\n",
    "                    if closest_left_donor!= 1:\n",
    "                        support_for_recomb_left = True\n",
    "                if closest_right_donor == closest_right_acceptor:\n",
    "                    if closest_right_donor != 29903:\n",
    "                        support_for_recomb_right = True\n",
    "\n",
    "                possible_window_of_recomb = [max(closest_left_donor, closest_left_acceptor), \n",
    "                                             min(closest_right_donor, closest_right_acceptor)]\n",
    "                    \n",
    "                    \n",
    "                    \n",
    "                possible_recomb_info.append({'donor': donor['node'], 'acceptor': acceptor['node'], \n",
    "                                             'emerging_lineage_donor': donor['emerging_lineage'], \n",
    "                                             'emerging_lineage_acceptor': acceptor['emerging_lineage'], \n",
    "                                             'support_for_recomb_left': support_for_recomb_left, \n",
    "                                             'support_for_recomb_right':support_for_recomb_right, \n",
    "                                             'possible_window_of_recomb': possible_window_of_recomb, \n",
    "                                             'on_deletion_branch': on_deletion_branch}) \n",
    "\n",
    "possible_recomb_df = pd.DataFrame(possible_recomb_info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "5f938ed1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>donor</th>\n",
       "      <th>acceptor</th>\n",
       "      <th>emerging_lineage_donor</th>\n",
       "      <th>emerging_lineage_acceptor</th>\n",
       "      <th>support_for_recomb_left</th>\n",
       "      <th>support_for_recomb_right</th>\n",
       "      <th>possible_window_of_recomb</th>\n",
       "      <th>on_deletion_branch</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9867, 15324]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 15324]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[6613, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[6954, 14676]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[8593, 14407]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9891, 15324]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>NODE_0000031</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.619/B.1.620</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 20262]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9867, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9867, 14676]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9867, 14407]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9891, 20262]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>NODE_0004363</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.526 (Iota)</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 14676]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 14407]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 20578]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NODE_0004416</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.351 (Beta)</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10323, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[6954, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[8593, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9891, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NODE_0008062</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>P.1 (Gamma)</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 12778]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[8593, 14407]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9891, 14676]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NODE_0005404</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.1.7 (Alpha)</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[9891, 14407]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NODE_0000152</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.525 (Eta)</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NODE_0005215</td>\n",
       "      <td>NODE_0006885</td>\n",
       "      <td>B.1.1.318</td>\n",
       "      <td>C.37 (Lambda)</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>[10097, 13536]</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           donor      acceptor emerging_lineage_donor  \\\n",
       "0   NODE_0000031  NODE_0004363        B.1.619/B.1.620   \n",
       "1   NODE_0000031  NODE_0004416        B.1.619/B.1.620   \n",
       "2   NODE_0000031  NODE_0008062        B.1.619/B.1.620   \n",
       "3   NODE_0000031  NODE_0005404        B.1.619/B.1.620   \n",
       "4   NODE_0000031  NODE_0000152        B.1.619/B.1.620   \n",
       "5   NODE_0000031  NODE_0005215        B.1.619/B.1.620   \n",
       "6   NODE_0000031  NODE_0006885        B.1.619/B.1.620   \n",
       "7   NODE_0004363  NODE_0004416         B.1.526 (Iota)   \n",
       "8   NODE_0004363  NODE_0008062         B.1.526 (Iota)   \n",
       "9   NODE_0004363  NODE_0005404         B.1.526 (Iota)   \n",
       "10  NODE_0004363  NODE_0000152         B.1.526 (Iota)   \n",
       "11  NODE_0004363  NODE_0005215         B.1.526 (Iota)   \n",
       "12  NODE_0004363  NODE_0006885         B.1.526 (Iota)   \n",
       "13  NODE_0004416  NODE_0008062         B.1.351 (Beta)   \n",
       "14  NODE_0004416  NODE_0005404         B.1.351 (Beta)   \n",
       "15  NODE_0004416  NODE_0000152         B.1.351 (Beta)   \n",
       "16  NODE_0004416  NODE_0005215         B.1.351 (Beta)   \n",
       "17  NODE_0004416  NODE_0006885         B.1.351 (Beta)   \n",
       "18  NODE_0008062  NODE_0005404            P.1 (Gamma)   \n",
       "19  NODE_0008062  NODE_0000152            P.1 (Gamma)   \n",
       "20  NODE_0008062  NODE_0005215            P.1 (Gamma)   \n",
       "21  NODE_0008062  NODE_0006885            P.1 (Gamma)   \n",
       "22  NODE_0005404  NODE_0000152        B.1.1.7 (Alpha)   \n",
       "23  NODE_0005404  NODE_0005215        B.1.1.7 (Alpha)   \n",
       "24  NODE_0005404  NODE_0006885        B.1.1.7 (Alpha)   \n",
       "25  NODE_0000152  NODE_0005215          B.1.525 (Eta)   \n",
       "26  NODE_0000152  NODE_0006885          B.1.525 (Eta)   \n",
       "27  NODE_0005215  NODE_0006885              B.1.1.318   \n",
       "\n",
       "   emerging_lineage_acceptor  support_for_recomb_left  \\\n",
       "0             B.1.526 (Iota)                    False   \n",
       "1             B.1.351 (Beta)                    False   \n",
       "2                P.1 (Gamma)                    False   \n",
       "3            B.1.1.7 (Alpha)                    False   \n",
       "4              B.1.525 (Eta)                    False   \n",
       "5                  B.1.1.318                    False   \n",
       "6              C.37 (Lambda)                    False   \n",
       "7             B.1.351 (Beta)                    False   \n",
       "8                P.1 (Gamma)                    False   \n",
       "9            B.1.1.7 (Alpha)                    False   \n",
       "10             B.1.525 (Eta)                    False   \n",
       "11                 B.1.1.318                    False   \n",
       "12             C.37 (Lambda)                    False   \n",
       "13               P.1 (Gamma)                    False   \n",
       "14           B.1.1.7 (Alpha)                    False   \n",
       "15             B.1.525 (Eta)                    False   \n",
       "16                 B.1.1.318                    False   \n",
       "17             C.37 (Lambda)                    False   \n",
       "18           B.1.1.7 (Alpha)                    False   \n",
       "19             B.1.525 (Eta)                    False   \n",
       "20                 B.1.1.318                    False   \n",
       "21             C.37 (Lambda)                    False   \n",
       "22             B.1.525 (Eta)                    False   \n",
       "23                 B.1.1.318                    False   \n",
       "24             C.37 (Lambda)                    False   \n",
       "25                 B.1.1.318                    False   \n",
       "26             C.37 (Lambda)                    False   \n",
       "27             C.37 (Lambda)                    False   \n",
       "\n",
       "    support_for_recomb_right possible_window_of_recomb  on_deletion_branch  \n",
       "0                      False             [9867, 15324]                True  \n",
       "1                      False            [10323, 15324]                True  \n",
       "2                      False             [6613, 12778]                True  \n",
       "3                      False             [6954, 14676]                True  \n",
       "4                      False             [8593, 14407]                True  \n",
       "5                      False             [9891, 15324]                True  \n",
       "6                      False            [10097, 13536]                True  \n",
       "7                      False            [10323, 20262]                True  \n",
       "8                      False             [9867, 12778]                True  \n",
       "9                      False             [9867, 14676]                True  \n",
       "10                     False             [9867, 14407]                True  \n",
       "11                     False             [9891, 20262]                True  \n",
       "12                     False            [10097, 13536]                True  \n",
       "13                     False            [10323, 12778]                True  \n",
       "14                     False            [10323, 14676]                True  \n",
       "15                     False            [10323, 14407]                True  \n",
       "16                     False            [10323, 20578]                True  \n",
       "17                     False            [10323, 13536]                True  \n",
       "18                     False             [6954, 12778]                True  \n",
       "19                     False             [8593, 12778]                True  \n",
       "20                     False             [9891, 12778]                True  \n",
       "21                     False            [10097, 12778]                True  \n",
       "22                     False             [8593, 14407]                True  \n",
       "23                     False             [9891, 14676]                True  \n",
       "24                     False            [10097, 13536]                True  \n",
       "25                     False             [9891, 14407]                True  \n",
       "26                     False            [10097, 13536]                True  \n",
       "27                     False            [10097, 13536]                True  "
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "possible_recomb_df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-kernel",
   "language": "python",
   "name": "ncov-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
