{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "75a377e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from utils import sars2_genome_info, get_parent, add_mut_accumulation_attr\n",
    "from utils_plotting import get_color_palette, convert_linege_names, get_linear_reg_stats\n",
    "from augur.utils import json_to_tree\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as patches\n",
    "import matplotlib.gridspec as gridspec\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "from scipy import stats\n",
    "import requests\n",
    "import random\n",
    "import math\n",
    "import json\n",
    "import re"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21d9d288",
   "metadata": {},
   "source": [
    "## Table 1: Accumulation of nonsynonymous mutations across the entire genome\n",
    "\n",
    "#### Table 1 gives the same information as Figure 1 except for every gene in the SARS-CoV-2 genome\n",
    "\n",
    "##### The accumulation of mutations in different regions of the genome is counted and the slope of mutation accumulation over time is used to give a rate of mutation accumulation\n",
    "\n",
    "##### Mutation accumulation is then compared to logistic growth rate and the R-value of this correlation and p-value of that R-value are reported\n",
    "\n",
    "#### This notebook completes the analysis for Table 1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63bfe45",
   "metadata": {},
   "source": [
    "Import the tree and convert to Bio Phylo format. This is a time-resolved phylogeny built from 9544 SARS-CoV-2 genomes sampled between December 2019 and May 15, 2021. The tree can be viewed at https://nextstrain.org/groups/blab/ncov/adaptive-evolution/2021-05-15"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d644857e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_url = 'https://nextstrain-blab.s3.amazonaws.com/ncov_adaptive-evolution_2021-05-15.json'\n",
    "\n",
    "tree_json = requests.get(tree_url).json()\n",
    "\n",
    "#Put tree in Bio.Phylo format\n",
    "tree = json_to_tree(tree_json)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "94ff0346",
   "metadata": {},
   "source": [
    "Get information about genome position and length of each gene "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f3836cd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_gene_locations, reference_gene_codon, gene_lengths_aa = sars2_genome_info()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74037b5c",
   "metadata": {},
   "source": [
    "Add an attribute to each node that gives the total number of mutations (synonymous SNPs, or nonsynonymous SNPs plus deletions) accumulated between the tree root and that node (including mutations on the node). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b96cfa73",
   "metadata": {},
   "outputs": [],
   "source": [
    "tree = add_mut_accumulation_attr(tree)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2942e915",
   "metadata": {},
   "source": [
    "List all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "caf9ac92",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes = ['Nsp1', 'Nsp2', 'Nsp3', 'Nsp4', 'Nsp5', 'Nsp6', 'Nsp7', 'Nsp8', 'Nsp9', \n",
    "             'Nsp10', 'RdRp', 'Nsp13', 'Nsp14', 'Nsp15', 'Nsp16', 'S1', 'S2', 'ORF3a', 'E', \n",
    "             'M', 'ORF6', 'ORF7a', 'ORF7b', 'ORF8', 'N', 'ORF9b']\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e4bde020",
   "metadata": {},
   "source": [
    "Now make a table that includes EVERY gene. Table will include rate of mutation accumulation and stats for the correlation between mutation accumulation and logistic growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f2f22549",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes_information = []\n",
    "\n",
    "# only look at internal nodes\n",
    "for node in tree.find_clades(terminal=False):\n",
    "\n",
    "    # only nodes within 6 weeks of May 15 have logisitic growth rates\n",
    "    logistic_growth = None\n",
    "    if \"logistic_growth\" in node.node_attrs:\n",
    "        logistic_growth = node.node_attrs[\"logistic_growth\"][\"value\"]\n",
    "    \n",
    "    if hasattr(node, \"node_attrs\"):\n",
    "        # get inferred node date\n",
    "        date = node.node_attrs[\"num_date\"][\"value\"]\n",
    "        \n",
    "        # get emerging lineage assignment of node \n",
    "        if 'emerging_lineage' in node.node_attrs:\n",
    "            emerging_lineage = node.node_attrs['emerging_lineage']['value']\n",
    "            emerging_lineage = convert_linege_names(emerging_lineage)\n",
    "\n",
    "        # make tidy df for seaborn plotting\n",
    "        for r in all_genes:\n",
    "            # name of attr for nonsyn mutations in this gene\n",
    "            accumulation_attr = f'{r.lower()}_accumulation'\n",
    "            # there are no synonymous deletions, but S1 deletions will be stored with S1_syn mut information-> these will not be plotted\n",
    "            all_genes_information.append({'clade': node.name, 'date': date,\n",
    "                                     'emerging_lineage': emerging_lineage,\n",
    "                                     'mut_location': r, \n",
    "                                     'num_muts': node.node_attrs[accumulation_attr],  \n",
    "                                     'muts_per_codon': node.node_attrs[accumulation_attr]/gene_lengths_aa[r],\n",
    "                                     'logistic_growth': logistic_growth})\n",
    "\n",
    "\n",
    "# make list into dataframe\n",
    "all_genes_df = pd.DataFrame(all_genes_information)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "60fb977a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def all_genes_table(data_files):\n",
    "    \"\"\"\n",
    "    Make table listing rate of mutation accumulation in each gene, \n",
    "    the R-value for the correlation between mutation accumulation and logistic growth rate, \n",
    "    and the p-value for this r-value compared to randomized trees\n",
    "    \"\"\"\n",
    "    # initialize dataframe to store information from all .json files where randomized data is stored\n",
    "    expected_df = pd.DataFrame(columns = ['data', 'iteration', 'gene', 'nonsyn_syn', 'r_value'])\n",
    "    \n",
    "    # read in all randomization R values\n",
    "    for data_file in data_files:\n",
    "        with open(data_file) as json_file:\n",
    "            data = json.load(json_file)\n",
    "            df = pd.DataFrame(data['growth_rate_stats'])\n",
    "            expected_df = pd.concat([expected_df, df])\n",
    "    \n",
    "    table1_stats = []\n",
    "    \n",
    "    for r in all_genes:\n",
    "        # get expected r-values for this gene\n",
    "        expected_gene_data = expected_df[(expected_df['gene']==r) & (expected_df['nonsyn_syn']=='nonsyn')]\n",
    "        \n",
    "        num_iterations = len(expected_gene_data)\n",
    "        \n",
    "        # get the mutation accumulation rate for this gene\n",
    "        mut_accumulation_rate, rv = get_linear_reg_stats(all_genes_df, r, \n",
    "                                              'date', 'muts_per_codon')\n",
    "        \n",
    "        # get the empirical r-value for the correlation between mut accumulation and logisitic growth rate\n",
    "        s, observed_r_value = get_linear_reg_stats(all_genes_df, r, \n",
    "                                              'logistic_growth', 'muts_per_codon')\n",
    "        \n",
    "        # find the p-value \n",
    "        pvalue= len(expected_gene_data[expected_gene_data['r_value']>observed_r_value])/num_iterations\n",
    "        \n",
    "        # make pvalue legible for table\n",
    "        if pvalue<0.005:\n",
    "            pvalue = '<0.005'\n",
    "        else:\n",
    "            pvalue = round(pvalue, 2)\n",
    "        \n",
    "        table1_stats.append({'gene': r, \n",
    "                             'mut_accumulation_rate': f'{format(mut_accumulation_rate, \"10.2E\")}', \n",
    "                            'mut_growth_r_value':observed_r_value, \n",
    "                             'pvalue': pvalue})\n",
    "        \n",
    "    table1 = pd.DataFrame(table1_stats)\n",
    "        \n",
    "    return table1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "80e15779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>mut_accumulation_rate</th>\n",
       "      <th>mut_growth_r_value</th>\n",
       "      <th>pvalue</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Nsp1</td>\n",
       "      <td>1.00E-04</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.43</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Nsp2</td>\n",
       "      <td>2.00E-04</td>\n",
       "      <td>-0.10</td>\n",
       "      <td>0.89</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Nsp3</td>\n",
       "      <td>1.30E-03</td>\n",
       "      <td>0.30</td>\n",
       "      <td>0.09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Nsp4</td>\n",
       "      <td>8.00E-04</td>\n",
       "      <td>0.28</td>\n",
       "      <td>0.06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Nsp5</td>\n",
       "      <td>5.00E-04</td>\n",
       "      <td>-0.09</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Nsp6</td>\n",
       "      <td>3.40E-03</td>\n",
       "      <td>0.35</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Nsp7</td>\n",
       "      <td>3.00E-04</td>\n",
       "      <td>-0.42</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Nsp8</td>\n",
       "      <td>3.00E-04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Nsp9</td>\n",
       "      <td>7.00E-04</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Nsp10</td>\n",
       "      <td>1.00E-04</td>\n",
       "      <td>-0.05</td>\n",
       "      <td>0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>RdRp</td>\n",
       "      <td>5.00E-04</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Nsp13</td>\n",
       "      <td>8.00E-04</td>\n",
       "      <td>0.20</td>\n",
       "      <td>0.15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Nsp14</td>\n",
       "      <td>3.00E-04</td>\n",
       "      <td>-0.03</td>\n",
       "      <td>0.69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Nsp15</td>\n",
       "      <td>3.00E-04</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Nsp16</td>\n",
       "      <td>3.00E-04</td>\n",
       "      <td>-0.22</td>\n",
       "      <td>0.96</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>S1</td>\n",
       "      <td>8.40E-03</td>\n",
       "      <td>0.46</td>\n",
       "      <td>&lt;0.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S2</td>\n",
       "      <td>3.50E-03</td>\n",
       "      <td>0.24</td>\n",
       "      <td>0.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ORF3a</td>\n",
       "      <td>1.80E-03</td>\n",
       "      <td>-0.06</td>\n",
       "      <td>0.86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>E</td>\n",
       "      <td>1.60E-03</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.39</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M</td>\n",
       "      <td>7.00E-04</td>\n",
       "      <td>0.29</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ORF6</td>\n",
       "      <td>2.00E-04</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ORF7a</td>\n",
       "      <td>1.50E-03</td>\n",
       "      <td>0.43</td>\n",
       "      <td>&lt;0.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ORF7b</td>\n",
       "      <td>1.70E-03</td>\n",
       "      <td>0.22</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ORF8</td>\n",
       "      <td>1.65E-02</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>N</td>\n",
       "      <td>6.10E-03</td>\n",
       "      <td>0.21</td>\n",
       "      <td>0.22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ORF9b</td>\n",
       "      <td>2.10E-03</td>\n",
       "      <td>0.27</td>\n",
       "      <td>0.05</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene mut_accumulation_rate  mut_growth_r_value  pvalue\n",
       "0    Nsp1              1.00E-04                0.05    0.43\n",
       "1    Nsp2              2.00E-04               -0.10    0.89\n",
       "2    Nsp3              1.30E-03                0.30    0.09\n",
       "3    Nsp4              8.00E-04                0.28    0.06\n",
       "4    Nsp5              5.00E-04               -0.09    0.84\n",
       "5    Nsp6              3.40E-03                0.35    0.01\n",
       "6    Nsp7              3.00E-04               -0.42       1\n",
       "7    Nsp8              3.00E-04                0.04    0.44\n",
       "8    Nsp9              7.00E-04                0.06     0.3\n",
       "9   Nsp10              1.00E-04               -0.05    0.81\n",
       "10   RdRp              5.00E-04                0.15    0.26\n",
       "11  Nsp13              8.00E-04                0.20    0.15\n",
       "12  Nsp14              3.00E-04               -0.03    0.69\n",
       "13  Nsp15              3.00E-04                0.15    0.23\n",
       "14  Nsp16              3.00E-04               -0.22    0.96\n",
       "15     S1              8.40E-03                0.46  <0.005\n",
       "16     S2              3.50E-03                0.24     0.1\n",
       "17  ORF3a              1.80E-03               -0.06    0.86\n",
       "18      E              1.60E-03                0.03    0.39\n",
       "19      M              7.00E-04                0.29    0.04\n",
       "20   ORF6              2.00E-04                0.01     0.5\n",
       "21  ORF7a              1.50E-03                0.43  <0.005\n",
       "22  ORF7b              1.70E-03                0.22    0.04\n",
       "23   ORF8              1.65E-02                0.19    0.19\n",
       "24      N              6.10E-03                0.21    0.22\n",
       "25  ORF9b              2.10E-03                0.27    0.05"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_genes_table(['growth_rate_stats/growth_stats_50it.json', \n",
    "            'growth_rate_stats/growth_stats_50itmore.json', \n",
    "            'growth_rate_stats/growth_stats_200its.json', \n",
    "            'growth_rate_stats/growth_stats_400its.json', \n",
    "            'growth_rate_stats/growth_stats_300its.json',\n",
    "                 'growth_rate_stats/growth_stats-supp_100its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_150its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_250its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_300its.json', \n",
    "                  'growth_rate_stats/growth_stats-supp_200itsmore.json', \n",
    "                 'growth_rate_stats/growth_stats-other_orfs_10its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_orfs_100its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_orfs_200its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_orfs_300its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_nsps_10its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_nsps_100its.json',\n",
    "                 'growth_rate_stats/growth_stats-other_nsps_200its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_nsps_300its.json', \n",
    "                 'growth_rate_stats/growth_stats-other_nsps_250its.json'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "125dc26b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ncov-kernel",
   "language": "python",
   "name": "ncov-kernel"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
